// openclaw/dist/extensions/google/web-search-contract-api.js
import { a as setProviderWebSearchPluginConfigValue, i as resolveProviderWebSearchPluginConfig, o as setScopedCredentialValue, t as getScopedCredentialValue } from "../../web-search-provider-config-Bzw58qrQ.js";
import "../../provider-web-search-contract-BlcSOQwq.js";
//#region extensions/google/web-search-contract-api.ts
function createGeminiWebSearchProvider() {
	return {
		id: "gemini",
		label: "Gemini (Google Search)",
		hint: "Requires Google Gemini API key · Google Search grounding",
		onboardingScopes: ["text-inference"],
		credentialLabel: "Google Gemini API key",
		envVars: ["GEMINI_API_KEY"],
		placeholder: "AIza...",
		signupUrl: "https://aistudio.google.com/apikey",
		docsUrl: "https://docs.openclaw.ai/tools/web",
		autoDetectOrder: 20,
		credentialPath: "plugins.entries.google.config.webSearch.apiKey",
		inactiveSecretPaths: ["plugins.entries.google.config.webSearch.apiKey"],
		getCredentialValue: (searchConfig) => getScopedCredentialValue(searchConfig, "gemini"),
		setCredentialValue: (searchConfigTarget, value) => setScopedCredentialValue(searchConfigTarget, "gemini", value),
		getConfiguredCredentialValue: (config) => resolveProviderWebSearchPluginConfig(config, "google")?.apiKey,
		setConfiguredCredentialValue: (configTarget, value) => {
			setProviderWebSearchPluginConfigValue(configTarget, "google", "apiKey", value);
		},
		createTool: () => null
	};
}
//#endregion
export { createGeminiWebSearchProvider };
