// openclaw/dist/extensions/slack/contract-api.js
import { n as vi } from "../../test.p_J6dB8a-QET2Xqom.js";
import { A as primeChannelOutboundSendMock } from "../../testing-DtIZqkvR.js";
import { t as slackOutbound } from "../../outbound-adapter-CU77JWY-.js";
import { t as collectSlackSecurityAuditFindings } from "../../security-audit-DqHtSt0-.js";
import { n as normalizeCompatibilityConfig, t as legacyConfigRules } from "../../doctor-contract-Byp5_Sbd.js";
import { n as collectRuntimeConfigAssignments, r as secretTargetRegistryEntries } from "../../secret-contract-BF-zlPPn.js";
//#region extensions/slack/src/outbound-payload-harness.ts
function createSlackOutboundPayloadHarness(params) {
	const sendSlack = vi.fn();
	primeChannelOutboundSendMock(sendSlack, {
		messageId: "sl-1",
		channelId: "C12345",
		ts: "1234.5678"
	}, params.sendResults);
	const ctx = {
		cfg: {},
		to: "C12345",
		text: "",
		payload: params.payload,
		deps: { sendSlack }
	};
	return {
		run: async () => await slackOutbound.sendPayload(ctx),
		sendMock: sendSlack,
		to: ctx.to
	};
}
//#endregion
export { collectRuntimeConfigAssignments, collectSlackSecurityAuditFindings, createSlackOutboundPayloadHarness, legacyConfigRules, normalizeCompatibilityConfig, secretTargetRegistryEntries };
