// openclaw/dist/extensions/minimax/web-search-contract-api.js
import { a as setProviderWebSearchPluginConfigValue, i as resolveProviderWebSearchPluginConfig } from "../../web-search-provider-config-Bzw58qrQ.js";
import "../../provider-web-search-contract-BlcSOQwq.js";
//#region extensions/minimax/web-search-contract-api.ts
const MINIMAX_CODING_PLAN_ENV_VARS = ["MINIMAX_CODE_PLAN_KEY", "MINIMAX_CODING_API_KEY"];
function getTopLevelCredentialValue(searchConfig) {
	return searchConfig?.apiKey;
}
function setTopLevelCredentialValue(searchConfigTarget, value) {
	searchConfigTarget.apiKey = value;
}
function createMiniMaxWebSearchProvider() {
	return {
		id: "minimax",
		label: "MiniMax Search",
		hint: "Structured results via MiniMax Coding Plan search API",
		credentialLabel: "MiniMax Coding Plan key",
		envVars: [...MINIMAX_CODING_PLAN_ENV_VARS],
		placeholder: "sk-cp-...",
		signupUrl: "https://platform.minimax.io/user-center/basic-information/interface-key",
		docsUrl: "https://docs.openclaw.ai/tools/minimax-search",
		autoDetectOrder: 15,
		credentialPath: "plugins.entries.minimax.config.webSearch.apiKey",
		inactiveSecretPaths: ["plugins.entries.minimax.config.webSearch.apiKey"],
		getCredentialValue: getTopLevelCredentialValue,
		setCredentialValue: setTopLevelCredentialValue,
		getConfiguredCredentialValue: (config) => resolveProviderWebSearchPluginConfig(config, "minimax")?.apiKey,
		setConfiguredCredentialValue: (configTarget, value) => {
			setProviderWebSearchPluginConfigValue(configTarget, "minimax", "apiKey", value);
		},
		createTool: () => null
	};
}
//#endregion
export { createMiniMaxWebSearchProvider };
