// openclaw/dist/extensions/tavily/web-search-contract-api.js
import { t as enablePluginInConfig } from "../../provider-enable-config-Hm2ULiVx.js";
import { a as setProviderWebSearchPluginConfigValue, i as resolveProviderWebSearchPluginConfig, o as setScopedCredentialValue, t as getScopedCredentialValue } from "../../web-search-provider-config-Bzw58qrQ.js";
import "../../provider-web-search-contract-BlcSOQwq.js";
//#region extensions/tavily/web-search-contract-api.ts
function createTavilyWebSearchProvider() {
	return {
		id: "tavily",
		label: "Tavily Search",
		hint: "Structured results with domain filters and AI answer summaries",
		onboardingScopes: ["text-inference"],
		credentialLabel: "Tavily API key",
		envVars: ["TAVILY_API_KEY"],
		placeholder: "tvly-...",
		signupUrl: "https://tavily.com/",
		docsUrl: "https://docs.openclaw.ai/tools/tavily",
		autoDetectOrder: 70,
		credentialPath: "plugins.entries.tavily.config.webSearch.apiKey",
		inactiveSecretPaths: ["plugins.entries.tavily.config.webSearch.apiKey"],
		getCredentialValue: (searchConfig) => getScopedCredentialValue(searchConfig, "tavily"),
		setCredentialValue: (searchConfigTarget, value) => setScopedCredentialValue(searchConfigTarget, "tavily", value),
		getConfiguredCredentialValue: (config) => resolveProviderWebSearchPluginConfig(config, "tavily")?.apiKey,
		setConfiguredCredentialValue: (configTarget, value) => {
			setProviderWebSearchPluginConfigValue(configTarget, "tavily", "apiKey", value);
		},
		applySelectionConfig: (config) => enablePluginInConfig(config, "tavily").config,
		createTool: () => null
	};
}
//#endregion
export { createTavilyWebSearchProvider };
