// openclaw/dist/cli.runtime-CfvPn6GT.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import "./error-runtime-OaZzlNpe.js";
import { T as normalizeQaProviderMode, c as startQaMockOpenAiServer, f as startQaLabServer, n as resolveQaLiveTurnTimeoutMs, s as startQaGatewayChild, t as runQaSuite, w as defaultQaModelForMode } from "./suite-BSxkaTP1.js";
import { n as writeQaDockerHarnessFiles, t as buildQaDockerHarnessImage } from "./docker-harness-DCSoIQSh.js";
import path from "node:path";
import { execFile } from "node:child_process";
import { randomUUID } from "node:crypto";
import { createServer } from "node:net";
import { setTimeout } from "node:timers/promises";
//#region extensions/qa-lab/src/docker-up.runtime.ts
function resolveDefaultQaDockerDir(repoRoot) {
	return path.resolve(repoRoot, ".artifacts/qa-docker");
}
function describeError(error) {
	if (error instanceof Error) return error.message;
	if (typeof error === "string") return error;
	return JSON.stringify(error);
}
async function isPortFree(port) {
	return await new Promise((resolve) => {
		const server = createServer();
		server.once("error", () => resolve(false));
		server.listen(port, "127.0.0.1", () => {
			server.close(() => resolve(true));
		});
	});
}
async function findFreePort() {
	return await new Promise((resolve, reject) => {
		const server = createServer();
		server.once("error", reject);
		server.listen(0, () => {
			const address = server.address();
			if (!address || typeof address === "string") {
				server.close();
				reject(/* @__PURE__ */ new Error("failed to find free port"));
				return;
			}
			server.close((error) => {
				if (error) {
					reject(error);
					return;
				}
				resolve(address.port);
			});
		});
	});
}
async function resolveHostPort(preferredPort, pinned) {
	if (pinned || await isPortFree(preferredPort)) return preferredPort;
	return await findFreePort();
}
function trimCommandOutput(output) {
	const trimmed = output.trim();
	if (!trimmed) return "";
	const lines = trimmed.split("\n");
	return lines.length <= 120 ? trimmed : lines.slice(-120).join("\n");
}
async function execCommand(command, args, cwd) {
	return await new Promise((resolve, reject) => {
		execFile(command, args, {
			cwd,
			encoding: "utf8",
			maxBuffer: 10 * 1024 * 1024
		}, (error, stdout, stderr) => {
			if (error) {
				const renderedStdout = trimCommandOutput(stdout);
				const renderedStderr = trimCommandOutput(stderr);
				reject(new Error([
					`Command failed: ${[command, ...args].join(" ")}`,
					renderedStderr ? `stderr:\n${renderedStderr}` : "",
					renderedStdout ? `stdout:\n${renderedStdout}` : ""
				].filter(Boolean).join("\n\n")));
				return;
			}
			resolve({
				stdout,
				stderr
			});
		});
	});
}
async function waitForHealth(url, deps) {
	const timeoutMs = deps.timeoutMs ?? 36e4;
	const pollMs = deps.pollMs ?? 1e3;
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastError = null;
	while (Date.now() < deadline) {
		try {
			if ((await deps.fetchImpl(url)).ok) return;
			lastError = /* @__PURE__ */ new Error(`Health check returned non-OK for ${url}`);
		} catch (error) {
			lastError = error;
		}
		await deps.sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	const lines = [
		`${deps.label ?? url} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		lastError ? `Last error: ${describeError(lastError)}` : "",
		`Hint: check container logs with \`docker compose -f ${deps.composeFile ?? "<compose-file>"} logs\` and verify the port is not already in use.`
	];
	throw new Error(lines.filter(Boolean).join("\n"));
}
async function isHealthy(url, fetchImpl) {
	try {
		return (await fetchImpl(url)).ok;
	} catch {
		return false;
	}
}
async function waitForDockerServiceHealth(service, composeFile, repoRoot, runCommand, sleepImpl, timeoutMs = 36e4, pollMs = 1e3) {
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastStatus = "unknown";
	while (Date.now() < deadline) {
		try {
			const { stdout } = await runCommand("docker", [
				"compose",
				"-f",
				composeFile,
				"ps",
				"--format",
				"json",
				service
			], repoRoot);
			const row = stdout.trim().split("\n").map((line) => line.trim()).filter(Boolean).map((line) => JSON.parse(line))[0];
			lastStatus = row?.Health ?? row?.State ?? "unknown";
			if (lastStatus === "healthy" || lastStatus === "running") return;
		} catch (error) {
			lastStatus = describeError(error);
		}
		await sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	throw new Error([
		`${service} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		`Last status: ${lastStatus}`,
		`Hint: check container logs with \`docker compose -f ${composeFile} logs ${service}\`.`
	].join("\n"));
}
async function resolveComposeServiceUrl(service, port, composeFile, repoRoot, runCommand) {
	const { stdout: containerStdout } = await runCommand("docker", [
		"compose",
		"-f",
		composeFile,
		"ps",
		"-q",
		service
	], repoRoot);
	const containerId = containerStdout.trim();
	if (!containerId) return null;
	const { stdout: ipStdout } = await runCommand("docker", [
		"inspect",
		"--format",
		"{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}",
		containerId
	], repoRoot);
	const ip = ipStdout.trim();
	if (!ip) return null;
	return `http://${ip}:${port}/`;
}
async function runQaDockerUp(params, deps) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const resolveHostPortImpl = deps?.resolveHostPortImpl ?? resolveHostPort;
	const outputDir = path.resolve(params.outputDir ?? resolveDefaultQaDockerDir(repoRoot));
	const gatewayPort = await resolveHostPortImpl(params.gatewayPort ?? 18789, params.gatewayPort != null);
	const qaLabPort = await resolveHostPortImpl(params.qaLabPort ?? 43124, params.qaLabPort != null);
	const runCommand = deps?.runCommand ?? execCommand;
	const fetchImpl = deps?.fetchImpl ?? (async (input) => {
		return await fetch(input);
	});
	const sleepImpl = deps?.sleepImpl ?? setTimeout;
	if (!params.skipUiBuild) await runCommand("pnpm", ["qa:lab:build"], repoRoot);
	await writeQaDockerHarnessFiles({
		outputDir,
		repoRoot,
		gatewayPort,
		qaLabPort,
		providerBaseUrl: params.providerBaseUrl,
		imageName: params.image,
		usePrebuiltImage: params.usePrebuiltImage,
		bindUiDist: params.bindUiDist,
		includeQaLabUi: true
	});
	const composeFile = path.join(outputDir, "docker-compose.qa.yml");
	try {
		await runCommand("docker", [
			"compose",
			"-f",
			composeFile,
			"down",
			"--remove-orphans"
		], repoRoot);
	} catch {}
	const composeArgs = [
		"compose",
		"-f",
		composeFile,
		"up"
	];
	if (!params.usePrebuiltImage) composeArgs.push("--build");
	composeArgs.push("-d");
	await runCommand("docker", composeArgs, repoRoot);
	await sleepImpl(3e3);
	const qaLabUrl = `http://127.0.0.1:${qaLabPort}`;
	const hostGatewayUrl = `http://127.0.0.1:${gatewayPort}/`;
	await waitForHealth(`${qaLabUrl}/healthz`, {
		label: "QA Lab",
		fetchImpl,
		sleepImpl,
		composeFile
	});
	await waitForDockerServiceHealth("openclaw-qa-gateway", composeFile, repoRoot, runCommand, sleepImpl);
	let gatewayUrl = hostGatewayUrl;
	if (!await isHealthy(`${hostGatewayUrl}healthz`, fetchImpl)) {
		const containerGatewayUrl = await resolveComposeServiceUrl("openclaw-qa-gateway", 18789, composeFile, repoRoot, runCommand);
		if (containerGatewayUrl && await isHealthy(`${containerGatewayUrl}healthz`, fetchImpl)) gatewayUrl = containerGatewayUrl;
	}
	return {
		outputDir,
		composeFile,
		qaLabUrl,
		gatewayUrl,
		stopCommand: `docker compose -f ${composeFile} down`
	};
}
//#endregion
//#region extensions/qa-lab/src/manual-lane.runtime.ts
function resolveManualLaneTimeoutMs(params) {
	if (typeof params.timeoutMs === "number" && Number.isFinite(params.timeoutMs) && params.timeoutMs > 0) return params.timeoutMs;
	return resolveQaLiveTurnTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel
	}, 12e4, params.primaryModel);
}
async function runQaManualLane(params) {
	const sessionSuffix = params.primaryModel.replace(/[^a-z0-9._-]+/gi, "-");
	const lab = await startQaLabServer({
		repoRoot: params.repoRoot,
		embeddedGateway: "disabled"
	});
	const mock = params.providerMode === "mock-openai" ? await startQaMockOpenAiServer({
		host: "127.0.0.1",
		port: 0
	}) : null;
	const gateway = await startQaGatewayChild({
		repoRoot: params.repoRoot,
		providerBaseUrl: mock ? `${mock.baseUrl}/v1` : void 0,
		qaBusBaseUrl: lab.listenUrl,
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		fastMode: params.fastMode,
		controlUiEnabled: false
	});
	const timeoutMs = resolveManualLaneTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		timeoutMs: params.timeoutMs
	});
	try {
		const started = await gateway.call("agent", {
			idempotencyKey: randomUUID(),
			agentId: "qa",
			sessionKey: `agent:qa:manual:${sessionSuffix}`,
			message: params.message,
			deliver: true,
			channel: "qa-channel",
			to: "dm:qa-operator",
			replyChannel: "qa-channel",
			replyTo: "dm:qa-operator"
		}, { timeoutMs: 3e4 });
		if (!started.runId) throw new Error(`agent call did not return a runId: ${JSON.stringify(started)}`);
		const waited = await gateway.call("agent.wait", {
			runId: started.runId,
			timeoutMs
		}, { timeoutMs: timeoutMs + 5e3 });
		await setTimeout(500);
		const reply = lab.state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator").at(-1)?.text ?? null;
		return {
			model: params.primaryModel,
			waited,
			reply,
			watchUrl: lab.baseUrl
		};
	} catch (error) {
		throw new Error(formatErrorMessage(error), { cause: error });
	} finally {
		await gateway.stop();
		await mock?.stop();
		await lab.stop();
	}
}
//#endregion
//#region extensions/qa-lab/src/cli.runtime.ts
function resolveQaManualLaneModels(opts) {
	const primaryModel = opts.primaryModel?.trim() || defaultQaModelForMode(opts.providerMode);
	const alternateModel = opts.alternateModel?.trim();
	return {
		primaryModel,
		alternateModel: alternateModel && alternateModel.length > 0 ? alternateModel : opts.primaryModel?.trim() ? primaryModel : defaultQaModelForMode(opts.providerMode, true)
	};
}
async function runInterruptibleServer(label, server) {
	process.stdout.write(`${label}: ${server.baseUrl}\n`);
	process.stdout.write("Press Ctrl+C to stop.\n");
	const shutdown = async () => {
		process.off("SIGINT", onSignal);
		process.off("SIGTERM", onSignal);
		await server.stop();
		process.exit(0);
	};
	const onSignal = () => {
		shutdown();
	};
	process.on("SIGINT", onSignal);
	process.on("SIGTERM", onSignal);
	await new Promise(() => void 0);
}
async function runQaLabSelfCheckCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const server = await startQaLabServer({
		repoRoot,
		outputPath: opts.output ? path.resolve(repoRoot, opts.output) : void 0
	});
	try {
		const result = await server.runSelfCheck();
		process.stdout.write(`QA self-check report: ${result.outputPath}\n`);
	} finally {
		await server.stop();
	}
}
async function runQaSuiteCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const providerMode = normalizeQaProviderMode(opts.providerMode);
	const result = await runQaSuite({
		repoRoot,
		outputDir: opts.outputDir ? path.resolve(repoRoot, opts.outputDir) : void 0,
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel,
		fastMode: opts.fastMode,
		scenarioIds: opts.scenarioIds
	});
	process.stdout.write(`QA suite watch: ${result.watchUrl}\n`);
	process.stdout.write(`QA suite report: ${result.reportPath}\n`);
	process.stdout.write(`QA suite summary: ${result.summaryPath}\n`);
}
async function runQaManualLaneCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const providerMode = opts.providerMode === void 0 ? "live-frontier" : normalizeQaProviderMode(opts.providerMode);
	const models = resolveQaManualLaneModels({
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel
	});
	const result = await runQaManualLane({
		repoRoot,
		providerMode,
		primaryModel: models.primaryModel,
		alternateModel: models.alternateModel,
		fastMode: opts.fastMode,
		message: opts.message,
		timeoutMs: opts.timeoutMs
	});
	process.stdout.write(JSON.stringify(result, null, 2));
	process.stdout.write("\n");
}
async function runQaLabUiCommand(opts) {
	await runInterruptibleServer("QA Lab UI", await startQaLabServer({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0,
		advertiseHost: opts.advertiseHost,
		advertisePort: Number.isFinite(opts.advertisePort) ? opts.advertisePort : void 0,
		controlUiUrl: opts.controlUiUrl,
		controlUiToken: opts.controlUiToken,
		controlUiProxyTarget: opts.controlUiProxyTarget,
		uiDistDir: opts.uiDistDir,
		autoKickoffTarget: opts.autoKickoffTarget,
		embeddedGateway: opts.embeddedGateway,
		sendKickoffOnStart: opts.sendKickoffOnStart
	}));
}
async function runQaDockerScaffoldCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const result = await writeQaDockerHarnessFiles({
		outputDir: path.resolve(repoRoot, opts.outputDir),
		repoRoot,
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		imageName: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist
	});
	process.stdout.write(`QA docker scaffold: ${result.outputDir}\n`);
}
async function runQaDockerBuildImageCommand(opts) {
	const result = await buildQaDockerHarnessImage({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		imageName: opts.image
	});
	process.stdout.write(`QA docker image: ${result.imageName}\n`);
}
async function runQaDockerUpCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const result = await runQaDockerUp({
		repoRoot,
		outputDir: opts.outputDir ? path.resolve(repoRoot, opts.outputDir) : void 0,
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		image: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist,
		skipUiBuild: opts.skipUiBuild
	});
	process.stdout.write(`QA docker dir: ${result.outputDir}\n`);
	process.stdout.write(`QA Lab UI: ${result.qaLabUrl}\n`);
	process.stdout.write(`Gateway UI: ${result.gatewayUrl}\n`);
	process.stdout.write(`Stop: ${result.stopCommand}\n`);
}
async function runQaMockOpenAiCommand(opts) {
	await runInterruptibleServer("QA mock OpenAI", await startQaMockOpenAiServer({
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0
	}));
}
//#endregion
export { runQaDockerBuildImageCommand, runQaDockerScaffoldCommand, runQaDockerUpCommand, runQaLabSelfCheckCommand, runQaLabUiCommand, runQaManualLaneCommand, runQaMockOpenAiCommand, runQaSuiteCommand };
