// openclaw/dist/googlechat-BlnQ5Qb0.js
import "./message-channel-BdU8JCBW.js";
import "./types.secrets-CWvOd-Rt.js";
import "./config-schema-CDNbtTnD.js";
import "./setup-helpers-mv9ZHZDG.js";
import "./zod-schema.providers-core-DsqfCroU.js";
import "./common-CynacdnN.js";
import "./fetch-guard-Dp23M272.js";
import "./fetch-BgzO65-j.js";
import { n as resolveChannelGroupRequireMention } from "./group-policy-XJAXMNhx.js";
import "./channel-policy-DM1sGttY.js";
import "./dm-policy-shared-CXuIpNws.js";
import "./setup-wizard-helpers-BSnJD8SD.js";
import "./channel-reply-pipeline-B9vwXaxt.js";
import "./channel-pairing-CoLYYHOY.js";
import "./status-helpers-6o_CyIi4.js";
import "./webhook-ingress-CIE_jq7u.js";
import { t as createOptionalChannelSetupSurface } from "./channel-setup-URAkKcoj.js";
import "./web-media-B4SISn9h.js";
import "./outbound-media-CQRoQwJA.js";
//#region src/plugin-sdk/googlechat.ts
function resolveGoogleChatGroupRequireMention(params) {
	return resolveChannelGroupRequireMention({
		cfg: params.cfg,
		channel: "googlechat",
		groupId: params.groupId,
		accountId: params.accountId
	});
}
const googlechatSetup = createOptionalChannelSetupSurface({
	channel: "googlechat",
	label: "Google Chat",
	npmSpec: "@openclaw/googlechat",
	docsPath: "/channels/googlechat"
});
const googlechatSetupAdapter = googlechatSetup.setupAdapter;
const googlechatSetupWizard = googlechatSetup.setupWizard;
//#endregion
export { googlechatSetupWizard as n, resolveGoogleChatGroupRequireMention as r, googlechatSetupAdapter as t };
