// openclaw/dist/outbound-adapter-CU77JWY-.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { t as getGlobalHookRunner } from "./hook-runner-global-yij_VaUO.js";
import { _ as sendPayloadMediaSequenceAndFinalize, b as sendTextMediaPayload, f as resolvePayloadMediaUrls } from "./reply-payload-BRozqdR-.js";
import { o as resolveInteractiveTextFallback } from "./payload-DTU7wrd-.js";
import { t as resolveOutboundSendDep } from "./send-deps-BydTP1mg.js";
import "./text-runtime-Y7jeuzOm.js";
import "./outbound-runtime-HUYrqdL1.js";
import "./plugin-runtime-Wce2LD4Z.js";
import { i as createAttachedChannelResultAdapter, t as attachChannelToResult } from "./channel-send-result-COHJCNM4.js";
import { a as resolveSlackAccount } from "./accounts-D4W0Kgk6.js";
import { t as compileSlackInteractiveReplies } from "./interactive-replies-CSBBnoJ5.js";
import { n as parseSlackBlocksInput } from "./blocks-input-Chs9y5IW.js";
import { r as buildSlackInteractiveBlocks } from "./blocks-render-biOTS5cN.js";
import { t as SLACK_TEXT_LIMIT } from "./limits-xKi6QpqB.js";
//#region extensions/slack/src/outbound-adapter.ts
const SLACK_MAX_BLOCKS = 50;
let slackSendRuntimePromise;
async function loadSlackSendRuntime() {
	slackSendRuntimePromise ??= import("./send.runtime-DVgJQLrA.js");
	return await slackSendRuntimePromise;
}
function resolveRenderedInteractiveBlocks(interactive) {
	if (!interactive) return;
	const blocks = buildSlackInteractiveBlocks(interactive);
	return blocks.length > 0 ? blocks : void 0;
}
function resolveSlackSendIdentity(identity) {
	if (!identity) return;
	const username = normalizeOptionalString(identity.name);
	const iconUrl = normalizeOptionalString(identity.avatarUrl);
	const rawEmoji = normalizeOptionalString(identity.emoji);
	const iconEmoji = !iconUrl && rawEmoji && /^:[^:\s]+:$/.test(rawEmoji) ? rawEmoji : void 0;
	if (!username && !iconUrl && !iconEmoji) return;
	return {
		username,
		iconUrl,
		iconEmoji
	};
}
async function applySlackMessageSendingHooks(params) {
	const hookRunner = getGlobalHookRunner();
	if (!hookRunner?.hasHooks("message_sending")) return {
		cancelled: false,
		text: params.text
	};
	const account = resolveSlackAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	const hookResult = await hookRunner.runMessageSending({
		to: params.to,
		content: params.text,
		metadata: {
			threadTs: params.threadTs,
			channelId: params.to,
			...params.mediaUrl ? { mediaUrl: params.mediaUrl } : {}
		}
	}, {
		channelId: "slack",
		accountId: account.accountId
	});
	if (hookResult?.cancel) return {
		cancelled: true,
		text: params.text
	};
	return {
		cancelled: false,
		text: hookResult?.content ?? params.text
	};
}
async function sendSlackOutboundMessage(params) {
	const send = resolveOutboundSendDep(params.deps, "slack") ?? (await loadSlackSendRuntime()).sendMessageSlack;
	const threadTs = params.replyToId ?? (params.threadId != null ? String(params.threadId) : void 0);
	const hookResult = await applySlackMessageSendingHooks({
		cfg: params.cfg,
		to: params.to,
		text: params.text,
		threadTs,
		mediaUrl: params.mediaUrl,
		accountId: params.accountId ?? void 0
	});
	if (hookResult.cancelled) return {
		messageId: "cancelled-by-hook",
		channelId: params.to,
		meta: { cancelled: true }
	};
	const slackIdentity = resolveSlackSendIdentity(params.identity);
	return await send(params.to, hookResult.text, {
		cfg: params.cfg,
		threadTs,
		accountId: params.accountId ?? void 0,
		...params.mediaUrl ? {
			mediaUrl: params.mediaUrl,
			mediaAccess: params.mediaAccess,
			mediaLocalRoots: params.mediaLocalRoots,
			mediaReadFile: params.mediaReadFile
		} : {},
		...params.blocks ? { blocks: params.blocks } : {},
		...slackIdentity ? { identity: slackIdentity } : {}
	});
}
function resolveSlackBlocks(payload) {
	const slackData = payload.channelData?.slack;
	const renderedInteractive = resolveRenderedInteractiveBlocks(payload.interactive);
	if (!slackData || typeof slackData !== "object" || Array.isArray(slackData)) return renderedInteractive;
	const mergedBlocks = [...parseSlackBlocksInput(slackData.blocks) ?? [], ...renderedInteractive ?? []];
	if (mergedBlocks.length === 0) return;
	if (mergedBlocks.length > SLACK_MAX_BLOCKS) throw new Error(`Slack blocks cannot exceed ${SLACK_MAX_BLOCKS} items after interactive render`);
	return mergedBlocks;
}
const slackOutbound = {
	deliveryMode: "direct",
	chunker: null,
	textChunkLimit: SLACK_TEXT_LIMIT,
	normalizePayload: ({ payload }) => compileSlackInteractiveReplies(payload),
	sendPayload: async (ctx) => {
		const payload = {
			...ctx.payload,
			text: resolveInteractiveTextFallback({
				text: ctx.payload.text,
				interactive: ctx.payload.interactive
			}) ?? ""
		};
		const blocks = resolveSlackBlocks(payload);
		if (!blocks) return await sendTextMediaPayload({
			channel: "slack",
			ctx: {
				...ctx,
				payload
			},
			adapter: slackOutbound
		});
		return attachChannelToResult("slack", await sendPayloadMediaSequenceAndFinalize({
			text: "",
			mediaUrls: resolvePayloadMediaUrls(payload),
			send: async ({ text, mediaUrl }) => await sendSlackOutboundMessage({
				cfg: ctx.cfg,
				to: ctx.to,
				text,
				mediaUrl,
				mediaAccess: ctx.mediaAccess,
				mediaLocalRoots: ctx.mediaLocalRoots,
				mediaReadFile: ctx.mediaReadFile,
				accountId: ctx.accountId,
				deps: ctx.deps,
				replyToId: ctx.replyToId,
				threadId: ctx.threadId,
				identity: ctx.identity
			}),
			finalize: async () => await sendSlackOutboundMessage({
				cfg: ctx.cfg,
				to: ctx.to,
				text: payload.text ?? "",
				mediaAccess: ctx.mediaAccess,
				mediaLocalRoots: ctx.mediaLocalRoots,
				mediaReadFile: ctx.mediaReadFile,
				blocks,
				accountId: ctx.accountId,
				deps: ctx.deps,
				replyToId: ctx.replyToId,
				threadId: ctx.threadId,
				identity: ctx.identity
			})
		}));
	},
	...createAttachedChannelResultAdapter({
		channel: "slack",
		sendText: async ({ cfg, to, text, accountId, deps, replyToId, threadId, identity }) => await sendSlackOutboundMessage({
			cfg,
			to,
			text,
			accountId,
			deps,
			replyToId,
			threadId,
			identity
		}),
		sendMedia: async ({ cfg, to, text, mediaUrl, mediaAccess, mediaLocalRoots, mediaReadFile, accountId, deps, replyToId, threadId, identity }) => await sendSlackOutboundMessage({
			cfg,
			to,
			text,
			mediaUrl,
			mediaAccess,
			mediaLocalRoots,
			mediaReadFile,
			accountId,
			deps,
			replyToId,
			threadId,
			identity
		})
	})
};
//#endregion
export { slackOutbound as t };
