// openclaw/dist/plugin-runtime-Wce2LD4Z.js
import { n as resolveGlobalSingleton } from "./global-singleton-CqneY5Pr.js";
import { f as normalizePluginInteractiveNamespace, h as validatePluginInteractiveNamespace, m as toPluginInteractiveRegistryKey, p as resolvePluginInteractiveMatch } from "./types-Bc6OypEK.js";
import "./hook-runner-global-yij_VaUO.js";
import { n as resolveGlobalDedupeCache } from "./dedupe-DdEu-Fo2.js";
import "./lazy-service-module-BzYR2Bhd.js";
import { o as detachPluginConversationBinding, p as requestPluginConversationBinding, s as getCurrentPluginConversationBinding } from "./conversation-binding-CsWJgSK2.js";
import "./commands-BZTb3kyL.js";
import "./http-registry-DEukHr_r.js";
//#region src/plugins/interactive-binding-helpers.ts
function createInteractiveConversationBindingHelpers(params) {
	const { registration, senderId, conversation } = params;
	const pluginRoot = registration.pluginRoot;
	return {
		requestConversationBinding: async (binding = {}) => {
			if (!pluginRoot) return {
				status: "error",
				message: "This interaction cannot bind the current conversation."
			};
			return requestPluginConversationBinding({
				pluginId: registration.pluginId,
				pluginName: registration.pluginName,
				pluginRoot,
				requestedBySenderId: senderId,
				conversation,
				binding
			});
		},
		detachConversationBinding: async () => {
			if (!pluginRoot) return { removed: false };
			return detachPluginConversationBinding({
				pluginRoot,
				conversation
			});
		},
		getCurrentConversationBinding: async () => {
			if (!pluginRoot) return null;
			return getCurrentPluginConversationBinding({
				pluginRoot,
				conversation
			});
		}
	};
}
//#endregion
//#region src/plugins/interactive.ts
const PLUGIN_INTERACTIVE_STATE_KEY = Symbol.for("openclaw.pluginInteractiveState");
const getState = () => resolveGlobalSingleton(PLUGIN_INTERACTIVE_STATE_KEY, () => ({
	interactiveHandlers: /* @__PURE__ */ new Map(),
	callbackDedupe: resolveGlobalDedupeCache(Symbol.for("openclaw.pluginInteractiveCallbackDedupe"), {
		ttlMs: 5 * 6e4,
		maxSize: 4096
	})
}));
const getInteractiveHandlers = () => getState().interactiveHandlers;
const getCallbackDedupe = () => getState().callbackDedupe;
function resolveNamespaceMatch(channel, data) {
	return resolvePluginInteractiveMatch({
		interactiveHandlers: getInteractiveHandlers(),
		channel,
		data
	});
}
function registerPluginInteractiveHandler(pluginId, registration, opts) {
	const interactiveHandlers = getInteractiveHandlers();
	const namespace = normalizePluginInteractiveNamespace(registration.namespace);
	const validationError = validatePluginInteractiveNamespace(namespace);
	if (validationError) return {
		ok: false,
		error: validationError
	};
	const key = toPluginInteractiveRegistryKey(registration.channel, namespace);
	const existing = interactiveHandlers.get(key);
	if (existing) return {
		ok: false,
		error: `Interactive handler namespace "${namespace}" already registered by plugin "${existing.pluginId}"`
	};
	interactiveHandlers.set(key, {
		...registration,
		namespace,
		pluginId,
		pluginName: opts?.pluginName,
		pluginRoot: opts?.pluginRoot
	});
	return { ok: true };
}
function clearPluginInteractiveHandlers() {
	const interactiveHandlers = getInteractiveHandlers();
	const callbackDedupe = getCallbackDedupe();
	interactiveHandlers.clear();
	callbackDedupe.clear();
}
function clearPluginInteractiveHandlersForPlugin(pluginId) {
	const interactiveHandlers = getInteractiveHandlers();
	for (const [key, value] of interactiveHandlers.entries()) if (value.pluginId === pluginId) interactiveHandlers.delete(key);
}
async function dispatchPluginInteractiveHandler(params) {
	const callbackDedupe = getCallbackDedupe();
	const match = resolveNamespaceMatch(params.channel, params.data);
	if (!match) return {
		matched: false,
		handled: false,
		duplicate: false
	};
	const dedupeKey = params.dedupeId?.trim();
	if (dedupeKey && callbackDedupe.peek(dedupeKey)) return {
		matched: true,
		handled: true,
		duplicate: true
	};
	await params.onMatched?.();
	const resolved = await params.invoke(match);
	if (dedupeKey) callbackDedupe.check(dedupeKey);
	return {
		matched: true,
		handled: resolved?.handled ?? true,
		duplicate: false
	};
}
//#endregion
export { createInteractiveConversationBindingHelpers as a, registerPluginInteractiveHandler as i, clearPluginInteractiveHandlersForPlugin as n, dispatchPluginInteractiveHandler as r, clearPluginInteractiveHandlers as t };
