// openclaw/dist/provider-registration-D54ZujPi.js
import { t as buildProviderReplayFamilyHooks } from "./provider-model-shared-kd5jursx.js";
import { t as createProviderApiKeyAuthMethod } from "./provider-api-key-auth-DRImnNV3.js";
import "./provider-auth-api-key-CKMbVrod.js";
import { t as buildProviderStreamFamilyHooks } from "./provider-stream-CNyQJ5TJ.js";
import "./provider-stream-family-CBqsYMnG.js";
import { n as resolveGoogleGeminiForwardCompatModel, t as isModernGoogleModel } from "./provider-models-snK-Ql7-.js";
import { n as normalizeGoogleModelId } from "./model-id-BoMSRzao.js";
import { d as resolveGoogleGenerativeAiTransport, n as GOOGLE_GEMINI_DEFAULT_MODEL, r as applyGoogleGeminiModelDefault, s as normalizeGoogleProviderConfig } from "./api-By0PEIS6.js";
//#region extensions/google/provider-registration.ts
const GOOGLE_GEMINI_PROVIDER_HOOKS = {
	...buildProviderReplayFamilyHooks({ family: "google-gemini" }),
	...buildProviderStreamFamilyHooks("google-thinking")
};
function registerGoogleProvider(api) {
	api.registerProvider({
		id: "google",
		label: "Google AI Studio",
		docsPath: "/providers/models",
		hookAliases: ["google-antigravity", "google-vertex"],
		envVars: ["GEMINI_API_KEY", "GOOGLE_API_KEY"],
		auth: [createProviderApiKeyAuthMethod({
			providerId: "google",
			methodId: "api-key",
			label: "Google Gemini API key",
			hint: "AI Studio / Gemini API key",
			optionKey: "geminiApiKey",
			flagName: "--gemini-api-key",
			envVar: "GEMINI_API_KEY",
			promptMessage: "Enter Gemini API key",
			defaultModel: GOOGLE_GEMINI_DEFAULT_MODEL,
			expectedProviders: ["google"],
			applyConfig: (cfg) => applyGoogleGeminiModelDefault(cfg).next,
			wizard: {
				choiceId: "gemini-api-key",
				choiceLabel: "Google Gemini API key",
				groupId: "google",
				groupLabel: "Google",
				groupHint: "Gemini API key + OAuth"
			}
		})],
		normalizeTransport: ({ api, baseUrl }) => resolveGoogleGenerativeAiTransport({
			api,
			baseUrl
		}),
		normalizeConfig: ({ provider, providerConfig }) => normalizeGoogleProviderConfig(provider, providerConfig),
		normalizeModelId: ({ modelId }) => normalizeGoogleModelId(modelId),
		resolveDynamicModel: (ctx) => resolveGoogleGeminiForwardCompatModel({
			providerId: ctx.provider,
			ctx
		}),
		...GOOGLE_GEMINI_PROVIDER_HOOKS,
		isModernModelRef: ({ modelId }) => isModernGoogleModel(modelId)
	});
}
//#endregion
export { registerGoogleProvider as t };
