// openclaw/dist/qa-gateway-config-Mpw8qadZ.js
import { a as splitQaModelRef, i as normalizeQaProviderMode, r as isQaFastModeModelRef, t as defaultQaModelForMode } from "./model-selection-DwwC92P1.js";
//#region extensions/qa-lab/src/qa-gateway-config.ts
const DISABLED_BUNDLED_CHANNELS = Object.freeze({
	bluebubbles: { enabled: false },
	discord: { enabled: false },
	feishu: { enabled: false },
	googlechat: { enabled: false },
	imessage: { enabled: false },
	irc: { enabled: false },
	line: { enabled: false },
	mattermost: { enabled: false },
	matrix: { enabled: false },
	msteams: { enabled: false },
	qqbot: { enabled: false },
	signal: { enabled: false },
	slack: { enabled: false },
	"synology-chat": { enabled: false },
	telegram: { enabled: false },
	tlon: { enabled: false },
	whatsapp: { enabled: false },
	zalo: { enabled: false },
	zalouser: { enabled: false }
});
const DEFAULT_QA_CONTROL_UI_ALLOWED_ORIGINS = Object.freeze([
	"http://127.0.0.1:18789",
	"http://localhost:18789",
	"http://127.0.0.1:43124",
	"http://localhost:43124"
]);
function mergeQaControlUiAllowedOrigins(extraOrigins) {
	const normalizedExtra = (extraOrigins ?? []).map((origin) => origin.trim()).filter((origin) => origin.length > 0);
	return [...new Set([...DEFAULT_QA_CONTROL_UI_ALLOWED_ORIGINS, ...normalizedExtra])];
}
function buildQaGatewayConfig(params) {
	const mockOpenAiProvider = {
		baseUrl: params.providerBaseUrl ?? "http://127.0.0.1:44080/v1",
		apiKey: "test",
		api: "openai-responses",
		models: [
			{
				id: "gpt-5.4",
				name: "gpt-5.4",
				api: "openai-responses",
				reasoning: false,
				input: ["text", "image"],
				cost: {
					input: 0,
					output: 0,
					cacheRead: 0,
					cacheWrite: 0
				},
				contextWindow: 128e3,
				maxTokens: 4096
			},
			{
				id: "gpt-5.4-alt",
				name: "gpt-5.4-alt",
				api: "openai-responses",
				reasoning: false,
				input: ["text", "image"],
				cost: {
					input: 0,
					output: 0,
					cacheRead: 0,
					cacheWrite: 0
				},
				contextWindow: 128e3,
				maxTokens: 4096
			},
			{
				id: "gpt-image-1",
				name: "gpt-image-1",
				api: "openai-responses",
				reasoning: false,
				input: ["text"],
				cost: {
					input: 0,
					output: 0,
					cacheRead: 0,
					cacheWrite: 0
				},
				contextWindow: 128e3,
				maxTokens: 4096
			}
		]
	};
	const providerMode = normalizeQaProviderMode(params.providerMode ?? "mock-openai");
	const primaryModel = params.primaryModel ?? defaultQaModelForMode(providerMode);
	const alternateModel = params.alternateModel ?? defaultQaModelForMode(providerMode, { alternate: true });
	const modelProviderIds = [primaryModel, alternateModel].map((ref) => splitQaModelRef(ref)?.provider).filter((provider) => Boolean(provider));
	const imageGenerationModelRef = params.imageGenerationModel !== void 0 ? params.imageGenerationModel : providerMode === "mock-openai" ? "mock-openai/gpt-image-1" : modelProviderIds.includes("openai") ? "openai/gpt-image-1" : null;
	const selectedProviderIds = providerMode === "live-frontier" ? [...new Set([
		...params.enabledProviderIds ?? [],
		...modelProviderIds,
		imageGenerationModelRef
	].map((value) => typeof value === "string" ? splitQaModelRef(value)?.provider ?? value : null).filter((provider) => Boolean(provider)))] : [];
	const pluginEntries = providerMode === "live-frontier" ? Object.fromEntries(selectedProviderIds.map((providerId) => [providerId, { enabled: true }])) : {};
	const allowedPlugins = providerMode === "live-frontier" ? [
		"memory-core",
		...selectedProviderIds,
		"qa-channel"
	] : ["memory-core", "qa-channel"];
	const liveModelParams = providerMode === "live-frontier" ? (modelRef) => ({
		transport: "sse",
		openaiWsWarmup: false,
		...params.fastMode === true || isQaFastModeModelRef(modelRef) ? { fastMode: true } : {}
	}) : (_modelRef) => ({
		transport: "sse",
		openaiWsWarmup: false
	});
	const allowedOrigins = mergeQaControlUiAllowedOrigins(params.controlUiAllowedOrigins);
	return {
		plugins: {
			allow: allowedPlugins,
			entries: {
				acpx: { enabled: false },
				"memory-core": { enabled: true },
				...pluginEntries
			}
		},
		agents: {
			defaults: {
				workspace: params.workspaceDir,
				model: { primary: primaryModel },
				...imageGenerationModelRef ? { imageGenerationModel: { primary: imageGenerationModelRef } } : {},
				memorySearch: { sync: {
					watch: true,
					watchDebounceMs: 25,
					onSessionStart: true,
					onSearch: true
				} },
				models: {
					[primaryModel]: { params: liveModelParams(primaryModel) },
					[alternateModel]: { params: liveModelParams(alternateModel) }
				},
				subagents: {
					allowAgents: ["*"],
					maxConcurrent: 2
				}
			},
			list: [{
				id: "qa",
				default: true,
				model: { primary: primaryModel },
				identity: {
					name: "C-3PO QA",
					theme: "Flustered Protocol Droid",
					emoji: "🤖",
					avatar: "avatars/c3po.png"
				},
				subagents: { allowAgents: ["*"] }
			}]
		},
		memory: { backend: "builtin" },
		...providerMode === "mock-openai" ? { models: {
			mode: "replace",
			providers: { "mock-openai": mockOpenAiProvider }
		} } : {},
		gateway: {
			mode: "local",
			bind: params.bind,
			port: params.gatewayPort,
			auth: {
				mode: "token",
				token: params.gatewayToken
			},
			reload: { deferralTimeoutMs: 1e3 },
			controlUi: {
				enabled: params.controlUiEnabled ?? true,
				...(params.controlUiEnabled ?? true) && params.controlUiRoot ? { root: params.controlUiRoot } : {},
				...params.controlUiEnabled ?? true ? {
					allowInsecureAuth: true,
					allowedOrigins
				} : {}
			}
		},
		discovery: { mdns: { mode: "off" } },
		channels: {
			...DISABLED_BUNDLED_CHANNELS,
			"qa-channel": {
				enabled: true,
				baseUrl: params.qaBusBaseUrl,
				botUserId: "openclaw",
				botDisplayName: "OpenClaw QA",
				allowFrom: ["*"],
				pollTimeoutMs: 250
			}
		},
		messages: { groupChat: { mentionPatterns: ["\\b@?openclaw\\b"] } }
	};
}
//#endregion
export { buildQaGatewayConfig as n, mergeQaControlUiAllowedOrigins as r, DEFAULT_QA_CONTROL_UI_ALLOWED_ORIGINS as t };
