// openclaw/dist/sessions-Dz-6DI5z.js
import { c as loadConfig } from "./config-DMMR1XE_.js";
import "./store-CeqUIT7u.js";
import { i as resolveMainSessionKey } from "./main-session-DaD0xK0B.js";
import { l as resolveStorePath } from "./paths-iH379dx6.js";
import "./reset-ClbjyzZk.js";
import "./session-key-i2poQWVq.js";
import { t as loadSessionStore } from "./store-load-DF6H9tmN.js";
import "./session-file-Bqa5WMrx.js";
import { t as parseSessionThreadInfo } from "./thread-info-BRq__y-4.js";
import "./transcript-61cUaMiH.js";
import "./targets-CGUgeaD9.js";
//#region src/config/sessions/main-session.runtime.ts
function resolveMainSessionKeyFromConfig() {
	return resolveMainSessionKey(loadConfig());
}
//#endregion
//#region src/config/sessions/delivery-info.ts
function extractDeliveryInfo(sessionKey) {
	const { baseSessionKey, threadId } = parseSessionThreadInfo(sessionKey);
	if (!sessionKey || !baseSessionKey) return {
		deliveryContext: void 0,
		threadId
	};
	let deliveryContext;
	try {
		const store = loadSessionStore(resolveStorePath(loadConfig().session?.store));
		let entry = store[sessionKey];
		if (!entry?.deliveryContext && baseSessionKey !== sessionKey) entry = store[baseSessionKey];
		if (entry?.deliveryContext) {
			const resolvedThreadId = entry.deliveryContext.threadId ?? entry.lastThreadId ?? entry.origin?.threadId;
			deliveryContext = {
				channel: entry.deliveryContext.channel,
				to: entry.deliveryContext.to,
				accountId: entry.deliveryContext.accountId,
				threadId: resolvedThreadId != null ? String(resolvedThreadId) : void 0
			};
		}
	} catch {}
	return {
		deliveryContext,
		threadId
	};
}
//#endregion
export { resolveMainSessionKeyFromConfig as n, extractDeliveryInfo as t };
