// openclaw/dist/slash-dispatch.runtime-BDzNRPzS.js
import { t as resolveConversationLabel$1 } from "./conversation-label-DtKa5VU0.js";
import { s as resolveChunkMode$1 } from "./chunk-SRc53mDN.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-BP-6p4LO.js";
import "./routing-D0K9urg9.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-Bq2CmFVy.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-COjV0JaB.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-D17dZcew.js";
import "./config-runtime-Cl-JrKQb.js";
import "./reply-runtime-Bm_eA5lM.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-DTqP40Bm.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-Bp9F4urp.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
