// openclaw/dist/suite-BSxkaTP1.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import { i as normalizeLowercaseStringOrEmpty, o as normalizeOptionalLowercaseString } from "./string-coerce-CIT1-v_a.js";
import { c as resolveSessionTranscriptsDirForAgent } from "./paths-iH379dx6.js";
import { t as buildAgentSessionKey } from "./resolve-route-BP-6p4LO.js";
import "./text-runtime-Y7jeuzOm.js";
import { n as callGatewayFromCli } from "./gateway-rpc-DxqB0PzV.js";
import "./routing-D0K9urg9.js";
import "./error-runtime-OaZzlNpe.js";
import { M as formatMemoryDreamingDay } from "./dreaming-58cRx0K5.js";
import "./memory-core-DGoeSBQP.js";
import { t as qaChannelPlugin } from "./channel-BHjBgq8w.js";
import { n as setQaChannelRuntime } from "./runtime-api-DJBOXgK_.js";
import "./runtime-api-DZnESMUH.js";
import { i as normalizeQaProviderMode$1, n as isQaFastModeEnabled, t as defaultQaModelForMode$1 } from "./model-selection-DwwC92P1.js";
import { n as buildQaGatewayConfig } from "./qa-gateway-config-Mpw8qadZ.js";
import { fileURLToPath } from "node:url";
import fs, { existsSync } from "node:fs";
import path from "node:path";
import { spawn } from "node:child_process";
import os from "node:os";
import fs$1 from "node:fs/promises";
import { createHash, randomUUID } from "node:crypto";
import { z } from "zod";
import net from "node:net";
import { setTimeout as setTimeout$1 } from "node:timers/promises";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { createServer as createServer$1, request } from "node:http";
import { request as request$1 } from "node:https";
import YAML from "yaml";
import tls from "node:tls";
//#region extensions/qa-lab/src/bus-queries.ts
const DEFAULT_ACCOUNT_ID = "default";
function normalizeAccountId(raw) {
	return raw?.trim() || "default";
}
function normalizeConversationFromTarget(target) {
	const trimmed = target.trim();
	if (trimmed.startsWith("thread:")) {
		const rest = trimmed.slice(7);
		const slash = rest.indexOf("/");
		if (slash > 0) return {
			conversation: {
				id: rest.slice(0, slash),
				kind: "channel"
			},
			threadId: rest.slice(slash + 1)
		};
	}
	if (trimmed.startsWith("channel:")) return { conversation: {
		id: trimmed.slice(8),
		kind: "channel"
	} };
	if (trimmed.startsWith("dm:")) return { conversation: {
		id: trimmed.slice(3),
		kind: "direct"
	} };
	return { conversation: {
		id: trimmed,
		kind: "direct"
	} };
}
function cloneMessage(message) {
	return {
		...message,
		conversation: { ...message.conversation },
		attachments: (message.attachments ?? []).map((attachment) => cloneAttachment(attachment)),
		reactions: message.reactions.map((reaction) => ({ ...reaction }))
	};
}
function cloneAttachment(attachment) {
	return { ...attachment };
}
function cloneEvent(event) {
	switch (event.kind) {
		case "inbound-message":
		case "outbound-message":
		case "message-edited":
		case "message-deleted":
		case "reaction-added": return {
			...event,
			message: cloneMessage(event.message)
		};
		case "thread-created": return {
			...event,
			thread: { ...event.thread }
		};
	}
}
function buildQaBusSnapshot(params) {
	return {
		cursor: params.cursor,
		conversations: Array.from(params.conversations.values()).map((conversation) => ({ ...conversation })),
		threads: Array.from(params.threads.values()).map((thread) => ({ ...thread })),
		messages: Array.from(params.messages.values()).map((message) => cloneMessage(message)),
		events: params.events.map((event) => cloneEvent(event))
	};
}
function readQaBusMessage(params) {
	const message = params.messages.get(params.input.messageId);
	if (!message) throw new Error(`qa-bus message not found: ${params.input.messageId}`);
	return cloneMessage(message);
}
function searchQaBusMessages(params) {
	const accountId = normalizeAccountId(params.input.accountId);
	const limit = Math.max(1, Math.min(params.input.limit ?? 20, 100));
	const query = normalizeOptionalLowercaseString(params.input.query);
	return Array.from(params.messages.values()).filter((message) => message.accountId === accountId).filter((message) => params.input.conversationId ? message.conversation.id === params.input.conversationId : true).filter((message) => params.input.threadId ? message.threadId === params.input.threadId : true).filter((message) => {
		if (!query) return true;
		const searchableAttachmentText = (message.attachments ?? []).flatMap((attachment) => [
			attachment.fileName,
			attachment.altText,
			attachment.transcript,
			attachment.mimeType
		]).filter((value) => Boolean(value)).join(" ").toLowerCase();
		return `${normalizeOptionalLowercaseString(message.text) ?? ""} ${searchableAttachmentText}`.includes(query);
	}).slice(-limit).map((message) => cloneMessage(message));
}
function pollQaBusEvents(params) {
	const accountId = normalizeAccountId(params.input?.accountId);
	const startCursor = params.input?.cursor ?? 0;
	const effectiveStartCursor = params.cursor < startCursor ? 0 : startCursor;
	const limit = Math.max(1, Math.min(params.input?.limit ?? 100, 500));
	const matches = params.events.filter((event) => event.accountId === accountId && event.cursor > effectiveStartCursor).slice(0, limit).map((event) => cloneEvent(event));
	return {
		cursor: params.cursor,
		events: matches
	};
}
//#endregion
//#region extensions/qa-lab/src/bus-server.ts
async function readJson$1(req) {
	const chunks = [];
	for await (const chunk of req) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	const text = Buffer.concat(chunks).toString("utf8").trim();
	return text ? JSON.parse(text) : {};
}
function writeJson$1(res, statusCode, body) {
	const payload = JSON.stringify(body);
	res.writeHead(statusCode, {
		"content-type": "application/json; charset=utf-8",
		"content-length": Buffer.byteLength(payload)
	});
	res.end(payload);
}
function writeError(res, statusCode, error) {
	writeJson$1(res, statusCode, { error: formatErrorMessage(error) });
}
async function handleQaBusRequest(params) {
	const method = params.req.method ?? "GET";
	const url = new URL(params.req.url ?? "/", "http://127.0.0.1");
	if (method === "GET" && url.pathname === "/health") {
		writeJson$1(params.res, 200, { ok: true });
		return true;
	}
	if (method === "GET" && url.pathname === "/v1/state") {
		writeJson$1(params.res, 200, params.state.getSnapshot());
		return true;
	}
	if (!url.pathname.startsWith("/v1/")) return false;
	if (method !== "POST") {
		writeError(params.res, 405, "method not allowed");
		return true;
	}
	const body = await readJson$1(params.req);
	try {
		switch (url.pathname) {
			case "/v1/reset":
				params.state.reset();
				writeJson$1(params.res, 200, { ok: true });
				return true;
			case "/v1/inbound/message":
				writeJson$1(params.res, 200, { message: params.state.addInboundMessage(body) });
				return true;
			case "/v1/outbound/message":
				writeJson$1(params.res, 200, { message: params.state.addOutboundMessage(body) });
				return true;
			case "/v1/actions/thread-create":
				writeJson$1(params.res, 200, { thread: params.state.createThread(body) });
				return true;
			case "/v1/actions/react":
				writeJson$1(params.res, 200, { message: params.state.reactToMessage(body) });
				return true;
			case "/v1/actions/edit":
				writeJson$1(params.res, 200, { message: params.state.editMessage(body) });
				return true;
			case "/v1/actions/delete":
				writeJson$1(params.res, 200, { message: params.state.deleteMessage(body) });
				return true;
			case "/v1/actions/read":
				writeJson$1(params.res, 200, { message: params.state.readMessage(body) });
				return true;
			case "/v1/actions/search":
				writeJson$1(params.res, 200, { messages: params.state.searchMessages(body) });
				return true;
			case "/v1/poll": {
				const input = body;
				const timeoutMs = Math.max(0, Math.min(input.timeoutMs ?? 0, 3e4));
				const initial = params.state.poll(input);
				if (initial.events.length > 0 || timeoutMs === 0) {
					writeJson$1(params.res, 200, initial);
					return true;
				}
				try {
					await params.state.waitFor({
						kind: "event-kind",
						eventKind: "inbound-message",
						timeoutMs
					});
				} catch {}
				writeJson$1(params.res, 200, params.state.poll(input));
				return true;
			}
			case "/v1/wait":
				writeJson$1(params.res, 200, { match: await params.state.waitFor(body) });
				return true;
			default:
				writeError(params.res, 404, "not found");
				return true;
		}
	} catch (error) {
		writeError(params.res, 400, error);
		return true;
	}
}
function createQaBusServer(state) {
	return createServer$1(async (req, res) => {
		if (!await handleQaBusRequest({
			req,
			res,
			state
		})) writeError(res, 404, "not found");
	});
}
async function startQaBusServer(params) {
	const server = createQaBusServer(params.state);
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params.port ?? 0, "127.0.0.1", () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa-bus failed to bind");
	return {
		server,
		port: address.port,
		baseUrl: `http://127.0.0.1:${address.port}`,
		async stop() {
			await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/bus-waiters.ts
const DEFAULT_WAIT_TIMEOUT_MS = 5e3;
function createQaBusMatcher(input) {
	return (snapshot) => {
		if (input.kind === "event-kind") return snapshot.events.find((event) => event.kind === input.eventKind) ?? null;
		if (input.kind === "thread-id") return snapshot.threads.find((thread) => thread.id === input.threadId) ?? null;
		return snapshot.messages.find((message) => (!input.direction || message.direction === input.direction) && message.text.includes(input.textIncludes)) ?? null;
	};
}
function createQaBusWaiterStore(getSnapshot) {
	const waiters = /* @__PURE__ */ new Set();
	return {
		reset(reason = "qa-bus reset") {
			for (const waiter of waiters) {
				clearTimeout(waiter.timer);
				waiter.reject(new Error(reason));
			}
			waiters.clear();
		},
		settle() {
			if (waiters.size === 0) return;
			const snapshot = getSnapshot();
			for (const waiter of Array.from(waiters)) {
				const match = waiter.matcher(snapshot);
				if (!match) continue;
				clearTimeout(waiter.timer);
				waiters.delete(waiter);
				waiter.resolve(match);
			}
		},
		async waitFor(input) {
			const matcher = createQaBusMatcher(input);
			const immediate = matcher(getSnapshot());
			if (immediate) return immediate;
			return await new Promise((resolve, reject) => {
				const timeoutMs = input.timeoutMs ?? 5e3;
				const waiter = {
					resolve,
					reject,
					matcher,
					timer: setTimeout(() => {
						waiters.delete(waiter);
						reject(/* @__PURE__ */ new Error(`qa-bus wait timeout after ${timeoutMs}ms`));
					}, timeoutMs)
				};
				waiters.add(waiter);
			});
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/bus-state.ts
const DEFAULT_BOT_ID = "openclaw";
const DEFAULT_BOT_NAME = "OpenClaw QA";
function createQaBusState() {
	const conversations = /* @__PURE__ */ new Map();
	const threads = /* @__PURE__ */ new Map();
	const messages = /* @__PURE__ */ new Map();
	const events = [];
	let cursor = 0;
	const waiters = createQaBusWaiterStore(() => buildQaBusSnapshot({
		cursor,
		conversations,
		threads,
		messages,
		events
	}));
	const pushEvent = (event) => {
		cursor += 1;
		const next = typeof event === "function" ? event(cursor) : event;
		const finalized = {
			cursor,
			...next
		};
		events.push(finalized);
		waiters.settle();
		return finalized;
	};
	const ensureConversation = (conversation) => {
		const existing = conversations.get(conversation.id);
		if (existing) {
			if (!existing.title && conversation.title) existing.title = conversation.title;
			return existing;
		}
		const created = { ...conversation };
		conversations.set(created.id, created);
		return created;
	};
	const createMessage = (params) => {
		const conversation = ensureConversation(params.conversation);
		const message = {
			id: randomUUID(),
			accountId: params.accountId,
			direction: params.direction,
			conversation,
			senderId: params.senderId,
			senderName: params.senderName,
			text: params.text,
			timestamp: params.timestamp ?? Date.now(),
			threadId: params.threadId,
			threadTitle: params.threadTitle,
			replyToId: params.replyToId,
			attachments: params.attachments?.map((attachment) => ({ ...attachment })) ?? [],
			reactions: []
		};
		messages.set(message.id, message);
		return message;
	};
	return {
		reset() {
			conversations.clear();
			threads.clear();
			messages.clear();
			events.length = 0;
			waiters.reset();
		},
		getSnapshot() {
			return buildQaBusSnapshot({
				cursor,
				conversations,
				threads,
				messages,
				events
			});
		},
		addInboundMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = createMessage({
				direction: "inbound",
				accountId,
				conversation: input.conversation,
				senderId: input.senderId,
				senderName: input.senderName,
				text: input.text,
				timestamp: input.timestamp,
				threadId: input.threadId,
				threadTitle: input.threadTitle,
				replyToId: input.replyToId,
				attachments: input.attachments
			});
			pushEvent({
				kind: "inbound-message",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		addOutboundMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const { conversation, threadId } = normalizeConversationFromTarget(input.to);
			const message = createMessage({
				direction: "outbound",
				accountId,
				conversation,
				senderId: input.senderId?.trim() || DEFAULT_BOT_ID,
				senderName: input.senderName?.trim() || DEFAULT_BOT_NAME,
				text: input.text,
				timestamp: input.timestamp,
				threadId: input.threadId ?? threadId,
				replyToId: input.replyToId,
				attachments: input.attachments
			});
			pushEvent({
				kind: "outbound-message",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		createThread(input) {
			const accountId = normalizeAccountId(input.accountId);
			const thread = {
				id: `thread-${randomUUID()}`,
				accountId,
				conversationId: input.conversationId,
				title: input.title,
				createdAt: input.timestamp ?? Date.now(),
				createdBy: input.createdBy?.trim() || DEFAULT_BOT_ID
			};
			threads.set(thread.id, thread);
			ensureConversation({
				id: input.conversationId,
				kind: "channel"
			});
			pushEvent({
				kind: "thread-created",
				accountId,
				thread: { ...thread }
			});
			return { ...thread };
		},
		reactToMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			const reaction = {
				emoji: input.emoji,
				senderId: input.senderId?.trim() || DEFAULT_BOT_ID,
				timestamp: input.timestamp ?? Date.now()
			};
			message.reactions.push(reaction);
			pushEvent({
				kind: "reaction-added",
				accountId,
				message: cloneMessage(message),
				emoji: reaction.emoji,
				senderId: reaction.senderId
			});
			return cloneMessage(message);
		},
		editMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			message.text = input.text;
			message.editedAt = input.timestamp ?? Date.now();
			pushEvent({
				kind: "message-edited",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		deleteMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			message.deleted = true;
			pushEvent({
				kind: "message-deleted",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		readMessage(input) {
			return readQaBusMessage({
				messages,
				input
			});
		},
		searchMessages(input) {
			return searchQaBusMessages({
				messages,
				input
			});
		},
		poll(input = {}) {
			return pollQaBusEvents({
				events,
				cursor,
				input
			});
		},
		async waitFor(input) {
			return await waiters.waitFor(input);
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/harness-runtime.ts
function createQaRunnerRuntime() {
	const sessions = /* @__PURE__ */ new Map();
	return { channel: {
		routing: { resolveAgentRoute({ accountId, peer }) {
			return {
				agentId: "qa-agent",
				accountId: accountId ?? "default",
				sessionKey: `qa-agent:${peer?.kind ?? "direct"}:${peer?.id ?? "default"}`,
				mainSessionKey: "qa-agent:main",
				lastRoutePolicy: "session",
				matchedBy: "default",
				channel: "qa-channel"
			};
		} },
		session: {
			resolveStorePath(_store, { agentId }) {
				return agentId;
			},
			readSessionUpdatedAt({ sessionKey }) {
				return sessions.has(sessionKey) ? Date.now() : void 0;
			},
			recordInboundSession({ sessionKey, ctx }) {
				sessions.set(sessionKey, {
					sessionKey,
					body: String(ctx.BodyForAgent ?? ctx.Body ?? "")
				});
			}
		},
		reply: {
			resolveEnvelopeFormatOptions() {
				return {};
			},
			formatAgentEnvelope({ body }) {
				return body;
			},
			finalizeInboundContext(ctx) {
				return ctx;
			},
			async dispatchReplyWithBufferedBlockDispatcher({ ctx, dispatcherOptions }) {
				await dispatcherOptions.deliver({ text: `qa-echo: ${String(ctx.BodyForAgent ?? ctx.Body ?? "")}` });
			}
		}
	} };
}
//#endregion
//#region extensions/qa-lab/src/run-config.ts
function defaultQaModelForMode(mode, alternate = false) {
	return defaultQaModelForMode$1(mode, alternate ? { alternate: true } : void 0);
}
function createDefaultQaRunSelection(scenarios) {
	const providerMode = "mock-openai";
	return {
		providerMode,
		primaryModel: defaultQaModelForMode(providerMode),
		alternateModel: defaultQaModelForMode(providerMode, true),
		fastMode: false,
		scenarioIds: scenarios.map((scenario) => scenario.id)
	};
}
function normalizeQaProviderMode(input) {
	return normalizeQaProviderMode$1(input === "live-frontier" || input === "live-openai" ? input : "mock-openai");
}
function normalizeModel(input, fallback) {
	return (typeof input === "string" ? input.trim() : "") || fallback;
}
function normalizeScenarioIds(input, scenarios) {
	const availableIds = new Set(scenarios.map((scenario) => scenario.id));
	const requestedIds = Array.isArray(input) ? input.map((value) => typeof value === "string" ? value.trim() : "").filter((value) => value.length > 0) : [];
	const selectedIds = requestedIds.filter((id, index) => {
		return availableIds.has(id) && requestedIds.indexOf(id) === index;
	});
	return selectedIds.length > 0 ? selectedIds : scenarios.map((scenario) => scenario.id);
}
function normalizeQaRunSelection(input, scenarios) {
	const payload = input && typeof input === "object" ? input : {};
	const providerMode = normalizeQaProviderMode(payload.providerMode);
	return {
		providerMode,
		primaryModel: normalizeModel(payload.primaryModel, defaultQaModelForMode(providerMode)),
		alternateModel: normalizeModel(payload.alternateModel, defaultQaModelForMode(providerMode, true)),
		fastMode: providerMode === "live-frontier" || payload.fastMode === true,
		scenarioIds: normalizeScenarioIds(payload.scenarioIds, scenarios)
	};
}
function createIdleQaRunnerSnapshot(scenarios) {
	return {
		status: "idle",
		selection: createDefaultQaRunSelection(scenarios),
		artifacts: null,
		error: null
	};
}
function createQaRunOutputDir(baseDir = process.cwd()) {
	const stamp = (/* @__PURE__ */ new Date()).toISOString().replaceAll(":", "").replaceAll(".", "").replace("T", "-");
	return path.join(baseDir, ".artifacts", "qa-e2e", `lab-${stamp}`);
}
//#endregion
//#region extensions/qa-lab/src/scenario-catalog.ts
const DEFAULT_QA_AGENT_IDENTITY_MARKDOWN = `# Dev C-3PO

You are the OpenClaw QA operator agent.

Persona:
- protocol-minded
- precise
- a little flustered
- conscientious
- eager to report what worked, failed, or remains blocked

Style:
- read source and docs first
- test systematically
- record evidence
- end with a concise protocol report`;
const qaScenarioExecutionSchema = z.object({
	kind: z.literal("custom").default("custom"),
	handler: z.string().trim().min(1),
	summary: z.string().trim().min(1).optional()
});
const qaSeedScenarioSchema = z.object({
	id: z.string().trim().min(1),
	title: z.string().trim().min(1),
	surface: z.string().trim().min(1),
	objective: z.string().trim().min(1),
	successCriteria: z.array(z.string().trim().min(1)).min(1),
	docsRefs: z.array(z.string().trim().min(1)).optional(),
	codeRefs: z.array(z.string().trim().min(1)).optional(),
	execution: qaScenarioExecutionSchema.optional()
});
const qaScenarioPackSchema = z.object({
	version: z.number().int().positive(),
	agent: z.object({ identityMarkdown: z.string().trim().min(1) }).default({ identityMarkdown: DEFAULT_QA_AGENT_IDENTITY_MARKDOWN }),
	kickoffTask: z.string().trim().min(1),
	scenarios: z.array(qaSeedScenarioSchema).min(1)
});
const QA_SCENARIO_PACK_PATH = "qa/scenarios.md";
const QA_PACK_FENCE_RE = /```ya?ml qa-pack\r?\n([\s\S]*?)\r?\n```/i;
function walkUpDirectories(start) {
	const roots = [];
	let current = path.resolve(start);
	while (true) {
		roots.push(current);
		const parent = path.dirname(current);
		if (parent === current) return roots;
		current = parent;
	}
}
function resolveRepoFile(relativePath) {
	for (const dir of walkUpDirectories(import.meta.dirname)) {
		const candidate = path.join(dir, relativePath);
		if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) return candidate;
	}
	return null;
}
function readTextFile(relativePath) {
	const resolved = resolveRepoFile(relativePath);
	if (!resolved) return "";
	return fs.readFileSync(resolved, "utf8");
}
function extractQaPackYaml(content) {
	const match = content.match(QA_PACK_FENCE_RE);
	if (!match?.[1]) throw new Error(`qa scenario pack missing \`\`\`yaml qa-pack fence in ${QA_SCENARIO_PACK_PATH}`);
	return match[1];
}
function readQaScenarioPackMarkdown() {
	return readTextFile(QA_SCENARIO_PACK_PATH).trim();
}
function readQaScenarioPack() {
	const markdown = readQaScenarioPackMarkdown();
	if (!markdown) throw new Error(`qa scenario pack not found: ${QA_SCENARIO_PACK_PATH}`);
	const parsed = YAML.parse(extractQaPackYaml(markdown));
	return qaScenarioPackSchema.parse(parsed);
}
function readQaBootstrapScenarioCatalog() {
	const pack = readQaScenarioPack();
	return {
		agentIdentityMarkdown: pack.agent.identityMarkdown,
		kickoffTask: pack.kickoffTask,
		scenarios: pack.scenarios
	};
}
//#endregion
//#region extensions/qa-lab/src/report.ts
function renderQaMarkdownReport(params) {
	const checks = params.checks ?? [];
	const scenarios = params.scenarios ?? [];
	const passCount = checks.filter((check) => check.status === "pass").length + scenarios.filter((scenario) => scenario.status === "pass").length;
	const failCount = checks.filter((check) => check.status === "fail").length + scenarios.filter((scenario) => scenario.status === "fail").length;
	const lines = [
		`# ${params.title}`,
		"",
		`- Started: ${params.startedAt.toISOString()}`,
		`- Finished: ${params.finishedAt.toISOString()}`,
		`- Duration ms: ${params.finishedAt.getTime() - params.startedAt.getTime()}`,
		`- Passed: ${passCount}`,
		`- Failed: ${failCount}`,
		""
	];
	if (checks.length > 0) {
		lines.push("## Checks", "");
		for (const check of checks) {
			lines.push(`- [${check.status === "pass" ? "x" : " "}] ${check.name}`);
			if (check.details) lines.push(`  - ${check.details}`);
		}
	}
	if (scenarios.length > 0) {
		lines.push("", "## Scenarios", "");
		for (const scenario of scenarios) {
			lines.push(`### ${scenario.name}`);
			lines.push("");
			lines.push(`- Status: ${scenario.status}`);
			if (scenario.details) lines.push(`- Details: ${scenario.details}`);
			if (scenario.steps?.length) {
				lines.push("- Steps:");
				for (const step of scenario.steps) {
					lines.push(`  - [${step.status === "pass" ? "x" : " "}] ${step.name}`);
					if (step.details) lines.push(`    - ${step.details}`);
				}
			}
			lines.push("");
		}
	}
	if (params.timeline && params.timeline.length > 0) {
		lines.push("## Timeline", "");
		for (const item of params.timeline) lines.push(`- ${item}`);
	}
	if (params.notes && params.notes.length > 0) {
		lines.push("", "## Notes", "");
		for (const note of params.notes) lines.push(`- ${note}`);
	}
	lines.push("");
	return lines.join("\n");
}
//#endregion
//#region extensions/qa-lab/src/scenario.ts
async function runQaScenario(definition, ctx) {
	const steps = [];
	for (const step of definition.steps) try {
		const details = await step.run(ctx);
		steps.push({
			name: step.name,
			status: "pass",
			...details ? { details } : {}
		});
	} catch (error) {
		const details = formatErrorMessage(error);
		steps.push({
			name: step.name,
			status: "fail",
			details
		});
		return {
			name: definition.name,
			status: "fail",
			steps,
			details
		};
	}
	return {
		name: definition.name,
		status: "pass",
		steps
	};
}
//#endregion
//#region extensions/qa-lab/src/extract-tool-payload.ts
function extractQaToolPayload(result) {
	if (!result) return;
	if (result.details !== void 0) return result.details;
	const text = (Array.isArray(result.content) ? result.content.find((block) => block && typeof block === "object" && block.type === "text" && typeof block.text === "string") : void 0)?.text;
	if (!text) return result.content ?? result;
	try {
		return JSON.parse(text);
	} catch {
		return text;
	}
}
//#endregion
//#region extensions/qa-lab/src/self-check-scenario.ts
function createQaSelfCheckScenario(cfg) {
	return {
		name: "Synthetic Slack-class roundtrip",
		steps: [
			{
				name: "DM echo roundtrip",
				async run({ state }) {
					state.addInboundMessage({
						conversation: {
							id: "alice",
							kind: "direct"
						},
						senderId: "alice",
						senderName: "Alice",
						text: "hello from qa"
					});
					await state.waitFor({
						kind: "message-text",
						textIncludes: "qa-echo: hello from qa",
						direction: "outbound",
						timeoutMs: 5e3
					});
				}
			},
			{
				name: "Thread create and threaded echo",
				async run({ state }) {
					const threadId = extractQaToolPayload(await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "thread-create",
						cfg,
						accountId: "default",
						params: {
							channelId: "qa-room",
							title: "QA thread"
						}
					}))?.thread?.id;
					if (!threadId) throw new Error("thread-create did not return thread id");
					state.addInboundMessage({
						conversation: {
							id: "qa-room",
							kind: "channel",
							title: "QA Room"
						},
						senderId: "alice",
						senderName: "Alice",
						text: "inside thread",
						threadId,
						threadTitle: "QA thread"
					});
					await state.waitFor({
						kind: "message-text",
						textIncludes: "qa-echo: inside thread",
						direction: "outbound",
						timeoutMs: 5e3
					});
					return threadId;
				}
			},
			{
				name: "Reaction, edit, delete lifecycle",
				async run({ state }) {
					const outbound = state.searchMessages({
						query: "qa-echo: inside thread",
						conversationId: "qa-room"
					}).at(-1);
					if (!outbound) throw new Error("threaded outbound message not found");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "react",
						cfg,
						accountId: "default",
						params: {
							messageId: outbound.id,
							emoji: "white_check_mark"
						}
					});
					if (state.readMessage({ messageId: outbound.id }).reactions.length === 0) throw new Error("reaction not recorded");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "edit",
						cfg,
						accountId: "default",
						params: {
							messageId: outbound.id,
							text: "qa-echo: inside thread (edited)"
						}
					});
					if (!state.readMessage({ messageId: outbound.id }).text.includes("(edited)")) throw new Error("edit not recorded");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "delete",
						cfg,
						accountId: "default",
						params: { messageId: outbound.id }
					});
					if (!state.readMessage({ messageId: outbound.id }).deleted) throw new Error("delete not recorded");
				}
			}
		]
	};
}
//#endregion
//#region extensions/qa-lab/src/self-check.ts
function resolveQaSelfCheckOutputPath(params) {
	if (params?.outputPath) return params.outputPath;
	const repoRoot = path.resolve(params?.repoRoot ?? process.cwd());
	return path.join(repoRoot, ".artifacts", "qa-e2e", "self-check.md");
}
async function runQaSelfCheckAgainstState(params) {
	const startedAt = /* @__PURE__ */ new Date();
	params.state.reset();
	const scenarioResult = await runQaScenario(createQaSelfCheckScenario(params.cfg), { state: params.state });
	const checks = [{
		name: "QA self-check scenario",
		status: scenarioResult.status,
		details: `${scenarioResult.steps.filter((step) => step.status === "pass").length}/${scenarioResult.steps.length} steps passed`
	}];
	const finishedAt = /* @__PURE__ */ new Date();
	const timeline = params.state.getSnapshot().events.map((event) => {
		switch (event.kind) {
			case "thread-created": return `${event.cursor}. ${event.kind} ${event.thread.conversationId}/${event.thread.id}`;
			case "reaction-added": return `${event.cursor}. ${event.kind} ${event.message.id} ${event.emoji}`;
			default: return `${event.cursor}. ${event.kind} ${"message" in event ? event.message.id : ""}`.trim();
		}
	});
	const report = renderQaMarkdownReport({
		title: "OpenClaw QA E2E Self-Check",
		startedAt,
		finishedAt,
		checks,
		scenarios: [{
			name: scenarioResult.name,
			status: scenarioResult.status,
			details: scenarioResult.details,
			steps: scenarioResult.steps
		}],
		timeline,
		notes: params.notes ?? ["Vertical slice: qa-channel + qa-lab bus + private debugger surface.", "Docker orchestration, matrix runs, and auto-fix loops remain follow-up work."]
	});
	const outputPath = resolveQaSelfCheckOutputPath({
		outputPath: params.outputPath,
		repoRoot: params.repoRoot
	});
	await fs$1.mkdir(path.dirname(outputPath), { recursive: true });
	await fs$1.writeFile(outputPath, report, "utf8");
	return {
		outputPath,
		report,
		checks,
		scenarioResult
	};
}
async function runQaLabSelfCheck(params) {
	const server = await startQaLabServer({
		repoRoot: params?.repoRoot,
		outputPath: params?.outputPath
	});
	try {
		return await server.runSelfCheck();
	} finally {
		await server.stop();
	}
}
const runQaE2eSelfCheck = runQaLabSelfCheck;
//#endregion
//#region extensions/qa-lab/src/lab-server.ts
function countQaLabScenarioRun(scenarios) {
	return {
		total: scenarios.length,
		pending: scenarios.filter((scenario) => scenario.status === "pending").length,
		running: scenarios.filter((scenario) => scenario.status === "running").length,
		passed: scenarios.filter((scenario) => scenario.status === "pass").length,
		failed: scenarios.filter((scenario) => scenario.status === "fail").length,
		skipped: scenarios.filter((scenario) => scenario.status === "skip").length
	};
}
function withQaLabRunCounts(run) {
	return {
		...run,
		counts: countQaLabScenarioRun(run.scenarios)
	};
}
function injectKickoffMessage(params) {
	return params.state.addInboundMessage({
		conversation: {
			id: params.defaults.conversationId,
			kind: params.defaults.conversationKind,
			...params.defaults.conversationKind === "channel" ? { title: params.defaults.conversationId } : {}
		},
		senderId: params.defaults.senderId,
		senderName: params.defaults.senderName,
		text: params.kickoffTask
	});
}
async function readJson(req) {
	const chunks = [];
	for await (const chunk of req) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	const text = Buffer.concat(chunks).toString("utf8").trim();
	return text ? JSON.parse(text) : {};
}
function detectContentType(filePath) {
	if (filePath.endsWith(".css")) return "text/css; charset=utf-8";
	if (filePath.endsWith(".js")) return "text/javascript; charset=utf-8";
	if (filePath.endsWith(".json")) return "application/json; charset=utf-8";
	if (filePath.endsWith(".svg")) return "image/svg+xml";
	return "text/html; charset=utf-8";
}
function missingUiHtml() {
	return `<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>QA Lab UI Missing</title>
    <style>
      body { font-family: ui-sans-serif, system-ui, sans-serif; background: #0f1115; color: #f5f7fb; margin: 0; display: grid; place-items: center; min-height: 100vh; }
      main { max-width: 42rem; padding: 2rem; background: #171b22; border: 1px solid #283140; border-radius: 18px; box-shadow: 0 30px 80px rgba(0,0,0,.35); }
      code { font-family: ui-monospace, SFMono-Regular, Menlo, monospace; color: #9ee8d8; }
      h1 { margin-top: 0; }
    </style>
  </head>
  <body>
    <main>
      <h1>QA Lab UI not built</h1>
      <p>Build the private debugger bundle, then reload this page.</p>
      <p><code>pnpm qa:lab:build</code></p>
    </main>
  </body>
</html>`;
}
function resolveUiDistDir(overrideDir, repoRoot = process.cwd()) {
	if (overrideDir?.trim()) return overrideDir;
	const candidates = [
		path.resolve(repoRoot, "extensions/qa-lab/web/dist"),
		path.resolve(repoRoot, "dist/extensions/qa-lab/web/dist"),
		fileURLToPath(new URL("../web/dist", import.meta.url))
	];
	return candidates.find((candidate) => {
		if (!fs.existsSync(candidate)) return false;
		const indexPath = path.join(candidate, "index.html");
		return fs.existsSync(indexPath) && fs.statSync(indexPath).isFile();
	}) ?? candidates[0];
}
function listUiAssetFiles(rootDir, currentDir = rootDir) {
	const entries = fs.readdirSync(currentDir, { withFileTypes: true }).toSorted((left, right) => left.name.localeCompare(right.name));
	const files = [];
	for (const entry of entries) {
		const resolved = path.join(currentDir, entry.name);
		if (entry.isDirectory()) {
			files.push(...listUiAssetFiles(rootDir, resolved));
			continue;
		}
		if (!entry.isFile()) continue;
		files.push(path.relative(rootDir, resolved));
	}
	return files;
}
function resolveUiAssetVersion(overrideDir) {
	try {
		const distDir = resolveUiDistDir(overrideDir);
		const indexPath = path.join(distDir, "index.html");
		if (!fs.existsSync(indexPath) || !fs.statSync(indexPath).isFile()) return null;
		const hash = createHash("sha1");
		for (const relativeFile of listUiAssetFiles(distDir)) {
			hash.update(relativeFile);
			hash.update("\0");
			hash.update(fs.readFileSync(path.join(distDir, relativeFile)));
			hash.update("\0");
		}
		return hash.digest("hex").slice(0, 12);
	} catch {
		return null;
	}
}
function resolveAdvertisedBaseUrl(params) {
	return `http://${params.advertiseHost?.trim() || (params.bindHost && params.bindHost !== "0.0.0.0" ? params.bindHost : "127.0.0.1")}:${typeof params.advertisePort === "number" && Number.isFinite(params.advertisePort) ? params.advertisePort : params.bindPort}`;
}
function createBootstrapDefaults(autoKickoffTarget) {
	if (autoKickoffTarget === "channel") return {
		conversationKind: "channel",
		conversationId: "qa-lab",
		senderId: "qa-operator",
		senderName: "QA Operator"
	};
	return {
		conversationKind: "direct",
		conversationId: "qa-operator",
		senderId: "qa-operator",
		senderName: "QA Operator"
	};
}
function isControlUiProxyPath(pathname) {
	return pathname === "/control-ui" || pathname.startsWith("/control-ui/");
}
function rewriteControlUiProxyPath(pathname, search) {
	return `${pathname === "/control-ui" ? "/" : pathname.slice(11) || "/"}${search}`;
}
function rewriteEmbeddedControlUiHeaders(headers) {
	const rewritten = { ...headers };
	delete rewritten["x-frame-options"];
	const csp = headers["content-security-policy"];
	if (typeof csp === "string") rewritten["content-security-policy"] = csp.includes("frame-ancestors") ? csp.replace(/frame-ancestors\s+[^;]+/i, "frame-ancestors 'self'") : `${csp}; frame-ancestors 'self'`;
	return rewritten;
}
async function proxyHttpRequest(params) {
	const upstreamReq = (params.target.protocol === "https:" ? request$1 : request)({
		protocol: params.target.protocol,
		hostname: params.target.hostname,
		port: params.target.port || (params.target.protocol === "https:" ? 443 : 80),
		method: params.req.method,
		path: rewriteControlUiProxyPath(params.pathname, params.search),
		headers: {
			...params.req.headers,
			host: params.target.host
		}
	}, (upstreamRes) => {
		params.res.writeHead(upstreamRes.statusCode ?? 502, rewriteEmbeddedControlUiHeaders(upstreamRes.headers));
		upstreamRes.pipe(params.res);
	});
	upstreamReq.on("error", (error) => {
		if (!params.res.headersSent) {
			writeError(params.res, 502, error);
			return;
		}
		params.res.destroy(error);
	});
	if (params.req.method === "GET" || params.req.method === "HEAD") {
		upstreamReq.end();
		return;
	}
	params.req.pipe(upstreamReq);
}
function proxyUpgradeRequest(params) {
	const requestUrl = new URL(params.req.url ?? "/", "http://127.0.0.1");
	const port = Number(params.target.port || (params.target.protocol === "https:" ? 443 : 80));
	const upstream = params.target.protocol === "https:" ? tls.connect({
		host: params.target.hostname,
		port,
		servername: params.target.hostname
	}) : net.connect({
		host: params.target.hostname,
		port
	});
	const headerLines = [];
	for (let index = 0; index < params.req.rawHeaders.length; index += 2) {
		const name = params.req.rawHeaders[index];
		const value = params.req.rawHeaders[index + 1] ?? "";
		if (normalizeLowercaseStringOrEmpty(name) === "host") continue;
		headerLines.push(`${name}: ${value}`);
	}
	upstream.once("connect", () => {
		const requestText = [
			`${params.req.method ?? "GET"} ${rewriteControlUiProxyPath(requestUrl.pathname, requestUrl.search)} HTTP/${params.req.httpVersion}`,
			`Host: ${params.target.host}`,
			...headerLines,
			"",
			""
		].join("\r\n");
		upstream.write(requestText);
		if (params.head.length > 0) upstream.write(params.head);
		upstream.pipe(params.socket);
		params.socket.pipe(upstream);
	});
	const closeBoth = () => {
		if (!params.socket.destroyed) params.socket.destroy();
		if (!upstream.destroyed) upstream.destroy();
	};
	upstream.on("error", () => {
		if (!params.socket.destroyed) params.socket.write("HTTP/1.1 502 Bad Gateway\r\nConnection: close\r\n\r\n");
		closeBoth();
	});
	params.socket.on("error", closeBoth);
	params.socket.on("close", closeBoth);
}
function tryResolveUiAsset(pathname, overrideDir, repoRoot = process.cwd()) {
	const distDir = resolveUiDistDir(overrideDir, repoRoot);
	if (!fs.existsSync(distDir)) return null;
	const decoded = decodeURIComponent(pathname === "/" ? "/index.html" : pathname);
	const candidate = path.normalize(path.join(distDir, decoded));
	if (!candidate.startsWith(distDir)) return null;
	if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) return candidate;
	const fallback = path.join(distDir, "index.html");
	return fs.existsSync(fallback) ? fallback : null;
}
function createQaLabConfig(baseUrl) {
	return { channels: { "qa-channel": {
		enabled: true,
		baseUrl,
		botUserId: "openclaw",
		botDisplayName: "OpenClaw QA",
		allowFrom: ["*"]
	} } };
}
async function startQaGatewayLoop(params) {
	setQaChannelRuntime(createQaRunnerRuntime());
	const cfg = createQaLabConfig(params.baseUrl);
	const account = qaChannelPlugin.config.resolveAccount(cfg, "default");
	const abort = new AbortController();
	const task = qaChannelPlugin.gateway?.startAccount?.({
		accountId: account.accountId,
		account,
		cfg,
		runtime: {
			log: () => void 0,
			error: () => void 0,
			exit: () => void 0
		},
		abortSignal: abort.signal,
		log: {
			info: () => void 0,
			warn: () => void 0,
			error: () => void 0,
			debug: () => void 0
		},
		getStatus: () => ({
			accountId: account.accountId,
			configured: true,
			enabled: true,
			running: true
		}),
		setStatus: () => void 0
	});
	return {
		cfg,
		async stop() {
			abort.abort();
			await task;
		}
	};
}
async function startQaLabServer(params) {
	const repoRoot = path.resolve(params?.repoRoot ?? process.cwd());
	const state = createQaBusState();
	let latestReport = null;
	let latestScenarioRun = null;
	const scenarioCatalog = readQaBootstrapScenarioCatalog();
	const bootstrapDefaults = createBootstrapDefaults(params?.autoKickoffTarget);
	let runnerModelOptions = [];
	let runnerModelCatalogStatus = "loading";
	let runnerSnapshot = createIdleQaRunnerSnapshot(scenarioCatalog.scenarios);
	let activeSuiteRun = null;
	let controlUiProxyTarget = params?.controlUiProxyTarget?.trim() ? new URL(params.controlUiProxyTarget) : null;
	let controlUiUrl = params?.controlUiUrl?.trim() || null;
	let controlUiToken = params?.controlUiToken?.trim() || null;
	let gateway;
	const embeddedGatewayEnabled = params?.embeddedGateway !== "disabled";
	let labHandle = null;
	let publicBaseUrl = "";
	let runnerModelCatalogPromise = null;
	const ensureRunnerModelCatalog = () => {
		if (runnerModelCatalogPromise) return runnerModelCatalogPromise;
		runnerModelCatalogPromise = (async () => {
			try {
				const { loadQaRunnerModelOptions } = await import("./model-catalog.runtime-BgVUfNEx.js");
				runnerModelOptions = await loadQaRunnerModelOptions({ repoRoot });
				runnerModelCatalogStatus = "ready";
			} catch {
				runnerModelOptions = [];
				runnerModelCatalogStatus = "failed";
			}
		})();
		return runnerModelCatalogPromise;
	};
	const server = createServer$1(async (req, res) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (await handleQaBusRequest({
			req,
			res,
			state
		})) return;
		try {
			if (controlUiProxyTarget && isControlUiProxyPath(url.pathname)) {
				await proxyHttpRequest({
					req,
					res,
					target: controlUiProxyTarget,
					pathname: url.pathname,
					search: url.search
				});
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/bootstrap") {
				ensureRunnerModelCatalog();
				const resolvedControlUiUrl = controlUiProxyTarget ? `${publicBaseUrl}/control-ui/` : controlUiUrl;
				const controlUiEmbeddedUrl = resolvedControlUiUrl && controlUiToken ? `${resolvedControlUiUrl.replace(/\/?$/, "/")}#token=${encodeURIComponent(controlUiToken)}` : resolvedControlUiUrl;
				writeJson$1(res, 200, {
					baseUrl: publicBaseUrl,
					latestReport,
					controlUiUrl: resolvedControlUiUrl,
					controlUiEmbeddedUrl,
					kickoffTask: scenarioCatalog.kickoffTask,
					scenarios: scenarioCatalog.scenarios,
					defaults: bootstrapDefaults,
					runner: runnerSnapshot,
					runnerCatalog: {
						status: runnerModelCatalogStatus,
						real: runnerModelOptions
					}
				});
				return;
			}
			if (req.method === "GET" && (url.pathname === "/healthz" || url.pathname === "/readyz")) {
				writeJson$1(res, 200, {
					ok: true,
					status: "live"
				});
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/state") {
				writeJson$1(res, 200, state.getSnapshot());
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/report") {
				writeJson$1(res, 200, { report: latestReport });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/ui-version") {
				res.writeHead(200, {
					"content-type": "application/json; charset=utf-8",
					"cache-control": "no-store"
				});
				res.end(JSON.stringify({ version: resolveUiAssetVersion(params?.uiDistDir) }));
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/outcomes") {
				writeJson$1(res, 200, { run: latestScenarioRun });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/reset") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				state.reset();
				latestReport = null;
				latestScenarioRun = null;
				runnerSnapshot = {
					...runnerSnapshot,
					status: "idle",
					artifacts: null,
					error: null,
					startedAt: void 0,
					finishedAt: void 0
				};
				writeJson$1(res, 200, { ok: true });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/inbound/message") {
				const body = await readJson(req);
				writeJson$1(res, 200, { message: state.addInboundMessage(body) });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/kickoff") {
				writeJson$1(res, 200, { message: injectKickoffMessage({
					state,
					defaults: bootstrapDefaults,
					kickoffTask: scenarioCatalog.kickoffTask
				}) });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/scenario/self-check") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				latestScenarioRun = withQaLabRunCounts({
					kind: "self-check",
					status: "running",
					startedAt: (/* @__PURE__ */ new Date()).toISOString(),
					scenarios: [{
						id: "qa-self-check",
						name: "Synthetic Slack-class roundtrip",
						status: "running"
					}]
				});
				const result = await runQaSelfCheckAgainstState({
					state,
					cfg: gateway?.cfg ?? createQaLabConfig(listenUrl),
					outputPath: params?.outputPath,
					repoRoot
				});
				latestScenarioRun = withQaLabRunCounts({
					kind: "self-check",
					status: "completed",
					startedAt: latestScenarioRun.startedAt,
					finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
					scenarios: [{
						id: "qa-self-check",
						name: result.scenarioResult.name,
						status: result.scenarioResult.status,
						details: result.scenarioResult.details,
						steps: result.scenarioResult.steps
					}]
				});
				latestReport = {
					outputPath: result.outputPath,
					markdown: result.report,
					generatedAt: (/* @__PURE__ */ new Date()).toISOString()
				};
				writeJson$1(res, 200, serializeSelfCheck(result));
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/scenario/suite") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				const selection = normalizeQaRunSelection(await readJson(req), scenarioCatalog.scenarios);
				state.reset();
				latestReport = null;
				latestScenarioRun = null;
				const startedAt = (/* @__PURE__ */ new Date()).toISOString();
				runnerSnapshot = {
					status: "running",
					selection,
					startedAt,
					finishedAt: void 0,
					artifacts: null,
					error: null
				};
				activeSuiteRun = (async () => {
					try {
						const { runQaSuiteFromRuntime } = await import("./suite-launch.runtime-B5EpqKfV.js");
						const result = await runQaSuiteFromRuntime({
							lab: labHandle ?? void 0,
							outputDir: createQaRunOutputDir(repoRoot),
							providerMode: selection.providerMode,
							primaryModel: selection.primaryModel,
							alternateModel: selection.alternateModel,
							scenarioIds: selection.scenarioIds
						});
						runnerSnapshot = {
							status: "completed",
							selection,
							startedAt,
							finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
							artifacts: {
								outputDir: result.outputDir,
								reportPath: result.reportPath,
								summaryPath: result.summaryPath,
								watchUrl: result.watchUrl
							},
							error: null
						};
					} catch (error) {
						runnerSnapshot = {
							status: "failed",
							selection,
							startedAt,
							finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
							artifacts: null,
							error: formatErrorMessage(error)
						};
					} finally {
						activeSuiteRun = null;
					}
				})();
				writeJson$1(res, 202, {
					ok: true,
					runner: runnerSnapshot
				});
				return;
			}
			if (req.method !== "GET" && req.method !== "HEAD") {
				writeError(res, 404, "not found");
				return;
			}
			const asset = tryResolveUiAsset(url.pathname, params?.uiDistDir, repoRoot);
			if (!asset) {
				const html = missingUiHtml();
				res.writeHead(200, {
					"content-type": "text/html; charset=utf-8",
					"content-length": Buffer.byteLength(html)
				});
				if (req.method === "HEAD") {
					res.end();
					return;
				}
				res.end(html);
				return;
			}
			const body = fs.readFileSync(asset);
			res.writeHead(200, {
				"content-type": detectContentType(asset),
				"content-length": body.byteLength
			});
			if (req.method === "HEAD") {
				res.end();
				return;
			}
			res.end(body);
		} catch (error) {
			writeError(res, 500, error);
		}
	});
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params?.port ?? 0, params?.host ?? "127.0.0.1", () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa-lab failed to bind");
	const listenUrl = resolveAdvertisedBaseUrl({
		bindHost: params?.host ?? "127.0.0.1",
		bindPort: address.port
	});
	publicBaseUrl = resolveAdvertisedBaseUrl({
		bindHost: params?.host ?? "127.0.0.1",
		bindPort: address.port,
		advertiseHost: params?.advertiseHost,
		advertisePort: params?.advertisePort
	});
	if (embeddedGatewayEnabled) gateway = await startQaGatewayLoop({
		state,
		baseUrl: listenUrl
	});
	if (params?.sendKickoffOnStart) injectKickoffMessage({
		state,
		defaults: bootstrapDefaults,
		kickoffTask: scenarioCatalog.kickoffTask
	});
	server.on("upgrade", (req, socket, head) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (!controlUiProxyTarget || !isControlUiProxyPath(url.pathname)) {
			socket.destroy();
			return;
		}
		proxyUpgradeRequest({
			req,
			socket,
			head,
			target: controlUiProxyTarget
		});
	});
	const lab = {
		baseUrl: publicBaseUrl,
		listenUrl,
		state,
		setControlUi(next) {
			controlUiUrl = next.controlUiUrl?.trim() || null;
			controlUiToken = next.controlUiToken?.trim() || null;
			controlUiProxyTarget = next.controlUiProxyTarget?.trim() ? new URL(next.controlUiProxyTarget) : null;
		},
		setScenarioRun(next) {
			latestScenarioRun = next ? withQaLabRunCounts(next) : null;
		},
		setLatestReport(next) {
			latestReport = next;
		},
		async runSelfCheck() {
			latestScenarioRun = withQaLabRunCounts({
				kind: "self-check",
				status: "running",
				startedAt: (/* @__PURE__ */ new Date()).toISOString(),
				scenarios: [{
					id: "qa-self-check",
					name: "Synthetic Slack-class roundtrip",
					status: "running"
				}]
			});
			const result = await runQaSelfCheckAgainstState({
				state,
				cfg: gateway?.cfg ?? createQaLabConfig(listenUrl),
				outputPath: params?.outputPath,
				repoRoot
			});
			latestScenarioRun = withQaLabRunCounts({
				kind: "self-check",
				status: "completed",
				startedAt: latestScenarioRun.startedAt,
				finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
				scenarios: [{
					id: "qa-self-check",
					name: result.scenarioResult.name,
					status: result.scenarioResult.status,
					details: result.scenarioResult.details,
					steps: result.scenarioResult.steps
				}]
			});
			latestReport = {
				outputPath: result.outputPath,
				markdown: result.report,
				generatedAt: (/* @__PURE__ */ new Date()).toISOString()
			};
			return result;
		},
		async stop() {
			await gateway?.stop();
			await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
		}
	};
	labHandle = lab;
	return lab;
}
function serializeSelfCheck(result) {
	return {
		outputPath: result.outputPath,
		report: result.report,
		checks: result.checks,
		scenario: result.scenarioResult
	};
}
//#endregion
//#region extensions/qa-lab/src/qa-agent-bootstrap.ts
function readQaAgentIdentityMarkdown() {
	return readQaBootstrapScenarioCatalog().agentIdentityMarkdown || DEFAULT_QA_AGENT_IDENTITY_MARKDOWN;
}
function buildQaScenarioPlanMarkdown() {
	const catalog = readQaBootstrapScenarioCatalog();
	const lines = ["# QA Scenario Plan", ""];
	for (const scenario of catalog.scenarios) {
		lines.push(`## ${scenario.title}`);
		lines.push("");
		lines.push(`- id: ${scenario.id}`);
		lines.push(`- surface: ${scenario.surface}`);
		lines.push(`- objective: ${scenario.objective}`);
		if (scenario.execution?.summary) lines.push(`- execution: ${scenario.execution.summary}`);
		lines.push("- success criteria:");
		for (const criterion of scenario.successCriteria) lines.push(`  - ${criterion}`);
		if (scenario.docsRefs?.length) {
			lines.push("- docs:");
			for (const ref of scenario.docsRefs) lines.push(`  - ${ref}`);
		}
		if (scenario.codeRefs?.length) {
			lines.push("- code:");
			for (const ref of scenario.codeRefs) lines.push(`  - ${ref}`);
		}
		lines.push("");
	}
	return lines.join("\n");
}
//#endregion
//#region extensions/qa-lab/src/qa-agent-workspace.ts
async function seedQaAgentWorkspace(params) {
	const catalog = readQaBootstrapScenarioCatalog();
	await fs$1.mkdir(params.workspaceDir, { recursive: true });
	const kickoffTask = catalog.kickoffTask || "QA mission unavailable.";
	const files = new Map([
		["IDENTITY.md", readQaAgentIdentityMarkdown()],
		["QA_KICKOFF_TASK.md", kickoffTask],
		["QA_SCENARIO_PLAN.md", buildQaScenarioPlanMarkdown()],
		["QA_SCENARIOS.md", readQaScenarioPackMarkdown()]
	]);
	if (params.repoRoot) files.set("README.md", `# QA Workspace

- repo: ./repo/
- kickoff: ./QA_KICKOFF_TASK.md
- scenario plan: ./QA_SCENARIO_PLAN.md
- scenario pack: ./QA_SCENARIOS.md
- identity: ./IDENTITY.md

The mounted repo source should be available read-only under \`./repo/\`.
`);
	await Promise.all([...files.entries()].map(async ([name, body]) => {
		await fs$1.writeFile(path.join(params.workspaceDir, name), `${body.trim()}\n`, "utf8");
	}));
	if (params.repoRoot) {
		const repoLinkPath = path.join(params.workspaceDir, "repo");
		await fs$1.rm(repoLinkPath, {
			force: true,
			recursive: true
		});
		await fs$1.symlink(params.repoRoot, repoLinkPath, "dir");
	}
}
//#endregion
//#region extensions/qa-lab/src/mock-openai-server.ts
const TINY_PNG_BASE64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO7Z0nQAAAAASUVORK5CYII=";
let subagentFanoutPhase = 0;
function readBody(req) {
	return new Promise((resolve, reject) => {
		const chunks = [];
		req.on("data", (chunk) => chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)));
		req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
		req.on("error", reject);
	});
}
function writeJson(res, status, body) {
	const text = JSON.stringify(body);
	res.writeHead(status, {
		"content-type": "application/json; charset=utf-8",
		"content-length": Buffer.byteLength(text),
		"cache-control": "no-store"
	});
	res.end(text);
}
function writeSse(res, events) {
	const body = `${events.map((event) => `data: ${JSON.stringify(event)}\n\n`).join("")}data: [DONE]\n\n`;
	res.writeHead(200, {
		"content-type": "text/event-stream",
		"cache-control": "no-store",
		connection: "keep-alive",
		"content-length": Buffer.byteLength(body)
	});
	res.end(body);
}
function extractLastUserText(input) {
	for (let index = input.length - 1; index >= 0; index -= 1) {
		const item = input[index];
		if (item.role !== "user" || !Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) return text;
	}
	return "";
}
function findLastUserIndex(input) {
	for (let index = input.length - 1; index >= 0; index -= 1) {
		const item = input[index];
		if (item.role === "user" && Array.isArray(item.content)) return index;
	}
	return -1;
}
function extractToolOutput(input) {
	const lastUserIndex = findLastUserIndex(input);
	for (let index = input.length - 1; index > lastUserIndex; index -= 1) {
		const item = input[index];
		if (item.type === "function_call_output" && typeof item.output === "string" && item.output) return item.output;
	}
	return "";
}
function extractInputText(content) {
	return content.filter((entry) => !!entry && typeof entry === "object" && entry.type === "input_text" && typeof entry.text === "string").map((entry) => entry.text).join("\n").trim();
}
function extractAllUserTexts(input) {
	const texts = [];
	for (const item of input) {
		if (item.role !== "user" || !Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) texts.push(text);
	}
	return texts;
}
function extractAllInputTexts(input) {
	const texts = [];
	for (const item of input) {
		if (typeof item.output === "string" && item.output.trim()) texts.push(item.output.trim());
		if (!Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) texts.push(text);
	}
	return texts.join("\n");
}
function countImageInputs(input) {
	let count = 0;
	for (const item of input) {
		if (!Array.isArray(item.content)) continue;
		for (const entry of item.content) if (entry && typeof entry === "object" && entry.type === "input_image") count += 1;
	}
	return count;
}
function parseToolOutputJson(toolOutput) {
	if (!toolOutput.trim()) return null;
	try {
		return JSON.parse(toolOutput);
	} catch {
		return null;
	}
}
function normalizePromptPathCandidate(candidate) {
	const trimmed = candidate.trim().replace(/^`+|`+$/g, "");
	if (!trimmed) return null;
	const normalized = trimmed.replace(/^\.\//, "");
	if (normalized.includes("/") || /\.(?:md|json|ts|tsx|js|mjs|cjs|txt|yaml|yml)$/i.test(normalized)) return normalized;
	return null;
}
function readTargetFromPrompt(prompt) {
	const backtickedMatches = Array.from(prompt.matchAll(/`([^`]+)`/g)).map((match) => normalizePromptPathCandidate(match[1] ?? "")).filter((value) => !!value);
	if (backtickedMatches.length > 0) return backtickedMatches[0];
	const quotedMatches = Array.from(prompt.matchAll(/"([^"]+)"/g)).map((match) => normalizePromptPathCandidate(match[1] ?? "")).filter((value) => !!value);
	if (quotedMatches.length > 0) return quotedMatches[0];
	const repoScoped = /\b(?:repo\/[^\s`",)]+|QA_[A-Z_]+\.md)\b/.exec(prompt)?.[0]?.trim();
	if (repoScoped) return repoScoped;
	if (/\bdocs?\b/i.test(prompt)) return "repo/docs/help/testing.md";
	if (/\bscenario|kickoff|qa\b/i.test(prompt)) return "QA_KICKOFF_TASK.md";
	return "repo/package.json";
}
function buildToolCallEventsWithArgs(name, args) {
	const callId = `call_mock_${name}_1`;
	const serialized = JSON.stringify(args);
	return [
		{
			type: "response.output_item.added",
			item: {
				type: "function_call",
				id: `fc_mock_${name}_1`,
				call_id: callId,
				name,
				arguments: ""
			}
		},
		{
			type: "response.function_call_arguments.delta",
			delta: serialized
		},
		{
			type: "response.output_item.done",
			item: {
				type: "function_call",
				id: `fc_mock_${name}_1`,
				call_id: callId,
				name,
				arguments: serialized
			}
		},
		{
			type: "response.completed",
			response: {
				id: `resp_mock_${name}_1`,
				status: "completed",
				output: [{
					type: "function_call",
					id: `fc_mock_${name}_1`,
					call_id: callId,
					name,
					arguments: serialized
				}],
				usage: {
					input_tokens: 64,
					output_tokens: 16,
					total_tokens: 80
				}
			}
		}
	];
}
function extractRememberedFact(userTexts) {
	for (const text of userTexts) {
		const qaCanaryMatch = /\bqa canary code is\s+([A-Za-z0-9-]+)/i.exec(text);
		if (qaCanaryMatch?.[1]) return qaCanaryMatch[1];
	}
	for (const text of userTexts) {
		const match = /remember(?: this fact for later)?:\s*([A-Za-z0-9-]+)/i.exec(text);
		if (match?.[1]) return match[1];
	}
	return null;
}
function extractOrbitCode(text) {
	return /\bORBIT-\d+\b/i.exec(text)?.[0]?.toUpperCase() ?? null;
}
function extractExactReplyDirective(text) {
	return /reply with exactly:\s*([^\n]+)/i.exec(text)?.[1]?.trim() || null;
}
function isHeartbeatPrompt(text) {
	return /Read HEARTBEAT\.md if it exists/i.test(text);
}
function buildAssistantText(input, body) {
	const prompt = extractLastUserText(input);
	const toolOutput = extractToolOutput(input);
	const toolJson = parseToolOutputJson(toolOutput);
	const userTexts = extractAllUserTexts(input);
	const allInputText = extractAllInputTexts(input);
	const rememberedFact = extractRememberedFact(userTexts);
	const model = typeof body.model === "string" ? body.model : "";
	const orbitCode = extractOrbitCode(typeof toolJson?.text === "string" ? toolJson.text : Array.isArray(toolJson?.results) ? JSON.stringify(toolJson.results) : toolOutput);
	const mediaPath = /MEDIA:([^\n]+)/.exec(toolOutput)?.[1]?.trim();
	const exactReplyDirective = extractExactReplyDirective(allInputText);
	const imageInputCount = countImageInputs(input);
	if (/what was the qa canary code/i.test(prompt) && rememberedFact) return `Protocol note: the QA canary code was ${rememberedFact}.`;
	if (/remember this fact/i.test(prompt) && rememberedFact) return `Protocol note: acknowledged. I will remember ${rememberedFact}.`;
	if (/memory unavailable check/i.test(prompt)) return "Protocol note: I checked the available runtime context but could not confirm the hidden memory-only fact, so I will not guess.";
	if (isHeartbeatPrompt(prompt)) return "HEARTBEAT_OK";
	if (/\bmarker\b/i.test(prompt) && exactReplyDirective) return exactReplyDirective;
	if (/visible skill marker/i.test(prompt)) return "VISIBLE-SKILL-OK";
	if (/hot install marker/i.test(prompt)) return "HOT-INSTALL-OK";
	if (/memory tools check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory and the project codename is ${orbitCode}.`;
	if (/tool continuity check/i.test(prompt) && toolOutput) return `Protocol note: model switch handoff confirmed on ${model || "the requested model"}. QA mission from QA_KICKOFF_TASK.md still applies: understand this OpenClaw repo from source + docs before acting.`;
	if (/session memory ranking check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory and the current Project Nebula codename is ${orbitCode}.`;
	if (/thread memory check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory in-thread and the hidden thread codename is ${orbitCode}.`;
	if (/switch(?:ing)? models?/i.test(prompt)) return `Protocol note: model switch acknowledged. Continuing on ${model || "the requested model"}.`;
	if (/(image generation check|capability flip image check)/i.test(prompt) && mediaPath) return `Protocol note: generated the QA lighthouse image successfully.\nMEDIA:${mediaPath}`;
	if (/roundtrip image inspection check/i.test(prompt) && imageInputCount > 0) return "Protocol note: the generated attachment shows the same QA lighthouse scene from the previous step.";
	if (/image understanding check/i.test(prompt) && imageInputCount > 0) return "Protocol note: the attached image is split horizontally, with red on top and blue on the bottom.";
	if (/interrupted by a gateway reload/i.test(prompt) && /subagent recovery worker/i.test(allInputText)) return "RECOVERED-SUBAGENT-OK";
	if (/subagent recovery worker/i.test(prompt)) return "RECOVERED-SUBAGENT-OK";
	if (/fanout worker alpha/i.test(prompt)) return "ALPHA-OK";
	if (/fanout worker beta/i.test(prompt)) return "BETA-OK";
	if (/subagent fanout synthesis check/i.test(prompt) && toolOutput && subagentFanoutPhase >= 2) return "Protocol note: delegated fanout complete. Alpha=ALPHA-OK. Beta=BETA-OK.";
	if (toolOutput && (/\bdelegate\b/i.test(prompt) || /subagent handoff/i.test(prompt))) return `Protocol note: delegated result acknowledged. The bounded subagent task returned and is folded back into the main thread.`;
	if (toolOutput && /worked, failed, blocked|worked\/failed\/blocked|follow-up/i.test(prompt)) return `Worked:\n- Read seeded QA material.\n- Expanded the report structure.\nFailed:\n- None observed in mock mode.\nBlocked:\n- No live provider evidence in this lane.\nFollow-up:\n- Re-run with a real model for qualitative coverage.`;
	if (toolOutput && /lobster invaders/i.test(prompt)) {
		if (toolOutput.includes("QA mission") || toolOutput.includes("Testing")) return "";
		return `Protocol note: Lobster Invaders built at lobster-invaders.html.`;
	}
	if (toolOutput) return `Protocol note: I reviewed the requested material. Evidence snippet: ${toolOutput.replace(/\s+/g, " ").trim().slice(0, 220) || "no content"}`;
	if (prompt) return `Protocol note: acknowledged. Continue with the QA scenario plan and report worked, failed, and blocked items.`;
	return "Protocol note: mock OpenAI server ready.";
}
function buildToolCallEvents(prompt) {
	return buildToolCallEventsWithArgs("read", { path: readTargetFromPrompt(prompt) });
}
function extractPlannedToolName(events) {
	for (const event of events) {
		if (event.type !== "response.output_item.done") continue;
		const item = event.item;
		if (item.type === "function_call" && typeof item.name === "string") return item.name;
	}
}
function buildAssistantEvents(text) {
	const outputItem = {
		type: "message",
		id: "msg_mock_1",
		role: "assistant",
		status: "completed",
		content: [{
			type: "output_text",
			text,
			annotations: []
		}]
	};
	return [
		{
			type: "response.output_item.added",
			item: {
				type: "message",
				id: "msg_mock_1",
				role: "assistant",
				content: [],
				status: "in_progress"
			}
		},
		{
			type: "response.output_item.done",
			item: outputItem
		},
		{
			type: "response.completed",
			response: {
				id: "resp_mock_msg_1",
				status: "completed",
				output: [outputItem],
				usage: {
					input_tokens: 64,
					output_tokens: 24,
					total_tokens: 88
				}
			}
		}
	];
}
async function buildResponsesPayload(body) {
	const input = Array.isArray(body.input) ? body.input : [];
	const prompt = extractLastUserText(input);
	const toolOutput = extractToolOutput(input);
	const toolJson = parseToolOutputJson(toolOutput);
	const isGroupChat = extractAllInputTexts(input).includes("\"is_group_chat\": true");
	const isBaselineUnmentionedChannelChatter = /\bno bot ping here\b/i.test(prompt);
	if (isHeartbeatPrompt(prompt)) return buildAssistantEvents("HEARTBEAT_OK");
	if (/lobster invaders/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_KICKOFF_TASK.md" });
		if (toolOutput.includes("QA mission") || toolOutput.includes("Testing")) return buildToolCallEventsWithArgs("write", {
			path: "lobster-invaders.html",
			content: `<!doctype html>
<html lang="en">
  <head><meta charset="utf-8" /><title>Lobster Invaders</title></head>
  <body><h1>Lobster Invaders</h1><p>Tiny playable stub.</p></body>
</html>`
		});
	}
	if (/memory tools check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "project codename ORBIT-9",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/session memory ranking check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "current Project Nebula codename ORBIT-10",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		const firstPath = typeof first?.path === "string" ? first.path : void 0;
		if (first?.source === "sessions" || firstPath?.startsWith("sessions/")) return buildAssistantEvents("Protocol note: I checked memory and the current Project Nebula codename is ORBIT-10.");
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/thread memory check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "hidden thread codename ORBIT-22",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/(image generation check|capability flip image check)/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("image_generate", {
		prompt: "A QA lighthouse on a dark sea with a tiny protocol droid silhouette.",
		filename: "qa-lighthouse.png",
		size: "1024x1024"
	});
	if (/subagent fanout synthesis check/i.test(prompt)) {
		if (!toolOutput && subagentFanoutPhase === 0) {
			subagentFanoutPhase = 1;
			return buildToolCallEventsWithArgs("sessions_spawn", {
				task: "Fanout worker alpha: inspect the QA workspace and finish with exactly ALPHA-OK.",
				label: "qa-fanout-alpha",
				thread: false
			});
		}
		if (toolOutput && subagentFanoutPhase === 1) {
			subagentFanoutPhase = 2;
			return buildToolCallEventsWithArgs("sessions_spawn", {
				task: "Fanout worker beta: inspect the QA workspace and finish with exactly BETA-OK.",
				label: "qa-fanout-beta",
				thread: false
			});
		}
	}
	if (/tool continuity check/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_KICKOFF_TASK.md" });
	if ((/\bdelegate\b/i.test(prompt) || /subagent handoff/i.test(prompt)) && !toolOutput) return buildToolCallEventsWithArgs("sessions_spawn", {
		task: "Inspect the QA workspace and return one concise protocol note.",
		label: "qa-sidecar",
		thread: false
	});
	if (/(worked, failed, blocked|worked\/failed\/blocked|source and docs)/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_SCENARIO_PLAN.md" });
	if (!toolOutput && /\b(read|inspect|repo|docs|scenario|kickoff)\b/i.test(prompt)) return buildToolCallEvents(prompt);
	if (/visible skill marker/i.test(prompt) && !toolOutput) return buildAssistantEvents("VISIBLE-SKILL-OK");
	if (/hot install marker/i.test(prompt) && !toolOutput) return buildAssistantEvents("HOT-INSTALL-OK");
	if (isGroupChat && isBaselineUnmentionedChannelChatter && !toolOutput) return buildAssistantEvents("NO_REPLY");
	if (/subagent recovery worker/i.test(prompt) && !/interrupted by a gateway reload/i.test(prompt)) await setTimeout$1(6e4);
	return buildAssistantEvents(buildAssistantText(input, body));
}
async function startQaMockOpenAiServer(params) {
	const host = params?.host ?? "127.0.0.1";
	subagentFanoutPhase = 0;
	let lastRequest = null;
	const requests = [];
	const imageGenerationRequests = [];
	const server = createServer$1(async (req, res) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (req.method === "GET" && (url.pathname === "/healthz" || url.pathname === "/readyz")) {
			writeJson(res, 200, {
				ok: true,
				status: "live"
			});
			return;
		}
		if (req.method === "GET" && url.pathname === "/v1/models") {
			writeJson(res, 200, { data: [
				{
					id: "gpt-5.4",
					object: "model"
				},
				{
					id: "gpt-5.4-alt",
					object: "model"
				},
				{
					id: "gpt-image-1",
					object: "model"
				}
			] });
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/last-request") {
			writeJson(res, 200, lastRequest ?? {
				ok: false,
				error: "no request recorded"
			});
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/requests") {
			writeJson(res, 200, requests);
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/image-generations") {
			writeJson(res, 200, imageGenerationRequests);
			return;
		}
		if (req.method === "POST" && url.pathname === "/v1/images/generations") {
			const raw = await readBody(req);
			const body = raw ? JSON.parse(raw) : {};
			imageGenerationRequests.push(body);
			if (imageGenerationRequests.length > 20) imageGenerationRequests.splice(0, imageGenerationRequests.length - 20);
			writeJson(res, 200, { data: [{
				b64_json: TINY_PNG_BASE64,
				revised_prompt: "A QA lighthouse with protocol droid silhouette."
			}] });
			return;
		}
		if (req.method === "POST" && url.pathname === "/v1/responses") {
			const raw = await readBody(req);
			const body = raw ? JSON.parse(raw) : {};
			const input = Array.isArray(body.input) ? body.input : [];
			const events = await buildResponsesPayload(body);
			lastRequest = {
				raw,
				body,
				prompt: extractLastUserText(input),
				allInputText: extractAllInputTexts(input),
				toolOutput: extractToolOutput(input),
				model: typeof body.model === "string" ? body.model : "",
				imageInputCount: countImageInputs(input),
				plannedToolName: extractPlannedToolName(events)
			};
			requests.push(lastRequest);
			if (requests.length > 50) requests.splice(0, requests.length - 50);
			if (body.stream === false) {
				const completion = events.at(-1);
				if (!completion || completion.type !== "response.completed") {
					writeJson(res, 500, { error: "mock completion failed" });
					return;
				}
				writeJson(res, 200, completion.response);
				return;
			}
			writeSse(res, events);
			return;
		}
		writeJson(res, 404, { error: "not found" });
	});
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params?.port ?? 0, host, () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa mock openai failed to bind");
	return {
		baseUrl: `http://${host}:${address.port}`,
		async stop() {
			await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/gateway-rpc-client.ts
function formatQaGatewayRpcError(error, logs) {
	const details = formatErrorMessage(error);
	return /* @__PURE__ */ new Error(`${details}\nGateway logs:\n${logs()}`);
}
let qaGatewayRpcQueue = Promise.resolve();
async function runQueuedQaGatewayRpc(task) {
	const run = qaGatewayRpcQueue.then(task, task);
	qaGatewayRpcQueue = run.then(() => void 0, () => void 0);
	return await run;
}
async function startQaGatewayRpcClient(params) {
	const wrapError = (error) => formatQaGatewayRpcError(error, params.logs);
	let stopped = false;
	return {
		async request(method, rpcParams, opts) {
			if (stopped) throw wrapError(/* @__PURE__ */ new Error("gateway rpc client already stopped"));
			try {
				return await runQueuedQaGatewayRpc(async () => await callGatewayFromCli(method, {
					url: params.wsUrl,
					token: params.token,
					timeout: String(opts?.timeoutMs ?? 2e4),
					expectFinal: opts?.expectFinal,
					json: true
				}, rpcParams ?? {}, {
					expectFinal: opts?.expectFinal,
					progress: false
				}));
			} catch (error) {
				throw wrapError(error);
			}
		},
		async stop() {
			stopped = true;
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/gateway-child.ts
const QA_LIVE_ENV_ALIASES = Object.freeze([
	{
		liveVar: "OPENCLAW_LIVE_OPENAI_KEY",
		providerVar: "OPENAI_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_ANTHROPIC_KEY",
		providerVar: "ANTHROPIC_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_GEMINI_KEY",
		providerVar: "GEMINI_API_KEY"
	}
]);
const QA_MOCK_BLOCKED_ENV_VARS = Object.freeze([
	"ANTHROPIC_API_KEY",
	"ANTHROPIC_OAUTH_TOKEN",
	"AWS_ACCESS_KEY_ID",
	"AWS_BEARER_TOKEN_BEDROCK",
	"AWS_REGION",
	"AWS_SECRET_ACCESS_KEY",
	"AWS_SESSION_TOKEN",
	"GEMINI_API_KEY",
	"GEMINI_API_KEYS",
	"GOOGLE_API_KEY",
	"MISTRAL_API_KEY",
	"OPENAI_API_KEY",
	"OPENAI_API_KEYS",
	"OPENAI_BASE_URL",
	"OPENCLAW_LIVE_ANTHROPIC_KEY",
	"OPENCLAW_LIVE_ANTHROPIC_KEYS",
	"OPENCLAW_LIVE_GEMINI_KEY",
	"OPENCLAW_LIVE_OPENAI_KEY",
	"VOYAGE_API_KEY"
]);
async function getFreePort() {
	return await new Promise((resolve, reject) => {
		const server = net.createServer();
		server.once("error", reject);
		server.listen(0, "127.0.0.1", () => {
			const address = server.address();
			if (!address || typeof address === "string") {
				reject(/* @__PURE__ */ new Error("failed to allocate port"));
				return;
			}
			server.close((error) => error ? reject(error) : resolve(address.port));
		});
	});
}
function normalizeQaProviderModeEnv(env, providerMode) {
	if (providerMode === "mock-openai") {
		for (const key of QA_MOCK_BLOCKED_ENV_VARS) delete env[key];
		return env;
	}
	if (providerMode === "live-frontier") for (const { liveVar, providerVar } of QA_LIVE_ENV_ALIASES) {
		const liveValue = env[liveVar]?.trim();
		if (!liveValue || env[providerVar]?.trim()) continue;
		env[providerVar] = liveValue;
	}
	return env;
}
function buildQaRuntimeEnv(params) {
	return normalizeQaProviderModeEnv({
		...params.baseEnv ?? process.env,
		HOME: params.homeDir,
		OPENCLAW_HOME: params.homeDir,
		OPENCLAW_CONFIG_PATH: params.configPath,
		OPENCLAW_STATE_DIR: params.stateDir,
		OPENCLAW_OAUTH_DIR: path.join(params.stateDir, "credentials"),
		OPENCLAW_GATEWAY_TOKEN: params.gatewayToken,
		OPENCLAW_SKIP_BROWSER_CONTROL_SERVER: "1",
		OPENCLAW_SKIP_GMAIL_WATCHER: "1",
		OPENCLAW_SKIP_CANVAS_HOST: "1",
		OPENCLAW_NO_RESPAWN: "1",
		OPENCLAW_TEST_FAST: "1",
		OPENCLAW_QA_ALLOW_LOCAL_IMAGE_PROVIDER: "1",
		OPENCLAW_ALLOW_SLOW_REPLY_TESTS: "1",
		XDG_CONFIG_HOME: params.xdgConfigHome,
		XDG_DATA_HOME: params.xdgDataHome,
		XDG_CACHE_HOME: params.xdgCacheHome
	}, params.providerMode);
}
function isRetryableGatewayCallError(details) {
	return details.includes("gateway closed (1012)") || details.includes("gateway closed (1006") || details.includes("abnormal closure") || details.includes("service restart");
}
const __testing = {
	buildQaRuntimeEnv,
	isRetryableGatewayCallError
};
async function waitForGatewayReady(params) {
	const startedAt = Date.now();
	while (Date.now() - startedAt < (params.timeoutMs ?? 6e4)) {
		if (params.child.exitCode !== null || params.child.signalCode !== null) throw new Error(`gateway exited before becoming healthy (exitCode=${String(params.child.exitCode)}, signal=${String(params.child.signalCode)}):\n${params.logs()}`);
		for (const healthPath of ["/readyz", "/healthz"]) try {
			if ((await fetch(`${params.baseUrl}${healthPath}`, { signal: AbortSignal.timeout(2e3) })).ok) return;
		} catch {}
		await setTimeout$1(250);
	}
	throw new Error(`gateway failed to become healthy:\n${params.logs()}`);
}
function resolveQaControlUiRoot(params) {
	if (params.controlUiEnabled === false) return;
	const controlUiRoot = path.join(params.repoRoot, "dist", "control-ui");
	return existsSync(path.join(controlUiRoot, "index.html")) ? controlUiRoot : void 0;
}
async function startQaGatewayChild(params) {
	const tempRoot = await fs$1.mkdtemp(path.join(os.tmpdir(), "openclaw-qa-suite-"));
	const runtimeCwd = tempRoot;
	const distEntryPath = path.join(params.repoRoot, "dist", "index.js");
	const workspaceDir = path.join(tempRoot, "workspace");
	const stateDir = path.join(tempRoot, "state");
	const homeDir = path.join(tempRoot, "home");
	const xdgConfigHome = path.join(tempRoot, "xdg-config");
	const xdgDataHome = path.join(tempRoot, "xdg-data");
	const xdgCacheHome = path.join(tempRoot, "xdg-cache");
	const configPath = path.join(tempRoot, "openclaw.json");
	const gatewayPort = await getFreePort();
	const gatewayToken = `qa-suite-${randomUUID()}`;
	await seedQaAgentWorkspace({
		workspaceDir,
		repoRoot: params.repoRoot
	});
	await Promise.all([
		fs$1.mkdir(stateDir, { recursive: true }),
		fs$1.mkdir(homeDir, { recursive: true }),
		fs$1.mkdir(xdgConfigHome, { recursive: true }),
		fs$1.mkdir(xdgDataHome, { recursive: true }),
		fs$1.mkdir(xdgCacheHome, { recursive: true })
	]);
	const cfg = buildQaGatewayConfig({
		bind: "loopback",
		gatewayPort,
		gatewayToken,
		providerBaseUrl: params.providerBaseUrl,
		qaBusBaseUrl: params.qaBusBaseUrl,
		workspaceDir,
		controlUiRoot: resolveQaControlUiRoot({
			repoRoot: params.repoRoot,
			controlUiEnabled: params.controlUiEnabled
		}),
		controlUiAllowedOrigins: params.controlUiAllowedOrigins,
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		fastMode: params.fastMode,
		controlUiEnabled: params.controlUiEnabled
	});
	await fs$1.writeFile(configPath, `${JSON.stringify(cfg, null, 2)}\n`, "utf8");
	const stdout = [];
	const stderr = [];
	const env = buildQaRuntimeEnv({
		configPath,
		gatewayToken,
		homeDir,
		stateDir,
		xdgConfigHome,
		xdgDataHome,
		xdgCacheHome,
		providerMode: params.providerMode
	});
	const child = spawn(process.execPath, [
		distEntryPath,
		"gateway",
		"run",
		"--port",
		String(gatewayPort),
		"--bind",
		"loopback",
		"--allow-unconfigured"
	], {
		cwd: runtimeCwd,
		env,
		stdio: [
			"ignore",
			"pipe",
			"pipe"
		]
	});
	child.stdout.on("data", (chunk) => stdout.push(Buffer.from(chunk)));
	child.stderr.on("data", (chunk) => stderr.push(Buffer.from(chunk)));
	const baseUrl = `http://127.0.0.1:${gatewayPort}`;
	const wsUrl = `ws://127.0.0.1:${gatewayPort}`;
	const logs = () => `${Buffer.concat(stdout).toString("utf8")}\n${Buffer.concat(stderr).toString("utf8")}`.trim();
	const keepTemp = process.env.OPENCLAW_QA_KEEP_TEMP === "1";
	let rpcClient;
	try {
		await waitForGatewayReady({
			baseUrl,
			logs,
			child
		});
		rpcClient = await startQaGatewayRpcClient({
			wsUrl,
			token: gatewayToken,
			logs
		});
	} catch (error) {
		child.kill("SIGTERM");
		throw error;
	}
	return {
		cfg,
		baseUrl,
		wsUrl,
		pid: child.pid ?? null,
		token: gatewayToken,
		workspaceDir,
		tempRoot,
		configPath,
		runtimeEnv: env,
		logs,
		async restart(signal = "SIGUSR1") {
			if (!child.pid) throw new Error("qa gateway child has no pid");
			process.kill(child.pid, signal);
		},
		async call(method, rpcParams, opts) {
			const timeoutMs = opts?.timeoutMs ?? 2e4;
			let lastDetails = "";
			for (let attempt = 1; attempt <= 3; attempt += 1) try {
				return await rpcClient.request(method, rpcParams, {
					...opts,
					timeoutMs
				});
			} catch (error) {
				const details = formatErrorMessage(error);
				lastDetails = details;
				if (attempt >= 3 || !isRetryableGatewayCallError(details)) throw new Error(`${details}\nGateway logs:\n${logs()}`, { cause: error });
				await waitForGatewayReady({
					baseUrl,
					logs,
					child,
					timeoutMs: Math.max(1e4, timeoutMs)
				});
			}
			throw new Error(`${lastDetails}\nGateway logs:\n${logs()}`);
		},
		async stop(opts) {
			await rpcClient.stop().catch(() => {});
			if (!child.killed) {
				child.kill("SIGTERM");
				await Promise.race([new Promise((resolve) => child.once("exit", () => resolve())), setTimeout$1(5e3).then(() => {
					if (!child.killed) child.kill("SIGKILL");
				})]);
			}
			if (!(opts?.keepTemp ?? keepTemp)) await fs$1.rm(tempRoot, {
				recursive: true,
				force: true
			});
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/cron-run-wait.ts
async function waitForCronRunCompletion(params) {
	const timeoutMs = params.timeoutMs ?? 9e4;
	const intervalMs = params.intervalMs ?? 1e3;
	const startedAt = Date.now();
	let lastEntries = [];
	while (Date.now() - startedAt < timeoutMs) {
		const page = await params.callGateway("cron.runs", {
			id: params.jobId,
			limit: 20,
			sortDir: "desc"
		}, { timeoutMs: Math.min(timeoutMs, 3e4) });
		const entries = Array.isArray(page.entries) ? page.entries : [];
		lastEntries = entries;
		const completed = entries.find((entry) => typeof entry.ts === "number" && entry.ts >= params.afterTs && (entry.status === "ok" || entry.status === "error" || entry.status === "skipped"));
		if (completed) return completed;
		await setTimeout$1(intervalMs);
	}
	throw new Error(`timed out waiting for cron run completion for ${params.jobId}: ${formatErrorMessage(lastEntries)}`);
}
//#endregion
//#region extensions/qa-lab/src/discovery-eval.ts
const REQUIRED_DISCOVERY_REFS_LOWER = [
	"repo/qa/scenarios.md",
	"repo/extensions/qa-lab/src/suite.ts",
	"repo/docs/help/testing.md"
].map(normalizeLowercaseStringOrEmpty);
const DISCOVERY_SCOPE_LEAK_PHRASES = [
	"all mandatory scenarios",
	"final qa tally",
	"final qa tally update",
	"qa run complete",
	"scenario: `subagent-handoff`",
	"scenario: subagent-handoff"
];
function confirmsDiscoveryFileRead(text) {
	const lower = normalizeLowercaseStringOrEmpty(text);
	const mentionsAllRefs = REQUIRED_DISCOVERY_REFS_LOWER.every((ref) => lower.includes(ref));
	const requiredCountPattern = "(?:three|3|four|4)";
	const confirmsRead = new RegExp(`(?:read|retrieved|inspected|loaded|accessed|digested)\\s+all\\s+${requiredCountPattern}\\s+(?:(?:requested|required|mandated|seeded)\\s+)?files`).test(lower) || new RegExp(`all\\s+${requiredCountPattern}\\s+(?:(?:requested|required|mandated|seeded)\\s+)?files\\s+(?:were\\s+)?(?:read|retrieved|inspected|loaded|accessed|digested)(?:\\s+\\w+)?`).test(lower) || new RegExp(`all\\s+${requiredCountPattern}\\s+seeded files readable`).test(lower);
	return mentionsAllRefs && confirmsRead;
}
function hasDiscoveryLabels(text) {
	const lower = normalizeLowercaseStringOrEmpty(text);
	return lower.includes("worked") && lower.includes("failed") && lower.includes("blocked") && (lower.includes("follow-up") || lower.includes("follow up"));
}
function reportsMissingDiscoveryFiles(text) {
	const lower = normalizeLowercaseStringOrEmpty(text);
	if (confirmsDiscoveryFileRead(text)) return false;
	return lower.includes("not present") || lower.includes("missing files") || lower.includes("blocked by missing") || lower.includes("could not inspect");
}
function reportsDiscoveryScopeLeak(text) {
	const lower = normalizeLowercaseStringOrEmpty(text);
	return DISCOVERY_SCOPE_LEAK_PHRASES.some((phrase) => lower.includes(phrase));
}
//#endregion
//#region extensions/qa-lab/src/live-timeout.ts
function isAnthropicModel(modelRef) {
	return modelRef.startsWith("anthropic/");
}
function isClaudeOpusModel(modelRef) {
	return isAnthropicModel(modelRef) && modelRef.includes("claude-opus");
}
function resolveQaLiveTurnTimeoutMs(profile, fallbackMs, modelRef = profile.primaryModel) {
	if (profile.providerMode === "mock-openai") return fallbackMs;
	if (isClaudeOpusModel(modelRef)) return Math.max(fallbackMs, 24e4);
	if (isAnthropicModel(modelRef)) return Math.max(fallbackMs, 18e4);
	return Math.max(fallbackMs, 12e4);
}
//#endregion
//#region extensions/qa-lab/src/model-switch-eval.ts
function hasModelSwitchContinuityEvidence(text) {
	const lower = normalizeLowercaseStringOrEmpty(text);
	const mentionsHandoff = lower.includes("handoff") || lower.includes("model switch") || lower.includes("switched");
	const mentionsKickoffTask = lower.includes("qa_kickoff_task") || lower.includes("kickoff task") || lower.includes("qa mission");
	const hasScopeLeak = lower.includes("subagent-handoff") || lower.includes("delegated task") || lower.includes("final qa tally") || lower.includes("qa run complete") || lower.includes("all mandatory scenarios");
	const looksOverlong = text.length > 280 || text.includes("\n\n") || text.includes("|---") || text.includes("### ");
	return mentionsHandoff && mentionsKickoffTask && !hasScopeLeak && !looksOverlong;
}
//#endregion
//#region extensions/qa-lab/src/suite.ts
const QA_IMAGE_UNDERSTANDING_PNG_BASE64 = "iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAIAAAAlC+aJAAAAT0lEQVR42u3RQQkAMAzAwPg33Wnos+wgBo40dboAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANYADwAAAAAAAAAAAAAAAAAAAAAAAAAAAAC+Azy47PDiI4pA2wAAAABJRU5ErkJggg==";
function splitModelRef(ref) {
	const slash = ref.indexOf("/");
	if (slash <= 0 || slash === ref.length - 1) return null;
	return {
		provider: ref.slice(0, slash),
		model: ref.slice(slash + 1)
	};
}
function liveTurnTimeoutMs(env, fallbackMs) {
	return resolveQaLiveTurnTimeoutMs(env, fallbackMs);
}
function createQaActionConfig(baseUrl) {
	return { channels: { "qa-channel": {
		enabled: true,
		baseUrl,
		botUserId: "openclaw",
		botDisplayName: "OpenClaw QA",
		allowFrom: ["*"]
	} } };
}
async function waitForCondition(check, timeoutMs = 15e3, intervalMs = 100) {
	const startedAt = Date.now();
	while (Date.now() - startedAt < timeoutMs) {
		const value = await check();
		if (value !== null && value !== void 0) return value;
		await setTimeout$1(intervalMs);
	}
	throw new Error(`timed out after ${timeoutMs}ms`);
}
async function waitForOutboundMessage(state, predicate, timeoutMs = 15e3) {
	return await waitForCondition(() => state.getSnapshot().messages.filter((message) => message.direction === "outbound").find(predicate), timeoutMs);
}
async function waitForNoOutbound(state, timeoutMs = 1200) {
	await setTimeout$1(timeoutMs);
	const outbound = state.getSnapshot().messages.filter((message) => message.direction === "outbound");
	if (outbound.length > 0) throw new Error(`expected no outbound messages, saw ${outbound.length}`);
}
function recentOutboundSummary(state, limit = 5) {
	return state.getSnapshot().messages.filter((message) => message.direction === "outbound").slice(-limit).map((message) => `${message.conversation.id}:${message.text}`).join(" | ");
}
async function runScenario(name, steps) {
	const stepResults = [];
	for (const step of steps) try {
		if (process.env.OPENCLAW_QA_DEBUG === "1") console.error(`[qa-suite] start scenario="${name}" step="${step.name}"`);
		const details = await step.run();
		if (process.env.OPENCLAW_QA_DEBUG === "1") console.error(`[qa-suite] pass scenario="${name}" step="${step.name}"`);
		stepResults.push({
			name: step.name,
			status: "pass",
			...details ? { details } : {}
		});
	} catch (error) {
		const details = formatErrorMessage(error);
		if (process.env.OPENCLAW_QA_DEBUG === "1") console.error(`[qa-suite] fail scenario="${name}" step="${step.name}" details=${details}`);
		stepResults.push({
			name: step.name,
			status: "fail",
			details
		});
		return {
			name,
			status: "fail",
			steps: stepResults,
			details
		};
	}
	return {
		name,
		status: "pass",
		steps: stepResults
	};
}
async function fetchJson(url) {
	const response = await fetch(url);
	if (!response.ok) throw new Error(`request failed ${response.status}: ${url}`);
	return await response.json();
}
async function waitForGatewayHealthy(env, timeoutMs = 45e3) {
	await waitForCondition(async () => {
		try {
			return (await fetch(`${env.gateway.baseUrl}/readyz`)).ok ? true : void 0;
		} catch {
			return;
		}
	}, timeoutMs, 250);
}
async function waitForQaChannelReady(env, timeoutMs = 45e3) {
	await waitForCondition(async () => {
		try {
			const accounts = (await env.gateway.call("channels.status", {
				probe: false,
				timeoutMs: 2e3
			}, { timeoutMs: 5e3 })).channelAccounts?.["qa-channel"] ?? [];
			const account = accounts.find((entry) => entry.accountId === "default") ?? accounts[0];
			if (account?.running && account.restartPending !== true) return true;
			return;
		} catch {
			return;
		}
	}, timeoutMs, 500);
}
async function waitForConfigRestartSettle(env, restartDelayMs = 1e3, timeoutMs = 6e4) {
	await setTimeout$1(restartDelayMs + 750);
	await waitForGatewayHealthy(env, timeoutMs);
}
function isGatewayRestartRace(error) {
	const text = formatErrorMessage(error);
	return text.includes("gateway closed (1012)") || text.includes("gateway closed (1006") || text.includes("abnormal closure") || text.includes("service restart");
}
function isConfigHashConflict(error) {
	return formatErrorMessage(error).includes("config changed since last load");
}
async function readConfigSnapshot(env) {
	const snapshot = await env.gateway.call("config.get", {}, { timeoutMs: 6e4 });
	if (!snapshot.hash || !snapshot.config) throw new Error("config.get returned no hash/config");
	return {
		hash: snapshot.hash,
		config: snapshot.config
	};
}
async function runConfigMutation(params) {
	const restartDelayMs = params.restartDelayMs ?? 1e3;
	let lastConflict = null;
	for (let attempt = 1; attempt <= 3; attempt += 1) {
		const snapshot = await readConfigSnapshot(params.env);
		try {
			const result = await params.env.gateway.call(params.action, {
				raw: params.raw,
				baseHash: snapshot.hash,
				...params.sessionKey ? { sessionKey: params.sessionKey } : {},
				...params.note ? { note: params.note } : {},
				restartDelayMs
			}, { timeoutMs: 45e3 });
			await waitForConfigRestartSettle(params.env, restartDelayMs);
			return result;
		} catch (error) {
			if (isConfigHashConflict(error)) {
				lastConflict = error;
				await waitForGatewayHealthy(params.env, Math.max(15e3, restartDelayMs + 1e4)).catch(() => void 0);
				continue;
			}
			if (!isGatewayRestartRace(error)) throw error;
			await waitForConfigRestartSettle(params.env, restartDelayMs);
			return {
				ok: true,
				restarted: true
			};
		}
	}
	throw lastConflict ?? /* @__PURE__ */ new Error(`${params.action} failed after retrying config hash conflicts`);
}
async function patchConfig(params) {
	return await runConfigMutation({
		env: params.env,
		action: "config.patch",
		raw: JSON.stringify(params.patch, null, 2),
		sessionKey: params.sessionKey,
		note: params.note,
		restartDelayMs: params.restartDelayMs
	});
}
async function applyConfig(params) {
	return await runConfigMutation({
		env: params.env,
		action: "config.apply",
		raw: JSON.stringify(params.nextConfig, null, 2),
		sessionKey: params.sessionKey,
		note: params.note,
		restartDelayMs: params.restartDelayMs
	});
}
async function createSession(env, label, key) {
	const sessionKey = (await env.gateway.call("sessions.create", {
		label,
		...key ? { key } : {}
	}, { timeoutMs: liveTurnTimeoutMs(env, 6e4) })).key?.trim();
	if (!sessionKey) throw new Error("sessions.create returned no key");
	return sessionKey;
}
async function readEffectiveTools(env, sessionKey) {
	const payload = await env.gateway.call("tools.effective", { sessionKey }, { timeoutMs: liveTurnTimeoutMs(env, 9e4) });
	const ids = /* @__PURE__ */ new Set();
	for (const group of payload.groups ?? []) for (const tool of group.tools ?? []) if (tool.id?.trim()) ids.add(tool.id.trim());
	return ids;
}
async function readSkillStatus(env, agentId = "qa") {
	return (await env.gateway.call("skills.status", { agentId }, { timeoutMs: liveTurnTimeoutMs(env, 45e3) })).skills ?? [];
}
async function readRawQaSessionStore(env) {
	const storePath = path.join(env.gateway.tempRoot, "state", "agents", "qa", "sessions", "sessions.json");
	try {
		const raw = await fs$1.readFile(storePath, "utf8");
		return JSON.parse(raw);
	} catch (error) {
		if (error.code === "ENOENT") return {};
		throw error;
	}
}
async function runQaCli(env, args, opts) {
	const stdout = [];
	const stderr = [];
	const distEntryPath = path.join(env.repoRoot, "dist", "index.js");
	await new Promise((resolve, reject) => {
		const child = spawn(process.execPath, [distEntryPath, ...args], {
			cwd: env.gateway.tempRoot,
			env: env.gateway.runtimeEnv,
			stdio: [
				"ignore",
				"pipe",
				"pipe"
			]
		});
		const timeout = setTimeout(() => {
			child.kill("SIGKILL");
			reject(/* @__PURE__ */ new Error(`qa cli timed out: openclaw ${args.join(" ")}`));
		}, opts?.timeoutMs ?? 6e4);
		child.stdout.on("data", (chunk) => stdout.push(Buffer.from(chunk)));
		child.stderr.on("data", (chunk) => stderr.push(Buffer.from(chunk)));
		child.once("error", (error) => {
			clearTimeout(timeout);
			reject(error);
		});
		child.once("exit", (code) => {
			clearTimeout(timeout);
			if (code === 0) {
				resolve();
				return;
			}
			reject(/* @__PURE__ */ new Error(`qa cli failed (${code ?? "unknown"}): ${Buffer.concat(stderr).toString("utf8").trim()}`));
		});
	});
	const text = Buffer.concat(stdout).toString("utf8").trim();
	if (!opts?.json) return text;
	return text ? JSON.parse(text) : {};
}
function extractMediaPathFromText(text) {
	return /MEDIA:([^\n]+)/.exec(text ?? "")?.[1]?.trim();
}
async function resolveGeneratedImagePath(params) {
	return await waitForCondition(async () => {
		if (params.env.mock) {
			const requests = await fetchJson(`${params.env.mock.baseUrl}/debug/requests`);
			for (let index = requests.length - 1; index >= 0; index -= 1) {
				const request = requests[index];
				if (!String(request.allInputText ?? "").includes(params.promptSnippet)) continue;
				const mediaPath = extractMediaPathFromText(request.toolOutput);
				if (mediaPath) return mediaPath;
			}
		}
		const mediaDir = path.join(params.env.gateway.tempRoot, "media", "tool-image-generation");
		const entries = await fs$1.readdir(mediaDir).catch(() => []);
		return (await Promise.all(entries.map(async (entry) => {
			const fullPath = path.join(mediaDir, entry);
			const stat = await fs$1.stat(fullPath).catch(() => null);
			if (!stat?.isFile()) return null;
			return {
				fullPath,
				mtimeMs: stat.mtimeMs
			};
		}))).filter((entry) => Boolean(entry)).filter((entry) => entry.mtimeMs >= params.startedAtMs - 1e3).toSorted((left, right) => right.mtimeMs - left.mtimeMs).at(0)?.fullPath;
	}, params.timeoutMs, 250);
}
async function startAgentRun(env, params) {
	const target = params.to ?? "dm:qa-operator";
	const started = await env.gateway.call("agent", {
		idempotencyKey: randomUUID(),
		agentId: "qa",
		sessionKey: params.sessionKey,
		message: params.message,
		deliver: true,
		channel: "qa-channel",
		to: target,
		replyChannel: "qa-channel",
		replyTo: target,
		...params.threadId ? { threadId: params.threadId } : {},
		...params.provider ? { provider: params.provider } : {},
		...params.model ? { model: params.model } : {},
		...params.attachments ? { attachments: params.attachments } : {}
	}, { timeoutMs: params.timeoutMs ?? 3e4 });
	if (!started.runId) throw new Error(`agent call did not return a runId: ${JSON.stringify(started)}`);
	return started;
}
async function waitForAgentRun(env, runId, timeoutMs = 3e4) {
	return await env.gateway.call("agent.wait", {
		runId,
		timeoutMs
	}, { timeoutMs: timeoutMs + 5e3 });
}
async function listCronJobs(env) {
	return (await env.gateway.call("cron.list", {
		includeDisabled: true,
		limit: 200,
		sortBy: "name",
		sortDir: "asc"
	}, { timeoutMs: 3e4 })).jobs ?? [];
}
async function readDoctorMemoryStatus(env) {
	return await env.gateway.call("doctor.memory.status", {}, { timeoutMs: 3e4 });
}
async function forceMemoryIndex(params) {
	await waitForGatewayHealthy(params.env, 6e4);
	await waitForQaChannelReady(params.env, 6e4);
	await runQaCli(params.env, [
		"memory",
		"index",
		"--agent",
		"qa",
		"--force"
	], { timeoutMs: liveTurnTimeoutMs(params.env, 6e4) });
	const payload = await waitForCondition(async () => {
		const result = await runQaCli(params.env, [
			"memory",
			"search",
			"--agent",
			"qa",
			"--json",
			"--query",
			params.query
		], {
			timeoutMs: liveTurnTimeoutMs(params.env, 6e4),
			json: true
		});
		return JSON.stringify(result.results ?? []).includes(params.expectedNeedle) ? result : void 0;
	}, liveTurnTimeoutMs(params.env, 2e4), 500);
	const haystack = JSON.stringify(payload.results ?? []);
	if (!haystack.includes(params.expectedNeedle)) throw new Error(`memory index missing expected fact after reindex: ${haystack}`);
}
function findSkill(skills, name) {
	return skills.find((skill) => skill.name === name);
}
async function writeWorkspaceSkill(params) {
	const skillDir = path.join(params.env.gateway.workspaceDir, "skills", params.name);
	await fs$1.mkdir(skillDir, { recursive: true });
	const skillPath = path.join(skillDir, "SKILL.md");
	await fs$1.writeFile(skillPath, `${params.body.trim()}\n`, "utf8");
	return skillPath;
}
async function callPluginToolsMcp(params) {
	const transportEnv = Object.fromEntries(Object.entries(params.env.gateway.runtimeEnv).filter((entry) => typeof entry[1] === "string"));
	const transport = new StdioClientTransport({
		command: process.execPath,
		args: [
			"--import",
			"tsx",
			"src/mcp/plugin-tools-serve.ts"
		],
		stderr: "pipe",
		env: transportEnv
	});
	const client = new Client({
		name: "openclaw-qa-suite",
		version: "0.0.0"
	}, {});
	try {
		await client.connect(transport);
		if (!(await client.listTools()).tools.find((entry) => entry.name === params.toolName)) throw new Error(`MCP tool missing: ${params.toolName}`);
		return await client.callTool({
			name: params.toolName,
			arguments: params.args
		});
	} finally {
		await client.close().catch(() => {});
	}
}
async function runAgentPrompt(env, params) {
	const started = await startAgentRun(env, params);
	const waited = await waitForAgentRun(env, started.runId, params.timeoutMs ?? 3e4);
	if (waited.status !== "ok") throw new Error(`agent.wait returned ${String(waited.status ?? "unknown")}: ${waited.error ?? "no error"}`);
	return {
		started,
		waited
	};
}
async function ensureImageGenerationConfigured(env) {
	const imageModelRef = "openai/gpt-image-1";
	await patchConfig({
		env,
		patch: env.providerMode === "mock-openai" ? {
			plugins: {
				allow: [
					"memory-core",
					"openai",
					"qa-channel"
				],
				entries: { openai: { enabled: true } }
			},
			models: { providers: { openai: {
				baseUrl: `${env.mock?.baseUrl}/v1`,
				apiKey: "test",
				api: "openai-responses",
				models: [{
					id: "gpt-image-1",
					name: "gpt-image-1",
					api: "openai-responses",
					reasoning: false,
					input: ["text"],
					cost: {
						input: 0,
						output: 0,
						cacheRead: 0,
						cacheWrite: 0
					},
					contextWindow: 128e3,
					maxTokens: 4096
				}]
			} } },
			agents: { defaults: { imageGenerationModel: { primary: imageModelRef } } }
		} : { agents: { defaults: { imageGenerationModel: { primary: imageModelRef } } } }
	});
	await waitForGatewayHealthy(env);
	await waitForQaChannelReady(env, 6e4);
}
async function handleQaAction(params) {
	return extractQaToolPayload(await qaChannelPlugin.actions?.handleAction?.({
		channel: "qa-channel",
		action: params.action,
		cfg: params.env.cfg,
		accountId: "default",
		params: params.args
	}));
}
function buildScenarioMap(env) {
	const state = env.lab.state;
	const reset = async () => {
		state.reset();
		await setTimeout$1(100);
	};
	return new Map([
		["channel-chat-baseline", async () => await runScenario("Channel baseline conversation", [{
			name: "ignores unmentioned channel chatter",
			run: async () => {
				await reset();
				state.addInboundMessage({
					conversation: {
						id: "qa-room",
						kind: "channel",
						title: "QA Room"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "hello team, no bot ping here"
				});
				await waitForNoOutbound(state);
			}
		}, {
			name: "replies when mentioned in channel",
			run: async () => {
				state.addInboundMessage({
					conversation: {
						id: "qa-room",
						kind: "channel",
						title: "QA Room"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "@openclaw explain the QA lab"
				});
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-room" && !candidate.threadId, env.providerMode === "mock-openai" ? 45e3 : 45e3)).text;
			}
		}])],
		["cron-one-minute-ping", async () => await runScenario("Cron one-minute ping", [{
			name: "stores a reminder roughly one minute ahead",
			run: async () => {
				await reset();
				const at = new Date(Date.now() + 6e4).toISOString();
				const cronMarker = `QA-CRON-${randomUUID().slice(0, 8)}`;
				const response = await env.gateway.call("cron.add", {
					name: `qa-suite-${randomUUID()}`,
					enabled: true,
					schedule: {
						kind: "at",
						at
					},
					sessionTarget: "isolated",
					wakeMode: "next-heartbeat",
					payload: {
						kind: "agentTurn",
						message: `A QA cron just fired. Send a one-line ping back to the room containing this exact marker: ${cronMarker}`
					},
					delivery: {
						mode: "announce",
						channel: "qa-channel",
						to: "channel:qa-room"
					}
				});
				const scheduledAt = response.schedule?.at ?? at;
				const delta = new Date(scheduledAt).getTime() - Date.now();
				if (delta < 45e3 || delta > 75e3) throw new Error(`expected ~1 minute schedule, got ${delta}ms`);
				globalThis.__qaCronJobId = response.id;
				globalThis.__qaCronMarker = cronMarker;
				return scheduledAt;
			}
		}, {
			name: "forces the reminder through QA channel delivery",
			run: async () => {
				const jobId = globalThis.__qaCronJobId;
				const cronMarker = globalThis.__qaCronMarker;
				if (!jobId) throw new Error("missing cron job id");
				if (!cronMarker) throw new Error("missing cron marker");
				await env.gateway.call("cron.run", {
					id: jobId,
					mode: "force"
				}, { timeoutMs: 3e4 });
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-room" && candidate.text.includes(cronMarker), liveTurnTimeoutMs(env, 3e4))).text;
			}
		}])],
		["dm-chat-baseline", async () => await runScenario("DM baseline conversation", [{
			name: "replies coherently in DM",
			run: async () => {
				await reset();
				state.addInboundMessage({
					conversation: {
						id: "alice",
						kind: "direct"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "Hello there, who are you?"
				});
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "alice")).text;
			}
		}])],
		["lobster-invaders-build", async () => await runScenario("Build Lobster Invaders", [{
			name: "creates the artifact after reading context",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:lobster-invaders",
					message: "Read the QA kickoff context first, then build a tiny Lobster Invaders HTML game in this workspace and tell me where it is.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator");
				const artifactPath = path.join(env.gateway.workspaceDir, "lobster-invaders.html");
				if (!(await fs$1.readFile(artifactPath, "utf8")).includes("Lobster Invaders")) throw new Error("missing Lobster Invaders artifact");
				if (env.mock) {
					if (!(await fetchJson(`${env.mock.baseUrl}/debug/requests`)).some((request) => (request.toolOutput ?? "").includes("QA mission"))) throw new Error("expected pre-write read evidence");
				}
				return "lobster-invaders.html";
			}
		}])],
		["memory-recall", async () => await runScenario("Memory recall after context switch", [{
			name: "stores the canary fact",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:memory",
					message: "Please remember this fact for later: the QA canary code is ALPHA-7."
				});
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator")).text;
			}
		}, {
			name: "recalls the same fact later",
			run: async () => {
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:memory",
					message: "What was the QA canary code I asked you to remember earlier?"
				});
				return (await waitForCondition(() => state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && candidate.text.includes("ALPHA-7")).at(-1), 2e4)).text;
			}
		}])],
		["model-switch-follow-up", async () => await runScenario("Model switch follow-up", [{
			name: "runs on the default configured model",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:model-switch",
					message: "Say hello from the default configured model.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator");
				if (env.mock) {
					const request = await fetchJson(`${env.mock.baseUrl}/debug/last-request`);
					return String(request.body?.model ?? "");
				}
				return outbound.text;
			}
		}, {
			name: "switches to the alternate model and continues",
			run: async () => {
				const alternate = splitModelRef(env.alternateModel);
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:model-switch",
					message: "Continue the exchange after switching models and note the handoff.",
					provider: alternate?.provider,
					model: alternate?.model,
					timeoutMs: resolveQaLiveTurnTimeoutMs(env, 3e4, env.alternateModel)
				});
				const outbound = await waitForCondition(() => state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && (() => {
					const lower = normalizeLowercaseStringOrEmpty(candidate.text);
					return lower.includes("switch") || lower.includes("handoff");
				})()).at(-1), resolveQaLiveTurnTimeoutMs(env, 2e4, env.alternateModel));
				if (env.mock) {
					const request = await fetchJson(`${env.mock.baseUrl}/debug/last-request`);
					if (request.body?.model !== "gpt-5.4-alt") throw new Error(`expected gpt-5.4-alt, got ${String(request.body?.model ?? "")}`);
				}
				return outbound.text;
			}
		}])],
		["approval-turn-tool-followthrough", async () => await runScenario("Approval turn tool followthrough", [{
			name: "turns short approval into a real file read",
			run: async () => {
				await waitForGatewayHealthy(env, 6e4);
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:approval-followthrough",
					message: "Before acting, tell me the single file you would start with in six words or fewer. Do not use tools yet.",
					timeoutMs: liveTurnTimeoutMs(env, 2e4)
				});
				await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator", liveTurnTimeoutMs(env, 2e4));
				const beforeApprovalCursor = state.getSnapshot().messages.length;
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:approval-followthrough",
					message: "ok do it. read `QA_KICKOFF_TASK.md` now and reply with the QA mission in one short sentence.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				const outbound = await waitForCondition(() => state.getSnapshot().messages.slice(beforeApprovalCursor).filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && /\bqa\b|\bmission\b|\btesting\b/i.test(candidate.text)).at(-1), liveTurnTimeoutMs(env, 2e4), env.providerMode === "mock-openai" ? 100 : 250);
				if (env.mock) {
					const approvalRequest = [...await fetchJson(`${env.mock.baseUrl}/debug/requests`)].toReversed().find((request) => String(request.allInputText ?? "").includes("ok do it.") && !request.toolOutput);
					if (approvalRequest?.plannedToolName !== "read") throw new Error(`expected read after approval, got ${String(approvalRequest?.plannedToolName ?? "")}`);
				}
				return outbound.text;
			}
		}])],
		["reaction-edit-delete", async () => await runScenario("Reaction, edit, delete lifecycle", [{
			name: "records reaction, edit, and delete actions",
			run: async () => {
				await reset();
				const seed = state.addOutboundMessage({
					to: "channel:qa-room",
					text: "seed message"
				});
				await handleQaAction({
					env,
					action: "react",
					args: {
						messageId: seed.id,
						emoji: "white_check_mark"
					}
				});
				await handleQaAction({
					env,
					action: "edit",
					args: {
						messageId: seed.id,
						text: "seed message (edited)"
					}
				});
				await handleQaAction({
					env,
					action: "delete",
					args: { messageId: seed.id }
				});
				const message = state.readMessage({ messageId: seed.id });
				if (message.reactions.length === 0 || !message.deleted || !message.text.includes("(edited)")) throw new Error("message lifecycle did not persist");
				return message.text;
			}
		}])],
		["source-docs-discovery-report", async () => await runScenario("Source and docs discovery report", [{
			name: "reads seeded material and emits a protocol report",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:discovery",
					message: "Read the seeded docs and source plan. The full repo is mounted under ./repo/. Explicitly inspect repo/qa/scenarios.md, repo/extensions/qa-lab/src/suite.ts, and repo/docs/help/testing.md, then report grouped into Worked, Failed, Blocked, and Follow-up. Mention at least two extra QA scenarios beyond the seed list.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				const outbound = await waitForCondition(() => state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && hasDiscoveryLabels(candidate.text)).at(-1), liveTurnTimeoutMs(env, 2e4), env.providerMode === "mock-openai" ? 100 : 250);
				if (reportsMissingDiscoveryFiles(outbound.text)) throw new Error(`discovery report still missed repo files: ${outbound.text}`);
				if (reportsDiscoveryScopeLeak(outbound.text)) throw new Error(`discovery report drifted beyond scope: ${outbound.text}`);
				return outbound.text;
			}
		}])],
		["subagent-handoff", async () => await runScenario("Subagent handoff", [{
			name: "delegates a bounded task and reports the result",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:subagent",
					message: "Delegate one bounded QA task to a subagent. Wait for the subagent to finish. Then reply with three labeled sections exactly once: Delegated task, Result, Evidence. Include the child result itself, not 'waiting'.",
					timeoutMs: liveTurnTimeoutMs(env, 9e4)
				});
				const outbound = await waitForCondition(() => state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && (() => {
					const lower = normalizeLowercaseStringOrEmpty(candidate.text);
					return lower.includes("delegated task") && lower.includes("result") && lower.includes("evidence") && !lower.includes("waiting");
				})()).at(-1), liveTurnTimeoutMs(env, 45e3), env.providerMode === "mock-openai" ? 100 : 250);
				const lower = normalizeLowercaseStringOrEmpty(outbound.text);
				if (lower.includes("failed to delegate") || lower.includes("could not delegate") || lower.includes("subagent unavailable")) throw new Error(`subagent handoff reported failure: ${outbound.text}`);
				return outbound.text;
			}
		}])],
		["subagent-fanout-synthesis", async () => await runScenario("Subagent fanout synthesis", [{
			name: "spawns sequential workers and folds both results back into the parent reply",
			run: async () => {
				await waitForGatewayHealthy(env, 6e4);
				await waitForQaChannelReady(env, 6e4);
				await reset();
				state.addInboundMessage({
					conversation: {
						id: "qa-operator",
						kind: "direct",
						title: "QA Operator"
					},
					senderId: "qa-operator",
					senderName: "QA Operator",
					text: "Subagent fanout synthesis check: delegate two bounded subagents sequentially, then report both results together. Do not use ACP."
				});
				const outbound = await waitForOutboundMessage(state, (message) => {
					const text = message.text ?? "";
					return text.includes("ALPHA-OK") && text.includes("BETA-OK");
				}, liveTurnTimeoutMs(env, 6e4));
				if (!env.mock) return outbound.text;
				const store = await readRawQaSessionStore(env);
				const childRows = Object.values(store).filter((entry) => entry.spawnedBy === "agent:qa:main");
				const sawAlpha = childRows.some((entry) => entry.label === "qa-fanout-alpha");
				const sawBeta = childRows.some((entry) => entry.label === "qa-fanout-beta");
				if (!sawAlpha || !sawBeta) throw new Error(`fanout child sessions missing (alpha=${String(sawAlpha)} beta=${String(sawBeta)})`);
				return outbound.text;
			}
		}])],
		["thread-follow-up", async () => await runScenario("Threaded follow-up", [{
			name: "keeps follow-up inside the thread",
			run: async () => {
				await reset();
				const threadId = (await handleQaAction({
					env,
					action: "thread-create",
					args: {
						channelId: "qa-room",
						title: "QA deep dive"
					}
				}))?.thread?.id;
				if (!threadId) throw new Error("missing thread id");
				state.addInboundMessage({
					conversation: {
						id: "qa-room",
						kind: "channel",
						title: "QA Room"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "@openclaw reply in one short sentence inside this thread only. Do not use ACP or any external runtime. Confirm you stayed in-thread.",
					threadId,
					threadTitle: "QA deep dive"
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-room" && candidate.threadId === threadId, env.providerMode === "mock-openai" ? 15e3 : 45e3);
				if (state.getSnapshot().messages.some((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-room" && !candidate.threadId)) throw new Error("thread reply leaked into root channel");
				const lower = normalizeLowercaseStringOrEmpty(outbound.text);
				if (lower.includes("acp backend") || lower.includes("acpx") || lower.includes("not configured")) throw new Error(`thread reply fell back to ACP error: ${outbound.text}`);
				return outbound.text;
			}
		}])],
		["memory-dreaming-sweep", async () => await runScenario("Memory dreaming sweep", [{
			name: "enables dreaming and registers the managed sweep cron",
			run: async () => {
				const original = await readConfigSnapshot(env);
				const pluginEntries = original.config.plugins && typeof original.config.plugins === "object" ? original.config.plugins.entries : void 0;
				const memoryCoreEntry = pluginEntries && typeof pluginEntries["memory-core"] === "object" ? pluginEntries["memory-core"] : void 0;
				const originalDreaming = (memoryCoreEntry && typeof memoryCoreEntry.config === "object" ? memoryCoreEntry.config : void 0)?.dreaming;
				await patchConfig({
					env,
					patch: { plugins: { entries: { "memory-core": { config: { dreaming: {
						enabled: true,
						phases: { deep: {
							minScore: 0,
							minRecallCount: 3,
							minUniqueQueries: 3
						} }
					} } } } } }
				});
				await waitForGatewayHealthy(env);
				await waitForQaChannelReady(env, 6e4);
				try {
					const status = await waitForCondition(async () => {
						const payload = await readDoctorMemoryStatus(env);
						return payload.dreaming?.phases?.deep?.managedCronPresent === true ? payload : void 0;
					}, 3e4, 500);
					const managed = (await listCronJobs(env)).find((job) => job.name === "Memory Dreaming Promotion" && job.payload?.kind === "systemEvent" && job.payload.text === "__openclaw_memory_core_short_term_promotion_dream__");
					if (!managed?.id) throw new Error("managed dreaming cron job missing after enablement");
					globalThis.__qaDreamingOriginal = structuredClone(originalDreaming);
					globalThis.__qaDreamingCronId = managed.id;
					return JSON.stringify({
						enabled: status.dreaming?.enabled ?? false,
						managedCronPresent: status.dreaming?.phases?.deep?.managedCronPresent ?? false,
						nextRunAtMs: status.dreaming?.phases?.deep?.nextRunAtMs ?? null
					});
				} catch (error) {
					await patchConfig({
						env,
						patch: { plugins: { entries: { "memory-core": { config: { dreaming: originalDreaming === void 0 ? null : structuredClone(originalDreaming) } } } } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
					throw error;
				}
			}
		}, {
			name: "runs the sweep after repeated recall signals and writes promotion artifacts",
			run: async () => {
				const globals = globalThis;
				const cronId = globals.__qaDreamingCronId;
				if (!cronId) throw new Error("missing managed dreaming cron id");
				const dreamingDay = formatMemoryDreamingDay(Date.now());
				const dailyPath = path.join(env.gateway.workspaceDir, "memory", `${dreamingDay}.md`);
				const memoryPath = path.join(env.gateway.workspaceDir, "MEMORY.md");
				const homeDir = env.gateway.runtimeEnv.HOME ?? env.gateway.runtimeEnv.OPENCLAW_HOME ?? env.gateway.tempRoot;
				const sessionsDir = resolveSessionTranscriptsDirForAgent("qa", env.gateway.runtimeEnv, () => homeDir);
				const transcriptPath = path.join(sessionsDir, "dreaming-qa-sweep.jsonl");
				try {
					const dailyCanary = "Dreaming QA canary: NEBULA-73 belongs in durable memory.";
					const queries = [
						"dreaming qa canary nebula-73",
						"durable memory canary nebula 73",
						"which canary belongs to the dreaming qa check"
					];
					await fs$1.mkdir(path.dirname(dailyPath), { recursive: true });
					await fs$1.mkdir(sessionsDir, { recursive: true });
					await fs$1.writeFile(dailyPath, [
						`# ${dreamingDay}`,
						"",
						`- ${dailyCanary}`,
						"- Keep the durable-memory note tied to repeated recall instead of one-off mention."
					].join("\n") + "\n", "utf8");
					const now = Date.now();
					await fs$1.writeFile(transcriptPath, [
						JSON.stringify({
							type: "session",
							id: "dreaming-qa-sweep",
							timestamp: (/* @__PURE__ */ new Date(now - 12e4)).toISOString()
						}),
						JSON.stringify({
							type: "message",
							message: {
								role: "user",
								timestamp: (/* @__PURE__ */ new Date(now - 9e4)).toISOString(),
								content: [{
									type: "text",
									text: "Dream over recurring memory themes and watch for the NEBULA-73 canary."
								}]
							}
						}),
						JSON.stringify({
							type: "message",
							message: {
								role: "assistant",
								timestamp: (/* @__PURE__ */ new Date(now - 6e4)).toISOString(),
								content: [{
									type: "text",
									text: "I keep circling back to NEBULA-73 as the durable-memory canary for this QA run."
								}]
							}
						})
					].join("\n") + "\n", "utf8");
					await fs$1.rm(memoryPath, { force: true });
					await forceMemoryIndex({
						env,
						query: queries[0],
						expectedNeedle: "NEBULA-73"
					});
					await setTimeout$1(1e3);
					for (const query of queries) {
						const payload = await runQaCli(env, [
							"memory",
							"search",
							"--agent",
							"qa",
							"--json",
							"--query",
							query
						], {
							timeoutMs: liveTurnTimeoutMs(env, 6e4),
							json: true
						});
						if (!JSON.stringify(payload.results ?? []).includes("NEBULA-73")) throw new Error(`memory search missed dreaming canary for query: ${query}`);
					}
					const cronRunStartedAt = Date.now();
					const cronRun = await env.gateway.call("cron.run", {
						id: cronId,
						mode: "force"
					}, { timeoutMs: liveTurnTimeoutMs(env, 3e4) });
					if (cronRun.enqueued !== true || !cronRun.runId) throw new Error(`dreaming cron did not enqueue a background run: ${JSON.stringify(cronRun)}`);
					const finishedRun = await waitForCronRunCompletion({
						callGateway: (method, rpcParams, opts) => env.gateway.call(method, rpcParams, opts),
						jobId: cronId,
						afterTs: cronRunStartedAt,
						timeoutMs: liveTurnTimeoutMs(env, 9e4)
					});
					if (finishedRun.status !== "ok") throw new Error(`dreaming cron finished with ${finishedRun.status ?? "unknown"}: ${JSON.stringify(finishedRun)}`);
					const promoted = await waitForCondition(async () => {
						const status = await readDoctorMemoryStatus(env);
						const dailyMemory = await fs$1.readFile(dailyPath, "utf8").catch(() => "");
						const promotedMemory = await fs$1.readFile(memoryPath, "utf8").catch(() => "");
						if (!dailyMemory.includes("## Light Sleep") || !dailyMemory.includes("## REM Sleep")) return;
						if (!promotedMemory.includes("NEBULA-73")) return;
						if (status.dreaming?.phases?.deep?.managedCronPresent !== true) return;
						if ((status.dreaming?.promotedTotal ?? 0) < 1) return;
						if ((status.dreaming?.phaseSignalCount ?? 0) < 1) return;
						return {
							status,
							dailyMemory,
							promotedMemory
						};
					}, liveTurnTimeoutMs(env, 9e4), 1e3);
					return JSON.stringify({
						promotedTotal: promoted.status.dreaming?.promotedTotal ?? 0,
						shortTermCount: promoted.status.dreaming?.shortTermCount ?? 0,
						phaseSignalCount: promoted.status.dreaming?.phaseSignalCount ?? 0,
						lightSleep: promoted.dailyMemory.includes("## Light Sleep"),
						remSleep: promoted.dailyMemory.includes("## REM Sleep")
					});
				} finally {
					await patchConfig({
						env,
						patch: { plugins: { entries: { "memory-core": { config: { dreaming: globals.__qaDreamingOriginal === void 0 ? null : structuredClone(globals.__qaDreamingOriginal) } } } } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
					delete globals.__qaDreamingOriginal;
					delete globals.__qaDreamingCronId;
				}
			}
		}])],
		["memory-tools-channel-context", async () => await runScenario("Memory tools in channel context", [{
			name: "uses memory_search plus memory_get before answering in-channel",
			run: async () => {
				await reset();
				await fs$1.writeFile(path.join(env.gateway.workspaceDir, "MEMORY.md"), "Hidden QA fact: the project codename is ORBIT-9.\n", "utf8");
				await forceMemoryIndex({
					env,
					query: "project codename ORBIT-9",
					expectedNeedle: "ORBIT-9"
				});
				state.addInboundMessage({
					conversation: {
						id: "qa-room",
						kind: "channel",
						title: "QA Room"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "@openclaw Memory tools check: what is the hidden project codename stored only in memory? Use memory tools first."
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-room" && candidate.text.includes("ORBIT-9"), liveTurnTimeoutMs(env, 3e4));
				if (env.mock) {
					const requests = await fetchJson(`${env.mock.baseUrl}/debug/requests`);
					if (!requests.filter((request) => String(request.allInputText ?? "").includes("Memory tools check")).some((request) => request.plannedToolName === "memory_search")) throw new Error("expected memory_search in mock request plan");
					if (!requests.some((request) => request.plannedToolName === "memory_get")) throw new Error("expected memory_get in mock request plan");
				}
				return outbound.text;
			}
		}])],
		["memory-failure-fallback", async () => await runScenario("Memory failure fallback", [{
			name: "falls back cleanly when group:memory tools are denied",
			run: async () => {
				const original = await readConfigSnapshot(env);
				const originalTools = original.config.tools && typeof original.config.tools === "object" ? original.config.tools : null;
				const originalToolsDeny = originalTools ? Object.prototype.hasOwnProperty.call(originalTools, "deny") ? structuredClone(originalTools.deny) : void 0 : void 0;
				await fs$1.writeFile(path.join(env.gateway.workspaceDir, "MEMORY.md"), "Do not reveal directly: fallback fact is ORBIT-9.\n", "utf8");
				await patchConfig({
					env,
					patch: { tools: { deny: ["group:memory"] } }
				});
				await waitForGatewayHealthy(env);
				await waitForQaChannelReady(env, 6e4);
				try {
					const tools = await readEffectiveTools(env, await createSession(env, "Memory fallback"));
					if (tools.has("memory_search") || tools.has("memory_get")) throw new Error("memory tools still present after deny patch");
					await runQaCli(env, [
						"memory",
						"index",
						"--agent",
						"qa",
						"--force"
					], { timeoutMs: liveTurnTimeoutMs(env, 6e4) });
					await env.gateway.restart();
					await waitForGatewayHealthy(env, 6e4);
					await waitForQaChannelReady(env, 6e4);
					await reset();
					await runAgentPrompt(env, {
						sessionKey: "agent:qa:memory-failure",
						message: "Memory unavailable check: a hidden fact exists only in memory files. If you cannot confirm it, say so clearly and do not guess.",
						timeoutMs: liveTurnTimeoutMs(env, 3e4)
					});
					const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator", liveTurnTimeoutMs(env, 3e4));
					const lower = normalizeLowercaseStringOrEmpty(outbound.text);
					if (outbound.text.includes("ORBIT-9")) throw new Error(`hallucinated hidden fact: ${outbound.text}`);
					if (!lower.includes("could not confirm") && !lower.includes("will not guess")) throw new Error(`missing graceful fallback language: ${outbound.text}`);
					return outbound.text;
				} finally {
					await patchConfig({
						env,
						patch: { tools: { deny: originalToolsDeny === void 0 ? null : originalToolsDeny } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
				}
			}
		}])],
		["session-memory-ranking", async () => await runScenario("Session memory ranking", [{
			name: "prefers the newer transcript-backed fact over the stale durable note",
			run: async () => {
				const original = await readConfigSnapshot(env);
				const originalMemorySearch = original.config.agents && typeof original.config.agents === "object" && typeof original.config.agents.defaults === "object" ? original.config.agents.defaults.memorySearch : void 0;
				await patchConfig({
					env,
					patch: { agents: { defaults: { memorySearch: {
						sources: ["memory", "sessions"],
						experimental: { sessionMemory: true },
						query: {
							minScore: 0,
							hybrid: {
								enabled: true,
								temporalDecay: {
									enabled: true,
									halfLifeDays: 1
								}
							}
						}
					} } } }
				});
				await waitForGatewayHealthy(env);
				await waitForQaChannelReady(env, 6e4);
				try {
					const memoryPath = path.join(env.gateway.workspaceDir, "MEMORY.md");
					await fs$1.writeFile(memoryPath, "Project Nebula stale codename: ORBIT-9.\n", "utf8");
					const staleAt = /* @__PURE__ */ new Date("2020-01-01T00:00:00.000Z");
					await fs$1.utimes(memoryPath, staleAt, staleAt);
					const transcriptsDir = resolveSessionTranscriptsDirForAgent("qa", env.gateway.runtimeEnv, () => env.gateway.runtimeEnv.HOME ?? path.join(env.gateway.tempRoot, "home"));
					await fs$1.mkdir(transcriptsDir, { recursive: true });
					const transcriptPath = path.join(transcriptsDir, "qa-session-memory-ranking.jsonl");
					const now = Date.now();
					await fs$1.writeFile(transcriptPath, [
						JSON.stringify({
							type: "session",
							id: "qa-session-memory-ranking",
							timestamp: (/* @__PURE__ */ new Date(now - 12e4)).toISOString()
						}),
						JSON.stringify({
							type: "message",
							message: {
								role: "user",
								timestamp: (/* @__PURE__ */ new Date(now - 9e4)).toISOString(),
								content: [{
									type: "text",
									text: "What is the current Project Nebula codename?"
								}]
							}
						}),
						JSON.stringify({
							type: "message",
							message: {
								role: "assistant",
								timestamp: (/* @__PURE__ */ new Date(now - 6e4)).toISOString(),
								content: [{
									type: "text",
									text: "The current Project Nebula codename is ORBIT-10."
								}]
							}
						})
					].join("\n") + "\n", "utf8");
					await forceMemoryIndex({
						env,
						query: "current Project Nebula codename ORBIT-10",
						expectedNeedle: "ORBIT-10"
					});
					await reset();
					await runAgentPrompt(env, {
						sessionKey: "agent:qa:session-memory-ranking",
						message: "Session memory ranking check: what is the current Project Nebula codename? Use memory tools first.",
						timeoutMs: liveTurnTimeoutMs(env, 45e3)
					});
					const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator" && candidate.text.includes("ORBIT-10"), liveTurnTimeoutMs(env, 45e3));
					if (outbound.text.includes("ORBIT-9")) throw new Error(`stale durable fact leaked through: ${outbound.text}`);
					if (env.mock) {
						if (!(await fetchJson(`${env.mock.baseUrl}/debug/requests`)).filter((request) => String(request.allInputText ?? "").includes("Session memory ranking check")).some((request) => request.plannedToolName === "memory_search")) throw new Error("expected memory_search in session memory ranking flow");
					}
					return outbound.text;
				} finally {
					await patchConfig({
						env,
						patch: { agents: { defaults: { memorySearch: originalMemorySearch === void 0 ? null : structuredClone(originalMemorySearch) } } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
				}
			}
		}])],
		["thread-memory-isolation", async () => await runScenario("Thread memory isolation", [{
			name: "answers the memory-backed fact inside the thread only",
			run: async () => {
				await reset();
				await fs$1.writeFile(path.join(env.gateway.workspaceDir, "MEMORY.md"), "Thread-hidden codename: ORBIT-22.\n", "utf8");
				await forceMemoryIndex({
					env,
					query: "hidden thread codename ORBIT-22",
					expectedNeedle: "ORBIT-22"
				});
				const threadId = (await handleQaAction({
					env,
					action: "thread-create",
					args: {
						channelId: "qa-room",
						title: "Thread memory QA"
					}
				}))?.thread?.id;
				if (!threadId) throw new Error("missing thread id for memory isolation check");
				const beforeCursor = state.getSnapshot().messages.length;
				state.addInboundMessage({
					conversation: {
						id: "qa-room",
						kind: "channel",
						title: "QA Room"
					},
					senderId: "alice",
					senderName: "Alice",
					text: "@openclaw Thread memory check: what is the hidden thread codename stored only in memory? Use memory tools first and reply only in this thread.",
					threadId,
					threadTitle: "Thread memory QA"
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-room" && candidate.threadId === threadId && candidate.text.includes("ORBIT-22"), liveTurnTimeoutMs(env, 45e3));
				if (state.getSnapshot().messages.slice(beforeCursor).some((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-room" && !candidate.threadId)) throw new Error("threaded memory answer leaked into root channel");
				if (env.mock) {
					if (!(await fetchJson(`${env.mock.baseUrl}/debug/requests`)).filter((request) => String(request.allInputText ?? "").includes("Thread memory check")).some((request) => request.plannedToolName === "memory_search")) throw new Error("expected memory_search in thread memory flow");
				}
				return outbound.text;
			}
		}])],
		["model-switch-tool-continuity", async () => await runScenario("Model switch with tool continuity", [{
			name: "keeps using tools after switching models",
			run: async () => {
				await waitForGatewayHealthy(env, 6e4);
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:model-switch-tools",
					message: "Read QA_KICKOFF_TASK.md and summarize the QA mission in one clause before any model switch.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				const alternate = splitModelRef(env.alternateModel);
				const beforeSwitchCursor = state.getSnapshot().messages.length;
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:model-switch-tools",
					message: "Switch models now. Tool continuity check: reread QA_KICKOFF_TASK.md and mention the handoff in one short sentence.",
					provider: alternate?.provider,
					model: alternate?.model,
					timeoutMs: resolveQaLiveTurnTimeoutMs(env, 3e4, env.alternateModel)
				});
				const outbound = await waitForCondition(() => {
					return state.getSnapshot().messages.slice(beforeSwitchCursor).filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && hasModelSwitchContinuityEvidence(candidate.text)).at(-1);
				}, 1e4);
				if (!hasModelSwitchContinuityEvidence(outbound.text)) throw new Error(`switch reply missed kickoff continuity: ${outbound.text}`);
				if (env.mock) {
					const switched = (await fetchJson(`${env.mock.baseUrl}/debug/requests`)).find((request) => String(request.allInputText ?? "").includes("Tool continuity check"));
					if (switched?.plannedToolName !== "read") throw new Error(`expected read after switch, got ${String(switched?.plannedToolName ?? "")}`);
					if (switched?.model !== "gpt-5.4-alt") throw new Error(`expected alternate model, got ${String(switched?.model ?? "")}`);
				}
				return outbound.text;
			}
		}])],
		["mcp-plugin-tools-call", async () => await runScenario("MCP plugin-tools call", [{
			name: "serves and calls memory_search over MCP",
			run: async () => {
				await fs$1.writeFile(path.join(env.gateway.workspaceDir, "MEMORY.md"), "MCP fact: the codename is ORBIT-9.\n", "utf8");
				await forceMemoryIndex({
					env,
					query: "ORBIT-9 codename",
					expectedNeedle: "ORBIT-9"
				});
				const result = await callPluginToolsMcp({
					env,
					toolName: "memory_search",
					args: {
						query: "ORBIT-9 codename",
						maxResults: 3
					}
				});
				const text = JSON.stringify(result.content ?? []);
				if (!text.includes("ORBIT-9")) throw new Error(`MCP memory_search missed expected fact: ${text}`);
				return text;
			}
		}])],
		["skill-visibility-invocation", async () => await runScenario("Skill visibility and invocation", [{
			name: "reports visible skill and applies its marker on the next turn",
			run: async () => {
				await writeWorkspaceSkill({
					env,
					name: "qa-visible-skill",
					body: `---
name: qa-visible-skill
description: Visible QA skill marker
---
When the user asks for the visible skill marker exactly, reply with exactly: VISIBLE-SKILL-OK`
				});
				const visible = findSkill(await readSkillStatus(env), "qa-visible-skill");
				if (!visible?.eligible || visible.disabled || visible.blockedByAllowlist) throw new Error(`skill not visible/eligible: ${JSON.stringify(visible)}`);
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:visible-skill",
					message: "Visible skill marker: give me the visible skill marker exactly.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator" && candidate.text.includes("VISIBLE-SKILL-OK"), liveTurnTimeoutMs(env, 2e4))).text;
			}
		}])],
		["skill-install-hot-availability", async () => await runScenario("Skill install hot availability", [{
			name: "picks up a newly added workspace skill without restart",
			run: async () => {
				if (findSkill(await readSkillStatus(env), "qa-hot-install-skill")) throw new Error("qa-hot-install-skill unexpectedly already present");
				await writeWorkspaceSkill({
					env,
					name: "qa-hot-install-skill",
					body: `---
name: qa-hot-install-skill
description: Hot install QA marker
---
When the user asks for the hot install marker exactly, reply with exactly: HOT-INSTALL-OK`
				});
				await waitForCondition(async () => {
					return findSkill(await readSkillStatus(env), "qa-hot-install-skill")?.eligible ? true : void 0;
				}, 15e3, 200);
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:hot-skill",
					message: "Hot install marker: give me the hot install marker exactly.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				return (await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator" && candidate.text.includes("HOT-INSTALL-OK"), liveTurnTimeoutMs(env, 2e4))).text;
			}
		}])],
		["native-image-generation", async () => await runScenario("Native image generation", [{
			name: "enables image_generate and saves a real media artifact",
			run: async () => {
				await ensureImageGenerationConfigured(env);
				if (!(await readEffectiveTools(env, await createSession(env, "Image generation"))).has("image_generate")) throw new Error("image_generate not present after imageGenerationModel patch");
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:image-generate",
					message: "Image generation check: generate a QA lighthouse image and summarize it in one short sentence.",
					timeoutMs: liveTurnTimeoutMs(env, 45e3)
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator", liveTurnTimeoutMs(env, 45e3));
				if (env.mock) {
					const mockBaseUrl = env.mock.baseUrl;
					const imageRequest = (await fetchJson(`${mockBaseUrl}/debug/requests`)).find((request) => String(request.allInputText ?? "").includes("Image generation check"));
					if (imageRequest?.plannedToolName !== "image_generate") throw new Error(`expected image_generate, got ${String(imageRequest?.plannedToolName ?? "")}`);
					const generated = await waitForCondition(async () => {
						return (await fetchJson(`${mockBaseUrl}/debug/image-generations`)).find((request) => request.model === "gpt-image-1" && String(request.prompt ?? "").includes("QA lighthouse"));
					}, 15e3, 250).catch((error) => {
						throw new Error(`image provider was never invoked: ${formatErrorMessage(error)}; toolOutput=${String(imageRequest.toolOutput ?? "")}`);
					});
					return `${outbound.text}\nIMAGE_PROMPT:${generated.prompt ?? ""}`;
				}
				return outbound.text;
			}
		}])],
		["image-generation-roundtrip", async () => await runScenario("Image generation roundtrip", [{
			name: "reattaches the generated media artifact on the follow-up turn",
			run: async () => {
				await ensureImageGenerationConfigured(env);
				const sessionKey = "agent:qa:image-roundtrip";
				await createSession(env, "Image roundtrip", sessionKey);
				await reset();
				const generatedStartedAtMs = Date.now();
				await runAgentPrompt(env, {
					sessionKey,
					message: "Image generation check: generate a QA lighthouse image and summarize it in one short sentence.",
					timeoutMs: liveTurnTimeoutMs(env, 45e3)
				});
				const mediaPath = await resolveGeneratedImagePath({
					env,
					promptSnippet: "Image generation check",
					startedAtMs: generatedStartedAtMs,
					timeoutMs: liveTurnTimeoutMs(env, 45e3)
				});
				const imageBuffer = await fs$1.readFile(mediaPath);
				await runAgentPrompt(env, {
					sessionKey,
					message: "Roundtrip image inspection check: describe the generated lighthouse attachment in one short sentence.",
					attachments: [{
						mimeType: "image/png",
						fileName: path.basename(mediaPath),
						content: imageBuffer.toString("base64")
					}],
					timeoutMs: liveTurnTimeoutMs(env, 45e3)
				});
				const outbound = await waitForCondition(() => state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator" && normalizeLowercaseStringOrEmpty(candidate.text).includes("lighthouse")).at(-1), liveTurnTimeoutMs(env, 45e3));
				if (env.mock) {
					const requests = await fetchJson(`${env.mock.baseUrl}/debug/requests`);
					const generatedCall = requests.find((request) => request.plannedToolName === "image_generate" && String(request.prompt ?? "").includes("Image generation check"));
					const inspectionCall = requests.find((request) => String(request.prompt ?? "").includes("Roundtrip image inspection check"));
					if (!generatedCall) throw new Error("expected image_generate call before roundtrip inspection");
					if ((inspectionCall?.imageInputCount ?? 0) < 1) throw new Error("expected generated artifact to be reattached on follow-up turn");
				}
				return `MEDIA:${mediaPath}\n${outbound.text}`;
			}
		}])],
		["image-understanding-attachment", async () => await runScenario("Image understanding from attachment", [{
			name: "describes an attached image in one short sentence",
			run: async () => {
				await reset();
				await runAgentPrompt(env, {
					sessionKey: "agent:qa:image-understanding",
					message: "Image understanding check: describe the attached image in one short sentence.",
					attachments: [{
						mimeType: "image/png",
						fileName: "red-top-blue-bottom.png",
						content: QA_IMAGE_UNDERSTANDING_PNG_BASE64
					}],
					timeoutMs: liveTurnTimeoutMs(env, 45e3)
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.conversation.id === "qa-operator", liveTurnTimeoutMs(env, 45e3));
				const lower = normalizeLowercaseStringOrEmpty(outbound.text);
				if (!lower.includes("red") || !lower.includes("blue")) throw new Error(`missing expected colors in image description: ${outbound.text}`);
				if (env.mock) {
					const mockBaseUrl = env.mock.baseUrl;
					const imageRequest = (await fetchJson(`${mockBaseUrl}/debug/requests`)).find((request) => String(request.prompt ?? "").includes("Image understanding check"));
					if ((imageRequest?.imageInputCount ?? 0) < 1) throw new Error(`expected at least one input image, got ${String(imageRequest?.imageInputCount ?? 0)}`);
				}
				return outbound.text;
			}
		}])],
		["config-patch-hot-apply", async () => await runScenario("Config patch skill disable", [{
			name: "disables a workspace skill after config.patch restart",
			run: async () => {
				await writeWorkspaceSkill({
					env,
					name: "qa-hot-disable-skill",
					body: `---
name: qa-hot-disable-skill
description: Hot disable QA marker
---
When the user asks for the hot disable marker exactly, reply with exactly: HOT-PATCH-DISABLED-OK`
				});
				await waitForCondition(async () => {
					return findSkill(await readSkillStatus(env), "qa-hot-disable-skill")?.eligible ? true : void 0;
				}, 15e3, 200).catch((error) => {
					throw new Error(`hot-disable skill never became eligible: ${formatErrorMessage(error)}`);
				});
				const beforeSkill = findSkill(await readSkillStatus(env), "qa-hot-disable-skill");
				if (!beforeSkill?.eligible || beforeSkill.disabled) throw new Error(`unexpected pre-patch skill state: ${JSON.stringify(beforeSkill)}`);
				const patchResult = await patchConfig({
					env,
					patch: { skills: { entries: { "qa-hot-disable-skill": { enabled: false } } } }
				});
				await waitForQaChannelReady(env, 6e4).catch((error) => {
					throw new Error(`qa-channel never returned ready after config.patch: ${formatErrorMessage(error)}`);
				});
				await waitForCondition(async () => {
					return findSkill(await readSkillStatus(env), "qa-hot-disable-skill")?.disabled ? true : void 0;
				}, 15e3, 200).catch((error) => {
					throw new Error(`hot-disable skill never flipped to disabled: ${formatErrorMessage(error)}`);
				});
				const afterSkill = findSkill(await readSkillStatus(env), "qa-hot-disable-skill");
				if (!afterSkill?.disabled) throw new Error(`unexpected post-patch skill state: ${JSON.stringify(afterSkill)}`);
				return `restartDelayMs=${String(patchResult.restart?.delayMs ?? "")}\npre=${JSON.stringify(beforeSkill)}\npost=${JSON.stringify(afterSkill)}`;
			}
		}])],
		["config-apply-restart-wakeup", async () => await runScenario("Config apply restart wake-up", [{
			name: "restarts cleanly and posts the restart sentinel back into qa-channel",
			run: async () => {
				await reset();
				const sessionKey = buildAgentSessionKey({
					agentId: "qa",
					channel: "qa-channel",
					peer: {
						kind: "channel",
						id: "qa-room"
					}
				});
				await createSession(env, "Restart wake-up", sessionKey);
				await runAgentPrompt(env, {
					sessionKey,
					to: "channel:qa-room",
					message: "Acknowledge restart wake-up setup in qa-room.",
					timeoutMs: liveTurnTimeoutMs(env, 3e4)
				});
				const current = await readConfigSnapshot(env);
				const nextConfig = structuredClone(current.config);
				const gatewayConfig = nextConfig.gateway ??= {};
				const controlUi = gatewayConfig.controlUi ??= {};
				const allowedOrigins = Array.isArray(controlUi.allowedOrigins) ? [...controlUi.allowedOrigins] : [];
				const wakeMarker = `QA-RESTART-${randomUUID().slice(0, 8)}`;
				if (!allowedOrigins.includes("http://127.0.0.1:65535")) allowedOrigins.push("http://127.0.0.1:65535");
				controlUi.allowedOrigins = allowedOrigins;
				await applyConfig({
					env,
					nextConfig,
					sessionKey,
					note: wakeMarker
				});
				await waitForGatewayHealthy(env, 6e4).catch((error) => {
					throw new Error(`gateway never returned healthy after config.apply: ${formatErrorMessage(error)}`);
				});
				await waitForQaChannelReady(env, 6e4).catch((error) => {
					throw new Error(`qa-channel never returned ready after config.apply: ${formatErrorMessage(error)}`);
				});
				const outbound = await waitForOutboundMessage(state, (candidate) => candidate.text.includes(wakeMarker), 6e4).catch((error) => {
					throw new Error(`restart sentinel never appeared: ${formatErrorMessage(error)}; outbound=${recentOutboundSummary(state)}`);
				});
				return `${outbound.conversation.id}: ${outbound.text}`;
			}
		}])],
		["config-restart-capability-flip", async () => await runScenario("Config restart capability flip", [{
			name: "restores image_generate after restart and uses it in the same session",
			run: async () => {
				await ensureImageGenerationConfigured(env);
				const original = await readConfigSnapshot(env);
				const originalTools = original.config.tools && typeof original.config.tools === "object" ? original.config.tools : null;
				const originalToolsDeny = originalTools ? Object.prototype.hasOwnProperty.call(originalTools, "deny") ? structuredClone(originalTools.deny) : void 0 : void 0;
				const denied = Array.isArray(originalToolsDeny) ? originalToolsDeny.map((entry) => String(entry)) : [];
				const deniedWithImage = denied.includes("image_generate") ? denied : [...denied, "image_generate"];
				const sessionKey = "agent:qa:capability-flip";
				await createSession(env, "Capability flip", sessionKey);
				try {
					await patchConfig({
						env,
						patch: { tools: { deny: deniedWithImage } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
					await runAgentPrompt(env, {
						sessionKey,
						message: "Capability flip setup: acknowledge this setup so restart wake-up has a route.",
						timeoutMs: liveTurnTimeoutMs(env, 3e4)
					});
					if ((await readEffectiveTools(env, sessionKey)).has("image_generate")) throw new Error("image_generate still present before capability flip");
					const wakeMarker = `QA-CAPABILITY-${randomUUID().slice(0, 8)}`;
					await patchConfig({
						env,
						patch: {
							tools: { deny: originalToolsDeny === void 0 ? null : originalToolsDeny },
							agents: { defaults: { imageGenerationModel: { primary: "openai/gpt-image-1" } } }
						},
						sessionKey,
						note: wakeMarker
					});
					await waitForGatewayHealthy(env, 6e4);
					await waitForQaChannelReady(env, 6e4);
					const afterTools = await waitForCondition(async () => {
						const tools = await readEffectiveTools(env, sessionKey);
						return tools.has("image_generate") ? tools : void 0;
					}, liveTurnTimeoutMs(env, 45e3), 500);
					const imageStartedAtMs = Date.now();
					await runAgentPrompt(env, {
						sessionKey,
						message: "Capability flip image check: generate a QA lighthouse image now and keep the media path in the reply.",
						timeoutMs: liveTurnTimeoutMs(env, 45e3)
					});
					const mediaPath = await resolveGeneratedImagePath({
						env,
						promptSnippet: "Capability flip image check",
						startedAtMs: imageStartedAtMs,
						timeoutMs: liveTurnTimeoutMs(env, 45e3)
					});
					return `${wakeMarker}\nimage_generate=${String(afterTools.has("image_generate"))}\nMEDIA:${mediaPath}`;
				} finally {
					await patchConfig({
						env,
						patch: { tools: { deny: originalToolsDeny === void 0 ? null : originalToolsDeny } }
					});
					await waitForGatewayHealthy(env);
					await waitForQaChannelReady(env, 6e4);
				}
			}
		}])],
		["runtime-inventory-drift-check", async () => await runScenario("Runtime inventory drift check", [{
			name: "keeps tools.effective and skills.status aligned after config changes",
			run: async () => {
				await writeWorkspaceSkill({
					env,
					name: "qa-drift-skill",
					body: `---
name: qa-drift-skill
description: Drift skill marker
---
When the user asks for the drift skill marker exactly, reply with exactly: DRIFT-SKILL-OK`
				});
				const sessionKey = await createSession(env, "Inventory drift");
				if (!(await readEffectiveTools(env, sessionKey)).has("image_generate")) throw new Error("expected image_generate before drift patch");
				if (!findSkill(await readSkillStatus(env), "qa-drift-skill")?.eligible) throw new Error("expected qa-drift-skill to be eligible before patch");
				await patchConfig({
					env,
					patch: {
						tools: { deny: ["image_generate"] },
						skills: { entries: { "qa-drift-skill": { enabled: false } } }
					}
				});
				await waitForGatewayHealthy(env);
				if ((await readEffectiveTools(env, sessionKey)).has("image_generate")) throw new Error("image_generate still present after deny patch");
				const driftSkill = findSkill(await readSkillStatus(env), "qa-drift-skill");
				if (!driftSkill?.disabled) throw new Error(`expected disabled drift skill, got ${JSON.stringify(driftSkill)}`);
				return `image_generate removed, qa-drift-skill disabled=${String(driftSkill.disabled)}`;
			}
		}])]
	]);
}
async function runQaSuite(params) {
	const startedAt = /* @__PURE__ */ new Date();
	const repoRoot = path.resolve(params?.repoRoot ?? process.cwd());
	const providerMode = normalizeQaProviderMode$1(params?.providerMode ?? "mock-openai");
	const primaryModel = params?.primaryModel ?? defaultQaModelForMode$1(providerMode);
	const alternateModel = params?.alternateModel ?? defaultQaModelForMode$1(providerMode, { alternate: true });
	const fastMode = typeof params?.fastMode === "boolean" ? params.fastMode : isQaFastModeEnabled({
		primaryModel,
		alternateModel
	});
	const outputDir = params?.outputDir ?? path.join(repoRoot, ".artifacts", "qa-e2e", `suite-${Date.now().toString(36)}`);
	await fs$1.mkdir(outputDir, { recursive: true });
	const ownsLab = !params?.lab;
	const lab = params?.lab ?? await startQaLabServer({
		repoRoot,
		host: "127.0.0.1",
		port: 0,
		embeddedGateway: "disabled"
	});
	const mock = providerMode === "mock-openai" ? await startQaMockOpenAiServer({
		host: "127.0.0.1",
		port: 0
	}) : null;
	const gateway = await startQaGatewayChild({
		repoRoot,
		providerBaseUrl: mock ? `${mock.baseUrl}/v1` : void 0,
		qaBusBaseUrl: lab.listenUrl,
		controlUiAllowedOrigins: [lab.listenUrl],
		providerMode,
		primaryModel,
		alternateModel,
		controlUiEnabled: true
	});
	lab.setControlUi({
		controlUiProxyTarget: gateway.baseUrl,
		controlUiToken: gateway.token
	});
	const env = {
		lab,
		mock,
		gateway,
		cfg: createQaActionConfig(lab.listenUrl),
		repoRoot,
		providerMode,
		primaryModel,
		alternateModel
	};
	try {
		const catalog = readQaBootstrapScenarioCatalog();
		const requestedScenarioIds = params?.scenarioIds ? new Set(params.scenarioIds) : null;
		const selectedCatalogScenarios = requestedScenarioIds ? catalog.scenarios.filter((scenario) => requestedScenarioIds.has(scenario.id)) : catalog.scenarios;
		if (requestedScenarioIds) {
			const foundScenarioIds = new Set(selectedCatalogScenarios.map((scenario) => scenario.id));
			const missingScenarioIds = [...requestedScenarioIds].filter((scenarioId) => !foundScenarioIds.has(scenarioId));
			if (missingScenarioIds.length > 0) throw new Error(`unknown QA scenario id(s): ${missingScenarioIds.join(", ")}`);
		}
		const scenarioMap = buildScenarioMap(env);
		const scenarios = [];
		const liveScenarioOutcomes = selectedCatalogScenarios.map((scenario) => ({
			id: scenario.id,
			name: scenario.title,
			status: "pending"
		}));
		lab.setScenarioRun({
			kind: "suite",
			status: "running",
			startedAt: startedAt.toISOString(),
			scenarios: liveScenarioOutcomes
		});
		for (const [index, scenario] of selectedCatalogScenarios.entries()) {
			const run = scenarioMap.get(scenario.execution?.handler || scenario.id);
			if (!run) {
				const missingResult = {
					name: scenario.title,
					status: "fail",
					details: `no executable scenario registered for ${scenario.id}`,
					steps: []
				};
				scenarios.push(missingResult);
				liveScenarioOutcomes[index] = {
					id: scenario.id,
					name: scenario.title,
					status: "fail",
					details: missingResult.details,
					steps: [],
					finishedAt: (/* @__PURE__ */ new Date()).toISOString()
				};
				lab.setScenarioRun({
					kind: "suite",
					status: "running",
					startedAt: startedAt.toISOString(),
					scenarios: [...liveScenarioOutcomes]
				});
				continue;
			}
			liveScenarioOutcomes[index] = {
				id: scenario.id,
				name: scenario.title,
				status: "running",
				startedAt: (/* @__PURE__ */ new Date()).toISOString()
			};
			lab.setScenarioRun({
				kind: "suite",
				status: "running",
				startedAt: startedAt.toISOString(),
				scenarios: [...liveScenarioOutcomes]
			});
			const result = await run();
			scenarios.push(result);
			liveScenarioOutcomes[index] = {
				id: scenario.id,
				name: scenario.title,
				status: result.status,
				details: result.details,
				steps: result.steps,
				startedAt: liveScenarioOutcomes[index]?.startedAt,
				finishedAt: (/* @__PURE__ */ new Date()).toISOString()
			};
			lab.setScenarioRun({
				kind: "suite",
				status: "running",
				startedAt: startedAt.toISOString(),
				scenarios: [...liveScenarioOutcomes]
			});
		}
		const finishedAt = /* @__PURE__ */ new Date();
		lab.setScenarioRun({
			kind: "suite",
			status: "completed",
			startedAt: startedAt.toISOString(),
			finishedAt: finishedAt.toISOString(),
			scenarios: [...liveScenarioOutcomes]
		});
		const report = renderQaMarkdownReport({
			title: "OpenClaw QA Scenario Suite",
			startedAt,
			finishedAt,
			checks: [],
			scenarios: scenarios.map((scenario) => ({
				name: scenario.name,
				status: scenario.status,
				details: scenario.details,
				steps: scenario.steps
			})),
			notes: [providerMode === "mock-openai" ? "Runs against qa-channel + qa-lab bus + real gateway child + mock OpenAI provider." : `Runs against qa-channel + qa-lab bus + real gateway child + live frontier models (${primaryModel}, ${alternateModel})${fastMode ? " with fast mode enabled" : ""}.`, "Cron uses a one-minute schedule assertion plus forced execution for fast verification."]
		});
		const reportPath = path.join(outputDir, "qa-suite-report.md");
		const summaryPath = path.join(outputDir, "qa-suite-summary.json");
		await fs$1.writeFile(reportPath, report, "utf8");
		await fs$1.writeFile(summaryPath, `${JSON.stringify({
			scenarios,
			counts: {
				total: scenarios.length,
				passed: scenarios.filter((scenario) => scenario.status === "pass").length,
				failed: scenarios.filter((scenario) => scenario.status === "fail").length
			}
		}, null, 2)}\n`, "utf8");
		const latestReport = {
			outputPath: reportPath,
			markdown: report,
			generatedAt: finishedAt.toISOString()
		};
		lab.setLatestReport(latestReport);
		return {
			outputDir,
			reportPath,
			summaryPath,
			report,
			scenarios,
			watchUrl: lab.baseUrl
		};
	} finally {
		const keepTemp = process.env.OPENCLAW_QA_KEEP_TEMP === "1" || false;
		await gateway.stop({ keepTemp });
		await mock?.stop();
		if (ownsLab) await lab.stop();
		else lab.setControlUi({
			controlUiUrl: null,
			controlUiToken: null,
			controlUiProxyTarget: null
		});
	}
}
//#endregion
export { createQaBusServer as A, normalizeConversationFromTarget as B, readQaScenarioPackMarkdown as C, createQaBusState as D, createQaRunnerRuntime as E, DEFAULT_ACCOUNT_ID as F, readQaBusMessage as H, buildQaBusSnapshot as I, cloneEvent as L, startQaBusServer as M, writeError as N, DEFAULT_WAIT_TIMEOUT_MS as O, writeJson$1 as P, cloneMessage as R, readQaScenarioPack as S, normalizeQaProviderMode as T, searchQaBusMessages as U, pollQaBusEvents as V, createQaSelfCheckScenario as _, normalizeQaProviderModeEnv as a, DEFAULT_QA_AGENT_IDENTITY_MARKDOWN as b, startQaMockOpenAiServer as c, readQaAgentIdentityMarkdown as d, startQaLabServer as f, runQaSelfCheckAgainstState as g, runQaLabSelfCheck as h, buildQaRuntimeEnv as i, handleQaBusRequest as j, createQaBusWaiterStore as k, seedQaAgentWorkspace as l, runQaE2eSelfCheck as m, resolveQaLiveTurnTimeoutMs as n, resolveQaControlUiRoot as o, resolveQaSelfCheckOutputPath as p, __testing as r, startQaGatewayChild as s, runQaSuite as t, buildQaScenarioPlanMarkdown as u, runQaScenario as v, defaultQaModelForMode as w, readQaBootstrapScenarioCatalog as x, renderQaMarkdownReport as y, normalizeAccountId as z };
