// openclaw/dist/video-generation-provider-BuxGmcEo.js
import { o as resolveProviderHttpRequestConfig, r as postJsonRequest, t as assertOkOrThrowHttpError } from "./shared-BkbRvjq1.js";
import "./provider-http-4dHXMdoD.js";
import { a as buildDashscopeVideoGenerationParameters, c as pollDashscopeVideoTaskUntilComplete, i as buildDashscopeVideoGenerationInput, n as DEFAULT_VIDEO_GENERATION_TIMEOUT_MS, o as downloadDashscopeGeneratedVideos, r as DEFAULT_VIDEO_RESOLUTION_TO_SIZE, s as extractDashscopeVideoUrls } from "./video-generation-CgzoQL_M.js";
import { t as isProviderApiKeyConfigured } from "./provider-auth-z3l1ppNs.js";
import { n as resolveApiKeyForProvider } from "./provider-auth-runtime-BTmYKp_r.js";
import "./models-BU6aYbF7.js";
//#region extensions/qwen/video-generation-provider.ts
const DEFAULT_QWEN_VIDEO_BASE_URL = "https://dashscope-intl.aliyuncs.com";
const DEFAULT_QWEN_VIDEO_MODEL = "wan2.6-t2v";
function resolveQwenVideoBaseUrl(req) {
	const direct = req.cfg?.models?.providers?.qwen?.baseUrl?.trim();
	if (!direct) return DEFAULT_QWEN_VIDEO_BASE_URL;
	try {
		return new URL(direct).toString();
	} catch {
		return DEFAULT_QWEN_VIDEO_BASE_URL;
	}
}
function resolveDashscopeAigcApiBaseUrl(baseUrl) {
	try {
		const url = new URL(baseUrl);
		if (url.hostname === "coding-intl.dashscope.aliyuncs.com" || url.hostname === "coding.dashscope.aliyuncs.com" || url.hostname === "dashscope-intl.aliyuncs.com" || url.hostname === "dashscope.aliyuncs.com") return url.origin;
	} catch {}
	if (baseUrl.startsWith("https://dashscope.aliyuncs.com/compatible-mode/v1")) return "https://dashscope.aliyuncs.com";
	if (baseUrl.startsWith("https://dashscope-intl.aliyuncs.com/compatible-mode/v1")) return DEFAULT_QWEN_VIDEO_BASE_URL;
	return baseUrl.replace(/\/+$/u, "");
}
function buildQwenVideoGenerationProvider() {
	return {
		id: "qwen",
		label: "Qwen Cloud",
		defaultModel: DEFAULT_QWEN_VIDEO_MODEL,
		models: [
			"wan2.6-t2v",
			"wan2.6-i2v",
			"wan2.6-r2v",
			"wan2.6-r2v-flash",
			"wan2.7-r2v"
		],
		isConfigured: ({ agentDir }) => isProviderApiKeyConfigured({
			provider: "qwen",
			agentDir
		}),
		capabilities: {
			generate: {
				maxVideos: 1,
				maxDurationSeconds: 10,
				supportsSize: true,
				supportsAspectRatio: true,
				supportsResolution: true,
				supportsAudio: true,
				supportsWatermark: true
			},
			imageToVideo: {
				enabled: true,
				maxVideos: 1,
				maxInputImages: 1,
				maxDurationSeconds: 10,
				supportsSize: true,
				supportsAspectRatio: true,
				supportsResolution: true,
				supportsAudio: true,
				supportsWatermark: true
			},
			videoToVideo: {
				enabled: true,
				maxVideos: 1,
				maxInputVideos: 4,
				maxDurationSeconds: 10,
				supportsSize: true,
				supportsAspectRatio: true,
				supportsResolution: true,
				supportsAudio: true,
				supportsWatermark: true
			}
		},
		async generateVideo(req) {
			const fetchFn = fetch;
			const auth = await resolveApiKeyForProvider({
				provider: "qwen",
				cfg: req.cfg,
				agentDir: req.agentDir,
				store: req.authStore
			});
			if (!auth.apiKey) throw new Error("Qwen API key missing");
			const { baseUrl, allowPrivateNetwork, headers, dispatcherPolicy } = resolveProviderHttpRequestConfig({
				baseUrl: resolveQwenVideoBaseUrl(req),
				defaultBaseUrl: DEFAULT_QWEN_VIDEO_BASE_URL,
				defaultHeaders: {
					Authorization: `Bearer ${auth.apiKey}`,
					"Content-Type": "application/json",
					"X-DashScope-Async": "enable"
				},
				provider: "qwen",
				capability: "video",
				transport: "http"
			});
			const model = req.model?.trim() || DEFAULT_QWEN_VIDEO_MODEL;
			const { response, release } = await postJsonRequest({
				url: `${resolveDashscopeAigcApiBaseUrl(baseUrl)}/api/v1/services/aigc/video-generation/video-synthesis`,
				headers,
				body: {
					model,
					input: buildDashscopeVideoGenerationInput({
						providerLabel: "Qwen",
						req
					}),
					parameters: buildDashscopeVideoGenerationParameters({
						...req,
						durationSeconds: req.durationSeconds ?? 5
					}, DEFAULT_VIDEO_RESOLUTION_TO_SIZE)
				},
				timeoutMs: req.timeoutMs,
				fetchFn,
				allowPrivateNetwork,
				dispatcherPolicy
			});
			try {
				await assertOkOrThrowHttpError(response, "Qwen video generation failed");
				const submitted = await response.json();
				const taskId = submitted.output?.task_id?.trim();
				if (!taskId) throw new Error("Qwen video generation response missing task_id");
				const completed = await pollDashscopeVideoTaskUntilComplete({
					providerLabel: "Qwen",
					taskId,
					headers,
					timeoutMs: req.timeoutMs,
					fetchFn,
					baseUrl: resolveDashscopeAigcApiBaseUrl(baseUrl),
					defaultTimeoutMs: DEFAULT_VIDEO_GENERATION_TIMEOUT_MS
				});
				const urls = extractDashscopeVideoUrls(completed);
				if (urls.length === 0) throw new Error("Qwen video generation completed without output video URLs");
				return {
					videos: await downloadDashscopeGeneratedVideos({
						providerLabel: "Qwen",
						urls,
						timeoutMs: req.timeoutMs,
						fetchFn,
						defaultTimeoutMs: DEFAULT_VIDEO_GENERATION_TIMEOUT_MS
					}),
					model,
					metadata: {
						requestId: submitted.request_id,
						taskId,
						taskStatus: completed.output?.task_status
					}
				};
			} finally {
				await release();
			}
		}
	};
}
//#endregion
export { buildQwenVideoGenerationProvider as t };
