// openclaw/dist/extensions/brave/web-search-contract-api.js
import { a as setProviderWebSearchPluginConfigValue, i as resolveProviderWebSearchPluginConfig } from "../../web-search-provider-config-Bzw58qrQ.js";
import "../../provider-web-search-contract-BlcSOQwq.js";
//#region extensions/brave/web-search-contract-api.ts
function getTopLevelCredentialValue(searchConfig) {
	return searchConfig?.apiKey;
}
function setTopLevelCredentialValue(searchConfigTarget, value) {
	searchConfigTarget.apiKey = value;
}
function createBraveWebSearchProvider() {
	return {
		id: "brave",
		label: "Brave Search",
		hint: "Structured results · country/language/time filters",
		onboardingScopes: ["text-inference"],
		credentialLabel: "Brave Search API key",
		envVars: ["BRAVE_API_KEY"],
		placeholder: "BSA...",
		signupUrl: "https://brave.com/search/api/",
		docsUrl: "https://docs.openclaw.ai/brave-search",
		autoDetectOrder: 10,
		credentialPath: "plugins.entries.brave.config.webSearch.apiKey",
		inactiveSecretPaths: ["plugins.entries.brave.config.webSearch.apiKey"],
		getCredentialValue: getTopLevelCredentialValue,
		setCredentialValue: setTopLevelCredentialValue,
		getConfiguredCredentialValue: (config) => resolveProviderWebSearchPluginConfig(config, "brave")?.apiKey,
		setConfiguredCredentialValue: (configTarget, value) => {
			setProviderWebSearchPluginConfigValue(configTarget, "brave", "apiKey", value);
		},
		createTool: () => null
	};
}
//#endregion
export { createBraveWebSearchProvider };
