// openclaw/dist/extensions/duckduckgo/web-search-contract-api.js
import { t as enablePluginInConfig } from "../../provider-enable-config-Hm2ULiVx.js";
import { o as setScopedCredentialValue, t as getScopedCredentialValue } from "../../web-search-provider-config-Bzw58qrQ.js";
import "../../provider-web-search-contract-BlcSOQwq.js";
//#region extensions/duckduckgo/web-search-contract-api.ts
function createDuckDuckGoWebSearchProvider() {
	return {
		id: "duckduckgo",
		label: "DuckDuckGo Search (experimental)",
		hint: "Free web search fallback with no API key required",
		requiresCredential: false,
		envVars: [],
		placeholder: "(no key needed)",
		signupUrl: "https://duckduckgo.com/",
		docsUrl: "https://docs.openclaw.ai/tools/web",
		autoDetectOrder: 100,
		credentialPath: "",
		inactiveSecretPaths: [],
		getCredentialValue: (searchConfig) => getScopedCredentialValue(searchConfig, "duckduckgo"),
		setCredentialValue: (searchConfigTarget, value) => setScopedCredentialValue(searchConfigTarget, "duckduckgo", value),
		applySelectionConfig: (config) => enablePluginInConfig(config, "duckduckgo").config,
		createTool: () => null
	};
}
//#endregion
export { createDuckDuckGoWebSearchProvider };
